<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  
  <meta name="renderer" content="webkit">
  <meta http-equiv="X-UA-Compatible" content="IE=edge" >
  <link rel="dns-prefetch" href="http://yoursite.com">
  <title>Weber Blog</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  <meta property="og:type" content="website">
<meta property="og:title" content="Weber Blog">
<meta property="og:url" content="http://yoursite.com/index.html">
<meta property="og:site_name" content="Weber Blog">
<meta property="og:locale" content="en_US">
<meta property="article:author" content="Weber Huang">
<meta name="twitter:card" content="summary">
  
    <link rel="alternative" href="/atom.xml" title="Weber Blog" type="application/atom+xml">
  
  
    <link rel="icon" href="/favicon.png">
  
  <link rel="stylesheet" type="text/css" href="/./main.0cf68a.css">
  <style type="text/css">
  
    #container.show {
      background: linear-gradient(200deg,#a0cfe4,#e8c37e);
    }
  </style>
  

  

<meta name="generator" content="Hexo 4.2.0"></head>

<body>
  <div id="container" q-class="show:isCtnShow">
    <canvas id="anm-canvas" class="anm-canvas"></canvas>
    <div class="left-col" q-class="show:isShow">
      
<div class="overlay" style="background: #4d4d4d"></div>
<div class="intrude-less">
	<header id="header" class="inner">
		<a href="/" class="profilepic">
			<img src="http://picture.hwb0307.top:8710/images/2020/02/09/85rOmI9uVcSXk3LC.jpg" class="js-avatar">
		</a>
		<hgroup>
		  <h1 class="header-author"><a href="/">Weber Huang</a></h1>
		</hgroup>
		

		<nav class="header-menu">
			<ul>
			
				<li><a href="/">主页</a></li>
	        
				<li><a href="/tags/%E9%9A%8F%E7%AC%94/">随笔</a></li>
	        
			</ul>
		</nav>
		<nav class="header-smart-menu">
    		
    			
    			<a q-on="click: openSlider(e, 'innerArchive')" href="javascript:void(0)">所有文章</a>
    			
            
    			
    			<a q-on="click: openSlider(e, 'friends')" href="javascript:void(0)">友链</a>
    			
            
    			
    			<a q-on="click: openSlider(e, 'aboutme')" href="javascript:void(0)">关于我</a>
    			
            
		</nav>
		<nav class="header-nav">
			<div class="social">
				
					<a class="github" target="_blank" href="#" title="github"><i class="icon-github"></i></a>
		        
					<a class="weibo" target="_blank" href="#" title="weibo"><i class="icon-weibo"></i></a>
		        
					<a class="rss" target="_blank" href="#" title="rss"><i class="icon-rss"></i></a>
		        
					<a class="zhihu" target="_blank" href="#" title="zhihu"><i class="icon-zhihu"></i></a>
		        
			</div>
		</nav>
	</header>		
</div>

    </div>
    <div class="mid-col" q-class="show:isShow,hide:isShow|isFalse">
      
<nav id="mobile-nav">
  	<div class="overlay js-overlay" style="background: #4d4d4d"></div>
	<div class="btnctn js-mobile-btnctn">
  		<div class="slider-trigger list" q-on="click: openSlider(e)"><i class="icon icon-sort"></i></div>
	</div>
	<div class="intrude-less">
		<header id="header" class="inner">
			<div class="profilepic">
				<img src="http://picture.hwb0307.top:8710/images/2020/02/09/85rOmI9uVcSXk3LC.jpg" class="js-avatar">
			</div>
			<hgroup>
			  <h1 class="header-author js-header-author">Weber Huang</h1>
			</hgroup>
			
			
			
				
			
				
			
			
			
			<nav class="header-nav">
				<div class="social">
					
						<a class="github" target="_blank" href="#" title="github"><i class="icon-github"></i></a>
			        
						<a class="weibo" target="_blank" href="#" title="weibo"><i class="icon-weibo"></i></a>
			        
						<a class="rss" target="_blank" href="#" title="rss"><i class="icon-rss"></i></a>
			        
						<a class="zhihu" target="_blank" href="#" title="zhihu"><i class="icon-zhihu"></i></a>
			        
				</div>
			</nav>

			<nav class="header-menu js-header-menu">
				<ul style="width: 50%">
				
				
					<li style="width: 50%"><a href="/">主页</a></li>
		        
					<li style="width: 50%"><a href="/tags/%E9%9A%8F%E7%AC%94/">随笔</a></li>
		        
				</ul>
			</nav>
		</header>				
	</div>
	<div class="mobile-mask" style="display:none" q-show="isShow"></div>
</nav>

      <div id="wrapper" class="body-wrap">
        <div class="menu-l">
          <div class="canvas-wrap">
            <canvas data-colors="#eaeaea" data-sectionHeight="100" data-contentId="js-content" id="myCanvas1" class="anm-canvas"></canvas>
          </div>
          <div id="js-content" class="content-ll">
            
  
    <article id="post-生信环境搭建" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="article-title" href="/2020/02/09/%E7%94%9F%E4%BF%A1%E7%8E%AF%E5%A2%83%E6%90%AD%E5%BB%BA/">生信环境搭建</a>
    </h1>
  

        
        <a href="/2020/02/09/%E7%94%9F%E4%BF%A1%E7%8E%AF%E5%A2%83%E6%90%AD%E5%BB%BA/" class="archive-article-date">
  	<time datetime="2020-02-09T07:24:47.484Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2020-02-09</time>
</a>
        
      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        <h1 id="生信环境必备配置"><a href="#生信环境必备配置" class="headerlink" title="生信环境必备配置"></a>生信环境必备配置</h1><h2 id="github"><a href="#github" class="headerlink" title="github"></a>github</h2><blockquote>
<p><a href="https://blog.csdn.net/tina_ttl/article/details/51326684" target="_blank" rel="noopener">如何在ubuntu下使用Github？</a></p>
</blockquote>
<h3 id="使用git"><a href="#使用git" class="headerlink" title="使用git"></a>使用git</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 初始化仓库</span></span></span><br><span class="line">cd ~/bin</span><br><span class="line">git init</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 更改本地文件</span></span></span><br><span class="line">touch README</span><br><span class="line">git add README</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 更新日志</span></span></span><br><span class="line">git commit -m 'A new update for bioinf functions'</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># push</span></span></span><br><span class="line">git push origin master</span><br><span class="line"><span class="meta">#</span><span class="bash"> git remote add origin https://github.com/huangwb8/bioinf.git <span class="comment">#可能会输入git的帐号和密码</span></span></span><br><span class="line"><span class="meta">#</span></span><br></pre></td></tr></table></figure>

<h3 id="安装git"><a href="#安装git" class="headerlink" title="安装git"></a>安装git</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 安装git</span></span></span><br><span class="line">sudo apt-get install git</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 检查SSH</span></span></span><br><span class="line">ssh -T git@github.com # Warning: Permanently added ‘github.com,204.232.175.90’ (RSA) to the list of known hosts. Permission denied (publickey).  说明可以连接</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 检查是否已井具有ssh keys</span></span></span><br><span class="line">cd ~/.ssh; ls</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 备份并移除已经存在的ssh keys</span></span></span><br><span class="line">mkdir key_backup; cp id_rsa* key_backup ; rm id_rsa* </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 执行如下命令</span></span></span><br><span class="line">ssh-keygen -t rsa -C "654751191@qq.com" #输入帐号和密码。后来，id_rsa（私钥）和id_rsa.pub（公钥）就被创建了</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 再次检查与github的连接</span></span></span><br><span class="line">ssh -T git@github.com # Hi alioth310! You’ve successfully uthenticated, but GitHub does not provide shell access.</span><br></pre></td></tr></table></figure>

<h3 id="解决每次git-pull-git-push都需输入密码设置"><a href="#解决每次git-pull-git-push都需输入密码设置" class="headerlink" title="解决每次git pull/git push都需输入密码设置"></a><a href="https://www.cnblogs.com/yulongcode/p/10882203.html" target="_blank" rel="noopener">解决每次git pull/git push都需输入密码设置</a></h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 执行这两条命令</span></span></span><br><span class="line">cd /</span><br><span class="line">git config --global credential.helper store</span><br></pre></td></tr></table></figure>

<p>执行完命令之后会在.gitconfig文件中多加红色字体</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">[user]</span><br><span class="line">        name &#x3D; your name</span><br><span class="line">        email &#x3D; your email</span><br><span class="line">[credential]</span><br><span class="line">        helper &#x3D; store</span><br></pre></td></tr></table></figure>

<p>使用：cd到项目目录，执行git pull,会提示输入账号密码。输完这一次以后git pull或git push就不在需要输入密码了（会在根目录生成一个 .git-credentials 文件）</p>
<h2 id="conda类"><a href="#conda类" class="headerlink" title="conda类"></a>conda类</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">conda activate base</span><br><span class="line">nohup conda install fastqc multiqc bowtie2 bwa samtools gatk4 qualimap --y &amp;</span><br></pre></td></tr></table></figure>

<h2 id="apt-get类"><a href="#apt-get类" class="headerlink" title="apt-get类"></a>apt-get类</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sudo apt-get install pigz</span><br></pre></td></tr></table></figure>

<h2 id="sratoolkit"><a href="#sratoolkit" class="headerlink" title="sratoolkit"></a>sratoolkit</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget https://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.9.6/sratoolkit.2.9.6-ubuntu64.tar.gz</span><br><span class="line">tar zxvf sratoolkit.2.9.6-ubuntu64.tar.gz</span><br><span class="line">cd sratoolkit.2.9.6-ubuntu64</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 添加环境路径；vim</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 测试软件是否安装成功</span></span></span><br><span class="line">prefetch --version</span><br></pre></td></tr></table></figure>

<h2 id="python"><a href="#python" class="headerlink" title="python"></a>python</h2><blockquote>
<p>主要是python3的环境.conda会自动安装python2环境.</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># python3</span></span></span><br><span class="line">cd ~/Downloads</span><br><span class="line">mkdir -p python &amp;&amp; cd python</span><br><span class="line">wget https://www.python.org/ftp/python/3.7.3/Python-3.7.3.tgz</span><br><span class="line">tar zxvf Python-3.7.3.tgz</span><br><span class="line">cd Python-3.7.3</span><br><span class="line">./configure --prefix="/home/huangwb8/Downloads/python/"</span><br><span class="line">make</span><br><span class="line">make install</span><br><span class="line">cd ../bin</span><br><span class="line">ls * | while read id; do ln -s /home/huangwb8/Downloads/python/bin/$&#123;id&#125; /usr/bin/$&#123;id&#125;; done</span><br><span class="line">which python3 </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># python2</span></span></span><br><span class="line">cd ~/Downloads</span><br><span class="line">mkdir -p python2 &amp;&amp; cd python2</span><br><span class="line">nohup wget https://www.python.org/ftp/python/2.7.8/Python-2.7.8.tgz &amp;</span><br><span class="line">tar zxvf Python-3.7.3.tgz</span><br><span class="line">cd Python-2.7.8</span><br><span class="line">./configure --prefix="/home/huangwb8/Downloads/python2/" #安装目录</span><br><span class="line">make</span><br><span class="line">make install</span><br><span class="line">cd ../bin</span><br><span class="line">ls * | while read id; do sudo ln -s /home/huangwb8/Downloads/python2/bin/$&#123;id&#125;  /home/huangwb8/bin/$&#123;id&#125;; done</span><br></pre></td></tr></table></figure>

<h2 id="cutadapt-amp-trim-galore"><a href="#cutadapt-amp-trim-galore" class="headerlink" title="cutadapt &amp; trim_galore"></a>cutadapt &amp; trim_galore</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 安装cutadapt（需要python3环境）</span></span></span><br><span class="line">pip3 install --user --upgrade cutadapt</span><br><span class="line">cutadapt --version</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 安装trim_galore（需要python3环境）</span></span></span><br><span class="line">mkdir -p ~/Download/trim_galore &amp;&amp; cd ~/Download/trim_galore</span><br><span class="line">curl -fsSL https://github.com/FelixKrueger/TrimGalore/archive/0.6.0.tar.gz -o trim_galore.tar.gz</span><br><span class="line">tar xvzf trim_galore.tar.gz</span><br><span class="line">sudo ln -s ~/Downloads/trim_galore/TrimGalore-0.6.0/trim_galore /usr/bin/trim_galore</span><br><span class="line">trim_galore --help</span><br></pre></td></tr></table></figure>

<h2 id="下载gnuplot-提供plot-bamstats函数"><a href="#下载gnuplot-提供plot-bamstats函数" class="headerlink" title="下载gnuplot(提供plot-bamstats函数)"></a>下载gnuplot(提供plot-bamstats函数)</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget https://jaist.dl.sourceforge.net/project/gnuplot/gnuplot/5.2.6/gnuplot-5.2.6.tar.gz </span><br><span class="line">tar -zxvf gnuplot-5.2.6.tar.gz</span><br><span class="line">cd gnuplot-5.2.6/</span><br><span class="line">./configure --prefix=/home/huangwb8/Downloads/gnuplot</span><br><span class="line">sudo make</span><br><span class="line">sudo make install</span><br></pre></td></tr></table></figure>

<h2 id="bcftools"><a href="#bcftools" class="headerlink" title="bcftools"></a>bcftools</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">p=~/Download/bcftools; mkdir -p $p; cd $p</span><br><span class="line">wget https://github.com/samtools/bcftools/releases/download/1.8/bcftools-1.8.tar.bz2 -b -c -T 120 -t inf  -o download_bcftools.log</span><br><span class="line">tar xjvf bcftools-1.8.tar.bz2</span><br><span class="line">cd bcftools-1.8/</span><br><span class="line">./configure</span><br><span class="line">make</span><br></pre></td></tr></table></figure>

<h2 id="ANNOVAR"><a href="#ANNOVAR" class="headerlink" title="ANNOVAR"></a>ANNOVAR</h2><blockquote>
<p><a href="http://www.bio-info-trainee.com/4007.html" target="_blank" rel="noopener">ANNOVAR 软件用法还可以更复杂</a></p>
<p><a href="https://mp.weixin.qq.com/s/GxqzKU7OPAMbjbZ5fPk7oQ" target="_blank" rel="noopener">【直播】我的基因组 32：使用annovar注释vcf</a></p>
<p>官方文档：<a href="http://annovar.openbioinformatics.org/en/latest/" target="_blank" rel="noopener">http://annovar.openbioinformatics.org/en/latest/</a></p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> 进入软件安装目录运行</span></span><br><span class="line">annovar=/data/annovar/annovar</span><br><span class="line">cd $annovar</span><br><span class="line">sudo vim ~/.bashrc # 添加/data/annovar/annovar到环境路径</span><br><span class="line">nohup annotate_variation.pl -downdb -webfrom annovar gnomad_genome --buildver hg38 ./humandb/ &gt; down.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 可以看到需要选择参考基因组版本，这里一般是 hg38, 很少看到hg19了。</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 然后是 数据库名字，在 http://annovar.openbioinformatics.org/en/latest/user-guide/download/ 全部列出。</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 其实就是相当于执行了下面的命令而已。</span></span></span><br><span class="line">wget -t 1 -T 30 -q -O hg38_gnomad_genome.txt.gz http://www.openbioinformatics.org/annovar/download/hg38_gnomad_genome.txt.gz</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 这个文件解压后有16G，注意云服务器的空间。</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 通常要下载的数据库非常多，包括：dbSNP,ExAC,ESP6500,cosmic,gnomad,1000genomes,clinvar,gwas等等</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 当然，注释上，并不等价于理解它们，这是很大的工作量。</span></span><br><span class="line">gversion=hg38</span><br><span class="line">nohup time annotate_variation.pl -buildver $gversion -downdb -webfrom annovar refGene humandb/ &gt; $&#123;gversion&#125;.refGene.log 2&gt;&amp;1 &amp;</span><br><span class="line">nohup time annotate_variation.pl -buildver $gversion -downdb cytoBand humandb/ &gt; $&#123;gversion&#125;.cytoBand.log 2&gt;&amp;1 &amp;</span><br><span class="line">nohup time annotate_variation.pl -buildver $gversion -downdb -webfrom annovar exac03 humandb/ &gt; $&#123;gversion&#125;.exac03.log 2&gt;&amp;1 &amp;</span><br><span class="line">nohup time annotate_variation.pl -buildver $gversion -downdb -webfrom annovar avsnp147 humandb/ &gt; $&#123;gversion&#125;.avsnp147.log 2&gt;&amp;1 &amp;</span><br><span class="line">nohup time annotate_variation.pl -buildver $gversion -downdb -webfrom annovar dbnsfp30a humandb/ &gt; $&#123;gversion&#125;.dbnsfp30a.log 2&gt;&amp;1 &amp;</span><br></pre></td></tr></table></figure>











<h1 id="WGS环境配置"><a href="#WGS环境配置" class="headerlink" title="WGS环境配置"></a>WGS环境配置</h1><h2 id="conda类-1"><a href="#conda类-1" class="headerlink" title="conda类"></a>conda类</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n wgs</span><br><span class="line">conda activate wgs</span><br><span class="line">nohup conda install  --y &amp;</span><br></pre></td></tr></table></figure>

<h1 id="WES环境配置"><a href="#WES环境配置" class="headerlink" title="WES环境配置"></a>WES环境配置</h1><h2 id="conda类-2"><a href="#conda类-2" class="headerlink" title="conda类"></a>conda类</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n wes</span><br><span class="line">conda activate wes</span><br><span class="line">nohup conda install bedtools --y &amp;</span><br></pre></td></tr></table></figure>





<h1 id="CHIP-Seq软件环境配置"><a href="#CHIP-Seq软件环境配置" class="headerlink" title="CHIP-Seq软件环境配置"></a>CHIP-Seq软件环境配置</h1><h2 id="conda类-3"><a href="#conda类-3" class="headerlink" title="conda类"></a>conda类</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n chip-seq</span><br><span class="line">conda activate chip-seq</span><br><span class="line">nohup conda install bedtools peakranger macs2 homer samtools deeptools --y &amp;</span><br></pre></td></tr></table></figure>

<h2 id="homer环境配置"><a href="#homer环境配置" class="headerlink" title="homer环境配置"></a>homer环境配置</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># homer环境配置</span></span></span><br><span class="line">ws_homer=/data/reference/homer</span><br><span class="line">path_log=$ws_homer/log/configureHomer_install</span><br><span class="line">mkdir -p $ws_homer $path_log</span><br><span class="line">cd $ws_homer #所有的工作都在homer当前目录下运行</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下载配置文件</span></span></span><br><span class="line">wget http://homer.salk.edu/homer/configureHomer.pl</span><br><span class="line">nohup perl configureHomer.pl -install &gt; $path_log/homer.log 2&gt;&amp;1 &amp;</span><br><span class="line">tail -f homer.log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 设置home/bin目录至该用户的永久环境路径，方便日后调用</span></span></span><br><span class="line">vim ~/.bashrc</span><br><span class="line"><span class="meta">#</span><span class="bash">添加/data/reference/homer/bin</span></span><br><span class="line">source ~/.bashrc</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 查看软件信息以获取下载名</span></span></span><br><span class="line">perl configureHomer.pl -list</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下载</span></span></span><br><span class="line">nohup time perl configureHomer.pl -install mm9 &gt; $path_log/configureHomer_install_mm9.log 2&gt;&amp;1 &amp; </span><br><span class="line"><span class="meta">#</span><span class="bash"> 鼠启动子</span></span><br><span class="line">nohup time perl configureHomer.pl -install mouse-p &gt; $path_log/configureHomer_install_mouse-p.log 2&gt;&amp;1 &amp; </span><br><span class="line"></span><br><span class="line">nohup time perl configureHomer.pl -install hg38 &gt; $path_log/configureHomer_install_hg38.log 2&gt;&amp;1 &amp; </span><br><span class="line"></span><br><span class="line">nohup time perl configureHomer.pl -install mm10 &gt; $path_log/configureHomer_install_mm10.log 2&gt;&amp;1 &amp; </span><br><span class="line"></span><br><span class="line">nohup time perl configureHomer.pl -install hg19 &gt; $path_log/configureHomer_install_hg19.log 2&gt;&amp;1 &amp; </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> 日志</span></span><br><span class="line">tail -f $path_log/*.log</span><br></pre></td></tr></table></figure>

<h1 id="RNA-seq环境"><a href="#RNA-seq环境" class="headerlink" title="RNA-seq环境"></a>RNA-seq环境</h1><h2 id="conda类-4"><a href="#conda类-4" class="headerlink" title="conda类"></a>conda类</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n rna-seq</span><br><span class="line">conda activate rna-seq</span><br><span class="line">nohup conda install htseq hisat2 subread cufflinks --y &amp;</span><br></pre></td></tr></table></figure>

<h1 id="Single-Cell-Seq环境"><a href="#Single-Cell-Seq环境" class="headerlink" title="Single Cell-Seq环境"></a>Single Cell-Seq环境</h1><h2 id="conda类-5"><a href="#conda类-5" class="headerlink" title="conda类"></a>conda类</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n scs</span><br><span class="line">conda activate scs</span><br><span class="line">conda install -y  subread &amp;</span><br></pre></td></tr></table></figure>

<h2 id="RSeQC"><a href="#RSeQC" class="headerlink" title="RSeQC"></a>RSeQC</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">ws=~/Downloads/RSeQC; mkdir -p $ws; cd$ws</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下载</span></span></span><br><span class="line">dl_wget https://jaist.dl.sourceforge.net/project/rseqc/RSeQC-3.0.1.tar.gz</span><br><span class="line">tar zxvf RSeQC-3.0.1.tar.gz; mv ./RSeQC-3.0.1 .; rm ./RSeQC-3.0.1</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 添加python3库</span></span></span><br><span class="line">sudo python3 ./setup.py install --root=/usr/bin</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 将库的软链添到python3现存库里</span></span></span><br><span class="line">lib=/home/huangwb8/Downloads/RSeQC/lib</span><br><span class="line">ls $lib | while read id; do (ln -s $lib/$id  /home/huangwb8/Downloads/python/lib/python3.7/site-packages); done </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 设置</span></span></span><br><span class="line">sudo ln -s python3.7 python</span><br></pre></td></tr></table></figure>









<h1 id="参考基因组"><a href="#参考基因组" class="headerlink" title="参考基因组"></a>参考基因组</h1><h2 id="人参考基因组h19"><a href="#人参考基因组h19" class="headerlink" title="人参考基因组h19"></a><a href="https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/" target="_blank" rel="noopener">人参考基因组h19</a></h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nohup wget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz &amp;  # h19</span><br></pre></td></tr></table></figure>
<h2 id="人参考基因组hg38"><a href="#人参考基因组hg38" class="headerlink" title="人参考基因组hg38"></a><a href="https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/" target="_blank" rel="noopener">人参考基因组hg38</a></h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">nohup wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.chromFa.tar.gz &amp; # hg38</span><br><span class="line">tar -zvf **chromFa.tar.gz</span><br><span class="line">touch hg38.fa</span><br><span class="line">cat *.fa &gt; hg38.fa</span><br><span class="line">rm chr*</span><br></pre></td></tr></table></figure>

<h2 id="鼠mm9基因组"><a href="#鼠mm9基因组" class="headerlink" title="鼠mm9基因组"></a><a href="http://hgdownload.cse.ucsc.edu/goldenPath/mm9/bigZips/chromFa.tar.gz" target="_blank" rel="noopener">鼠mm9基因组</a></h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">tar -zvf **chromFa.tar.gz</span><br><span class="line">mkdir chromFa;mv *fa ./chromFa</span><br><span class="line">touch mm9.fa</span><br><span class="line">cat ./chromFa/*.fa &gt; mm9.fa</span><br><span class="line">rm -r ./chromFa</span><br></pre></td></tr></table></figure>

<h2 id="鼠mm10基因组"><a href="#鼠mm10基因组" class="headerlink" title="鼠mm10基因组"></a><a href="http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz" target="_blank" rel="noopener">鼠mm10基因组</a></h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># fasta</span></span></span><br><span class="line">tar -zvf **chromFa.tar.gz</span><br><span class="line">mkdir chromFa;mv *fa ./chromFa</span><br><span class="line">touch mm10.fa</span><br><span class="line">cat ./chromFa/*.fa &gt; mm10.fa</span><br><span class="line">rm -r ./chromFa</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># gtf</span></span></span><br><span class="line">path_gtf=/data/reference/gtf/mm10_gencode.vM23.annotation.gtf.gz</span><br><span class="line">ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M23/gencode.vM23.annotation.gtf.gz</span><br></pre></td></tr></table></figure>



<h1 id="miRbase"><a href="#miRbase" class="headerlink" title="miRbase"></a>miRbase</h1><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ws=/data/reference/miRNA;</span><br><span class="line">wget ftp://mirbase.org/pub/mirbase/CURRENT/hairpin.fa.gz -b -c -T 120 -t inf -P $ws -o $ws/hairpin.log</span><br><span class="line">wget ftp://mirbase.org/pub/mirbase/CURRENT/mature.fa.gz -b -c -T 120 -t inf -P $ws -o $ws/mature.fa.log</span><br></pre></td></tr></table></figure>



<h1 id="索引文件制作"><a href="#索引文件制作" class="headerlink" title="索引文件制作"></a>索引文件制作</h1><blockquote>
<p>学习文献：<a href="https://www.jianshu.com/p/071c1757ded1" target="_blank" rel="noopener">bwa bowtie2 salmon subread hisat2建索引和比对</a></p>
</blockquote>
<h2 id="bowtie软件建立索引文件"><a href="#bowtie软件建立索引文件" class="headerlink" title="bowtie软件建立索引文件"></a>bowtie软件建立索引文件</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> 参考</span></span><br><span class="line">cd ~/reference</span><br><span class="line">mkdir -p index/bowtie &amp;&amp; cd index/bowtie</span><br><span class="line">nohup time ~/biosoft/bowtie/bowtie2-2.2.9/bowtie2-build  ~/reference/genome/hg19/hg19.fa  ~/reference/index/bowtie/hg19 1&gt;hg19.bowtie_index.log 2&gt;&amp;1 &amp;</span><br><span class="line">nohup time ~/biosoft/bowtie/bowtie2-2.2.9/bowtie2-build  ~/reference/genome/hg38/hg38.fa  ~/reference/index/bowtie/hg38 1&gt;hg38.bowtie_index.log 2&gt;&amp;1 &amp;</span><br><span class="line">nohup time ~/biosoft/bowtie/bowtie2-2.2.9/bowtie2-build  ~/reference/genome/mm10/mm10.fa  ~/reference/index/bowtie/mm10 1&gt;mm10.bowtie_index.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> 个人</span></span><br><span class="line">mkdir hg19 &amp;&amp; cd hg19</span><br><span class="line">nohup time bowtie2-build --threads 48 /data/reference/genome/hg19/hg19.fa hg19 &gt;hg19.bowtie_index.log 2&gt;&amp;1 &amp;</span><br><span class="line">mkdir hg38 &amp;&amp; cd hg38</span><br><span class="line">nohup time bowtie2-build --threads 48 /data/reference/genome/hg38/hg38.fa hg38 &gt;hg38.bowtie_index.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> pre-miRNA</span></span><br><span class="line">ws=/data/reference/miRNA; mkdir -p $ws/bowtie2</span><br><span class="line">nohup bowtie2-build --threads 48 $ws/hairpin.fa  $ws/bowtie2/hairpin &gt; $ws/bowtie2/hairpin.log 2&gt;&amp;1 &amp;</span><br><span class="line">tail -f $ws/bowtie2/hairpin.log</span><br></pre></td></tr></table></figure>

<h2 id="bwa软件建立索引文件"><a href="#bwa软件建立索引文件" class="headerlink" title="bwa软件建立索引文件"></a>bwa软件建立索引文件</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># hg19</span></span></span><br><span class="line">mkdir hg19 &amp;&amp; cd hg19</span><br><span class="line">nohup time bwa index -a bwtsw -p hg19 /data/reference/genome/hg19/hg19.fa 1&gt;hg19.bwa_index.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># hg38</span></span></span><br><span class="line">mkdir hg38 &amp;&amp; cd hg38</span><br><span class="line">nohup time bwa index -a bwtsw -p hg38 /data/reference/genome/hg38/hg38.fa 1&gt;hg38.bwa_index.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># mm9</span></span></span><br><span class="line">ws=/data/reference/mouse/mm9</span><br><span class="line">mkdir -p $ws/index/bwa</span><br><span class="line">cd $ws/index/bwa</span><br><span class="line">nohup time bwa index -a bwtsw -p $ws/index/bwa/mm9 $ws/genome/mm9.fa 1&gt;mm9.bwa_index.log 2&gt;&amp;1 &amp;</span><br><span class="line">tail -f mm9.bwa_index.log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># mm10</span></span></span><br><span class="line">ws=/data/reference/mouse/mm10</span><br><span class="line">mkdir -p $ws/index/bwa</span><br><span class="line">cd $ws/index/bwa</span><br><span class="line">nohup time bwa index -a bwtsw -p $ws/index/bwa/mm10 $ws/genome/mm10.fa 1&gt;mm10.bwa_index.log 2&gt;&amp;1 &amp;</span><br><span class="line">tail -f /data/reference/mouse/mm10/index/bwa/mm10.bwa_index.log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> pre-miRNA</span></span><br><span class="line">ws=/data/reference/miRNA; mkdir -p $ws/bwa</span><br><span class="line">nohup bwa index -p $ws/bwa/hairpin $ws/hairpin.fa &gt; $ws/bwa/hairpin.log 2&gt;&amp;1 &amp;</span><br><span class="line">tail -f $ws/bwa/hairpin.log</span><br></pre></td></tr></table></figure>



<h2 id="hisat软件建立索引文件"><a href="#hisat软件建立索引文件" class="headerlink" title="hisat软件建立索引文件"></a>hisat软件建立索引文件</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">cd ~/reference</span><br><span class="line">mkdir -p index/hisat &amp;&amp; cd index/hisat</span><br><span class="line">nohup wget ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/data/hg19.tar.gz  &amp;</span><br><span class="line">nohup wget ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/data/hg38.tar.gz  &amp;</span><br><span class="line">nohup wget ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/data/grcm38.tar.gz &amp;</span><br><span class="line">tar zxvf hg19.tar.gz</span><br><span class="line">tar zxvf grcm38.tar.gz</span><br><span class="line">tar zxvf hg38.tar.gz</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> pre-miRNA</span></span><br><span class="line">ws=/data/reference/miRNA; mkdir -p $ws/hisat2</span><br><span class="line">nohup hisat2-build -p 48 $ws/hairpin.fa  $ws/hisat2/hairpin &gt; $ws/hisat2/hairpin.log 2&gt;&amp;1 &amp;</span><br><span class="line">tail -f $ws/hisat2/hairpin.log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> mm10</span></span><br><span class="line">index=/data/reference/mouse/mm10/index/hisat2</span><br><span class="line">genome=/data/reference/mouse/mm10/genome/mm10.fa</span><br><span class="line">nohup hisat2-build -p 48 $genome  $index/mm10 &gt; $index/mm10.log 2&gt;&amp;1 &amp;</span><br><span class="line">tail -f $index/mm10.log</span><br></pre></td></tr></table></figure>



<h2 id="star索引文件"><a href="#star索引文件" class="headerlink" title="star索引文件"></a>star索引文件</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">注意--sjdbOverhang 参数为reads的长度-1</span></span><br><span class="line"><span class="meta">#</span><span class="bash">模式选择为 genomeGenerate </span></span><br><span class="line"><span class="meta">#</span><span class="bash"> gtf一定要是非压缩形式</span></span><br><span class="line"></span><br><span class="line">mkdir star &amp;&amp; cd star</span><br><span class="line"><span class="meta">#</span><span class="bash"> hg38</span></span><br><span class="line">mkdir hg38; cd hg38</span><br><span class="line">nohup STAR \</span><br><span class="line">    --runMode genomeGenerate \</span><br><span class="line">    --genomeDir . \</span><br><span class="line">    --genomeFastaFiles /data/reference/genome/hg38/hg38.fa \</span><br><span class="line">    --runThreadN 30 \</span><br><span class="line">    --sjdbGTFfile /data/reference/gtf/raw/hg38_gencode.v32.annotation.gtf \</span><br><span class="line">    &gt; star_index_hg38.log 2&gt;&amp;1 &amp;</span><br><span class="line">    </span><br><span class="line">nohup STAR \</span><br><span class="line">    --runMode genomeGenerate \</span><br><span class="line">    --genomeDir . \</span><br><span class="line">    --genomeFastaFiles /data/reference/genome/hg38/hg38.fa \</span><br><span class="line">    --runThreadN 48 \</span><br><span class="line">    &gt; star_index_hg38.log 2&gt;&amp;1 &amp;</span><br><span class="line">    </span><br><span class="line"><span class="meta">#</span><span class="bash"> hg19</span></span><br><span class="line">mkdir hg19; cd hg19</span><br><span class="line">nohup STAR \</span><br><span class="line">    --runMode genomeGenerate \</span><br><span class="line">    --genomeDir . \</span><br><span class="line">    --genomeFastaFiles /data/reference/genome/hg19/hg19.fa \</span><br><span class="line">    --runThreadN 30 \</span><br><span class="line">    --sjdbGTFfile /data/reference/gtf/raw/gencode.v19.annotation.gtf \</span><br><span class="line">    &gt; star_index_hg19.log 2&gt;&amp;1 &amp;</span><br></pre></td></tr></table></figure>



<h2 id="subjunc"><a href="#subjunc" class="headerlink" title="subjunc"></a>subjunc</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 用法</span></span></span><br><span class="line">subjunc -T 5 -i &lt;index.pre&gt; -r fq1 -R fq2 -o *.bam</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># index</span></span></span><br><span class="line">conda activate rna-seq</span><br><span class="line">conda install subread</span><br><span class="line">index=/data/reference/index/subread</span><br><span class="line">cd $index</span><br><span class="line">mkdir -p $index/&#123;mm10,mm9,hg19,hg38&#125;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># make index</span></span></span><br><span class="line">cd $index/hg38; nohup subread-buildindex -o $index/hg38/hg38 /data/reference/genome/hg38/hg38.fa &gt; ../subread_index_hg38.log 2&gt;&amp;1 &amp;</span><br><span class="line">cd $index/hg19; nohup subread-buildindex -o $index/hg19/hg19 /data/reference/genome/hg19/hg19.fa &gt; ../subread_index_hg19.log 2&gt;&amp;1 &amp;</span><br><span class="line">cd $index/mm10; nohup subread-buildindex -o $index/mm10/mm10  /data/reference/mouse/mm10/genome/mm10.fa &gt; ../subread_index_mm10.log 2&gt;&amp;1 &amp;</span><br><span class="line">cd $index/mm9; nohup subread-buildindex -o $index/mm9/mm9  /data/reference/mouse/mm9/genome/mm9.fa &gt; ../subread_index_mm9.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># log</span></span></span><br><span class="line">tail -f $index/subread_index*log</span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="built_in">kill</span> -9 `ps -ef | grep subread | awk <span class="string">'&#123;print $2&#125;'</span>`</span></span><br></pre></td></tr></table></figure>



<h1 id="基因组bed文件下载"><a href="#基因组bed文件下载" class="headerlink" title="基因组bed文件下载"></a>基因组bed文件下载</h1><h2 id="RNA-seq"><a href="#RNA-seq" class="headerlink" title="RNA-seq"></a>RNA-seq</h2><blockquote>
<p>需要提供bed文件，可以直接RSeQC的网站下载(看文件只有1M多，就没有考虑转换</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget https://downloads.sourceforge.net/project/rseqc/BED/Human_Homo_sapiens/hg19_RefSeq.bed.gz</span><br><span class="line">read_distribution.py -i SRR3589958_sorted.bam -r /data/reference/rseqc/hg19_RefSeq.bed</span><br></pre></td></tr></table></figure>
<h2 id="Chip-seq"><a href="#Chip-seq" class="headerlink" title="Chip-seq"></a>Chip-seq</h2><blockquote>
<p><a href="https://www.jianshu.com/p/31bb9d2b84b0" target="_blank" rel="noopener">如何从UCSC获取感兴趣区域的BED文件</a></p>
<p><a href="https://genome.ucsc.edu/cgi-bin/hgTables" target="_blank" rel="noopener">ucsc table browser</a></p>
</blockquote>
<p><img src="https://picture.hwb0307.top/images/2019/10/07/KQNRAsTJdqu89i0U.jpg" alt="基因组bed文件下载-right.jpg"></p>
<p>主要是这个output format要选对.否则分析的时候会错误.</p>
<h2 id="CCDS-exon"><a href="#CCDS-exon" class="headerlink" title="CCDS exon"></a>CCDS exon</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cd /data/reference/CCDS/mouse</span><br><span class="line">cat CCDS.current.txt | perl -alne '&#123;/\[(.*?)\]/;next unless $1;$gene=$F[2];$exons=$1;$exons=~s/\s//g;$exons=~s/-/\t/g;print "$F[0]\t$_\t$gene" foreach split/,/,$exons;&#125;'| sort -u | bedtools sort -i | awk '&#123;print "chr"$0"\t0\t+"&#125;'&gt; mm10.exon.bed</span><br><span class="line">bed=/data/reference/CCDS/mouse/mm10.exon.bed</span><br></pre></td></tr></table></figure>









      

      
    </div>
    <div class="article-info article-info-index">
      
      
      

      
        <p class="article-more-link">
          <a class="article-more-a" href="/2020/02/09/%E7%94%9F%E4%BF%A1%E7%8E%AF%E5%A2%83%E6%90%AD%E5%BB%BA/">展开全文 >></a>
        </p>
      

      
      <div class="clearfix"></div>
    </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-Decision curve analysis学习" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="article-title" href="/2020/02/09/Decision%20curve%20analysis%E5%AD%A6%E4%B9%A0/">Decision curve analysis学习</a>
    </h1>
  

        
        <a href="/2020/02/09/Decision%20curve%20analysis%E5%AD%A6%E4%B9%A0/" class="archive-article-date">
  	<time datetime="2020-02-09T07:24:47.483Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2020-02-09</time>
</a>
        
      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        <h2 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h2><ul>
<li>Vickers A J, Elkin E B. Decision curve analysis: a novel method for evaluating prediction models[J]. Med Decis Making, 2006, 26(6): 565-574.</li>
<li>Zhang Z, Rousson V, Lee W C, et al. Decision curve analysis: a technical note[J]. Ann Transl Med, 2018, 6(15): 308.</li>
</ul>
<h2 id="DCA的实践基础"><a href="#DCA的实践基础" class="headerlink" title="DCA的实践基础"></a>DCA的实践基础</h2><ul>
<li>临床中确诊通常是概率性的，即只有一定的把握判断患者是否处于疾病状态。</li>
<li>治疗通常是有副作用的，所以不可以盲目地进行治疗。</li>
<li>临床实践需要根据不同的个体情况，在权衡中作出选择</li>
</ul>
<h2 id="基本概念"><a href="#基本概念" class="headerlink" title="基本概念"></a>基本概念</h2><h3 id="1-threshold-probability-概率阈值"><a href="#1-threshold-probability-概率阈值" class="headerlink" title="1. threshold probability(概率阈值)"></a>1. threshold probability(概率阈值)</h3><ul>
<li>threshold probability, p<sub>t</sub>, is where the expected beneft of treatment is equal to the expected beneft of avoiding treatment.</li>
<li>the threshold probability can vary from patient to patient. For example, surgery-averse patients may opt to take surgery at a high risk of cancer, but cancer-averse patients are more likely to perform operation at lower risk of cancer.   </li>
</ul>
<p>在实际应用中，通常的情况是disease是无法被确诊的，只能说以$p$的概率判断疾病是否发生，这个$p$可以由某些预测模型$M$给出，但也许不完全准确。如下图，在患者的决策树中，$abcd$代表的是相应项目的量纲一致的获益程度(比如生活质量评分等)。患者根据什么选择是否接受treatment呢？</p>
<p><img src="https://picture.hwb0307.top/images/2019/11/12/VtBWK8CQJoUyC0mI.jpg" alt="患者决策树.jpg"></p>
<p>为了解决这个问题，此处引入了threshold probability——$p_t$，即概率阈值的这一统计量。它定义为：当疾病发生的概率=$p_t$时，选择treatment和non-treatment的获益是一样的。</p>
<p>基于此概念，$p_t$ 显然有以下等式：<br>$$<br>p_ta + (1-p_t)b = p_tc + (1-p_t)d<br>$$<br>换算后可有：<br>$$<br>\frac{a-c}{d-b} = \frac{1-p_t}{p_t}<br>$$<br>其中$a-c$即是真阳获益（假阴伤害，过度保守伤害），$d-b$即是假阳伤害（真阴获益，过度积极伤害）。举例理解此式：如果$p_t=10%$（即患病概率为10%时），当真阳获益与假阳伤害之比为9时，患者接收治疗与不接受治疗的获益是同等的。另外，此式还提供了两个信息：</p>
<ul>
<li>$p_t$<strong>是个性化指标，依患者不同而不同</strong>。对于特定场景中的treatment和disease，$ac$在群体中显然是稳定的，因为可以认为$a-c$是一个定值。而$bd$对于不同的患者而言是不固定的，即$d-b$是个动态值。举个例子。比如前列腺癌根治术的副作用是阳痿，$abcd$对应的是患者的性生活质量。对于十分年轻的患者，如果发生阳痿，患者的生活质量将急剧下降，此时$b$将会十分地小，$d$较大，相应地$p_t$变大；而老年患者或者阳痿症的患者，阳痿本身对后者的生活质量影响较小，其的$b$将会比较大，$d$较小，相应地$p_t$变小。</li>
<li>在实际运用中，计算$Net \ Benefit$的$abcd$均需要额外的资料才可以计算。而threshold probability将$abcd$的对应关系转化为$p_t$后，可以只根据$p_t$计算$Net \ Benefit$，因此更具有可行性。</li>
</ul>
<h3 id="2-Net-Benefit-净获益"><a href="#2-Net-Benefit-净获益" class="headerlink" title="2. Net Benefit(净获益)"></a>2. Net Benefit(净获益)</h3><p>如果给定一个$p_t$，即可计算Net benefit(NB)。公式如下：<br>$$<br>Net\ Benefit = \frac{TruePositiveCount}{n} \times (a-c) - \frac{FalsePositiveCount}{n} \times (d - b)<br>$$<br>由于$a-c$在特定场景下是固定的，此处为简化方程式，令$a-c=1$，右侧整体除以$a - c$则可换算为<br>$$<br>Net\ Benefit = \frac{TruePositiveCount}{n} - \frac{FalsePositiveCount}{n} \times \frac{d - b}{a - c} \ = \frac{TruePositiveCount}{n} - \frac{FalsePositiveCount}{n} \times \frac{p_t}{1-p_t}<br>$$<br>显然，此时$Net \ Benefit$是一个相对值，其大小与$a-c$的实际取值有关。因此$Net \ Benefit$只可用于获益比较，而不可用于描述真实获益（因为$a-c$是未知的）。</p>
<p>此式中的$TruePositiveCount$和$FalsePositiveCount$和$n$均属于四格表里的数据。由上述等式可知，当$p_t$趋于0时，$Net \ Benefit$趋向于$\frac{TruePositiveCount}{n}$；当$p_t$趋于1时，$Net \ Benefit$趋向于$-\infty$。</p>
<p>当$p_t$取值(0,1)时，此时Model对于Outcome的判断有所不同，则对应着某个四格表。</p>
<p>举个例子：前列腺癌有两种术式，保留精囊蒂（保守治疗）和去除精囊蒂（积极治疗）。积极治疗的肿瘤根除较完全，但容易因为神经损伤而导致阳痿；保守治疗的神经保存好，但肿瘤根除不完全，容易导致肿瘤复发。现共识认为<strong>发生SVI后则必须行积极治疗</strong>，否则肿瘤复发风险极高。然而SVI只有在术中、甚至是术后病理才可以确诊。因此人们需要可用于术前的SVI诊断模型M。现假设M是由肿瘤分期、病理分级和血清PSA的线性组合，并且我们已经获得了患者的SVI回顾性资料。我们取某cutoff值$c_i[min(M),max(M)]$，使得$\frac{Count(M &gt;= c_i)}{Count(M)}=10%$，即$p_t=10%$[1,0]。经资料查阅，此时M对Outcome的判断情况如下：</p>
<table>
<thead>
<tr>
<th align="center">预测模型（M）</th>
<th align="center">是否发生疾病</th>
<th align="center">Outcome(存在SVI)</th>
<th align="center">Outcome(不存在SVI)</th>
</tr>
</thead>
<tbody><tr>
<td align="center"></td>
<td align="center">n=902</td>
<td align="center">positive</td>
<td align="center">negative</td>
</tr>
<tr>
<td align="center">SVI发生率&gt;=p<sub>t</sub></td>
<td align="center">Yes</td>
<td align="center">65</td>
<td align="center">225</td>
</tr>
<tr>
<td align="center">SVI发生率&lt;p<sub>t</sub></td>
<td align="center">No</td>
<td align="center">22</td>
<td align="center">590</td>
</tr>
</tbody></table>
<p>此时$Net \ Benefit$的值则为：<br>$$<br>Net \ Benefit = \frac{65}{902} - \frac{225}{902} \times \frac{0.1}{1-0.1} = 0.0443459<br>$$</p>
<p>假设医生无法使用模型来预测患者的SVI情况，此时有两种极端选择：所有的患者都行积极治疗或行保守治疗。</p>
<p>如果患者全行保守治疗(<strong>None</strong>)，即认为SVI发生率=0（等价于$c_i=max(M)$），那么将无法获得患者的SVI情况。此时$TruePositiveCount$和$FalsePositiveCount$均为0，因而$Net \ Benefit=0$</p>
<p>如果患者全行积极治疗(<strong>All</strong>)，即认为SVI发生率=1 （等价于$c_i=min(M)$），那么所有患者都可以准确地获得SVI信息。此时$TruePositiveCount=65+22=87$，$FalsePositiveCount=225+590=815$。因此，对应的$Net \ Benefit = -0.0039$。</p>
<p>因此，我们可以说，当选择$p_t=10%$时，使用$M$享有更高的的$Net \ Benefit$。</p>
<h3 id="3-Decision-curve-analysis"><a href="#3-Decision-curve-analysis" class="headerlink" title="3. Decision curve analysis"></a>3. Decision curve analysis</h3><p>由于不同的患者的$b$不同，因此decision curveu将所有的$p_t$取值在坐标系中表示出来(通常是0.01~0.99这99个小数)。此时可以直观地比较不同的模型在某个区间的threshold probability下$Net \ Benefit$的差异。</p>
<p><img src="https://picture.hwb0307.top/images/2019/11/12/SY94BOO6KyDeTrEP.jpg" alt="DC.jpg"></p>
<p>此时要考虑另外一个因素，即实际应用中SVI的风险有多高时即行积极治疗。显然在SVI风险高达60<del>70%的患者当中，$M$是没有价值的。只有在SVI风险比较小的时候，比如20%</del>30%的时候，$M$可以比none或all的作法要有更高的获益。</p>
<ul>
<li>we need to consider the likely range of $p_t$ in the population, that is, the typical threshold probabilities of SVI at which surgeons would opt for complete dissection of seminal vesicles. If it were the case that all surgeons remove the seminal vesicle tip only if there was at least a 60 – 70% risk of SVI, the model clearly has no clinical role. But it is unlikely that any surgeon would consider removal of a healthy seminal vesicle tip to be worse than failing to remove a potentially cancerous one. If, on the other hand, we assume that the likely range of $p_t$ in the population is between 20% and 30%, we would use the model, because it is of clear benefit at these $p_t$’s </li>
</ul>

      

      
    </div>
    <div class="article-info article-info-index">
      
      
      

      
        <p class="article-more-link">
          <a class="article-more-a" href="/2020/02/09/Decision%20curve%20analysis%E5%AD%A6%E4%B9%A0/">展开全文 >></a>
        </p>
      

      
      <div class="clearfix"></div>
    </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-Linux常用语句" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="article-title" href="/2020/02/09/Linux%E5%B8%B8%E7%94%A8%E8%AF%AD%E5%8F%A5/">Linux常用语句</a>
    </h1>
  

        
        <a href="/2020/02/09/Linux%E5%B8%B8%E7%94%A8%E8%AF%AD%E5%8F%A5/" class="archive-article-date">
  	<time datetime="2020-02-09T07:24:47.481Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2020-02-09</time>
</a>
        
      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        <h2 id="常用的全局变量"><a href="#常用的全局变量" class="headerlink" title="常用的全局变量"></a>常用的全局变量</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">GATK=/home/huangwb8/Downloads/conda/location/envs/bioinf/bin/gatk # which gatk</span><br><span class="line">path_genome=/data/reference/genome/hg38</span><br><span class="line">path_bwa=/data/reference/index/bwa/hg38</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 函数常用选项</span></span></span><br><span class="line">ws=$1 #ws=/home/huangwb8/Test/wes_1</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/fastq</span><br><span class="line">path_bam=$&#123;ws&#125;/output/align</span><br><span class="line">path_log=$&#123;ws&#125;/log/sra2fq</span><br><span class="line">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span><br><span class="line"></span><br><span class="line">nthread=48</span><br><span class="line">path_sra=$&#123;ws&#125;/input/sra</span><br></pre></td></tr></table></figure>

<h2 id="kill某个程序的所有进程"><a href="#kill某个程序的所有进程" class="headerlink" title="kill某个程序的所有进程"></a>kill某个程序的所有进程</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 单个软件</span></span></span><br><span class="line">kill -9 `ps -ef | grep hisat2 | awk '&#123;print $2&#125;'`</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 多个软件</span></span></span><br><span class="line">kill -9 `ps -ef | grep -E 'hisat2|samtools' | awk '&#123;print $2&#125;'`</span><br></pre></td></tr></table></figure>



<h2 id="清空memory中的buff-cache空间"><a href="#清空memory中的buff-cache空间" class="headerlink" title="清空memory中的buff/cache空间"></a>清空memory中的buff/cache空间</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sync</span><br><span class="line">su root</span><br><span class="line">echo 1 &gt; /proc/sys/vm/drop_caches &amp;&amp; echo 2 &gt; /proc/sys/vm/drop_caches &amp;&amp; echo 3 &gt; /proc/sys/vm/drop_caches</span><br></pre></td></tr></table></figure>



<h2 id="java设置基本知识"><a href="#java设置基本知识" class="headerlink" title="java设置基本知识"></a>java设置基本知识</h2><p>在一些规模稍大的应用中，Java虚拟机（JVM）的内存设置尤为重要，想在项目中取得好的效率，GC（垃圾回收）的设置是第一步。<br><strong>PermGen space：</strong>全称是Permanent Generation space.就是说是永久保存的区域,用于存放Class和Meta信息,Class在被Load的时候被放入该区域Heap space：存放Instance。<br><strong>GC(Garbage Collection)</strong>应该不会对PermGen space进行清理,所以如果你的APP会LOAD很多CLASS的话,就很可能出现PermGen space错误。</p>
<p>Java Heap分为3个区：Young，Old和Permanent。Young保存刚实例化的对象。当该区被填满时，GC会将对象移到Old区。Permanent区则负责保存反射对象，本文不讨论该区。<br>JVM的Heap分配可以使用-X参数设定：</p>
<ul>
<li><p>-Xms表示初始Heap大小</p>
</li>
<li><p>-Xmx表示java heap最大值</p>
</li>
<li><p>-Xmn表示young generation的heap大小</p>
</li>
</ul>
<p>JVM有2个GC线程<br>第一个线程负责回收Heap的Young区<br>第二个线程在Heap不足时，遍历Heap，将Young 区升级为Older区<br>Older区的大小等于-Xmx减去-Xmn，不能将-Xms的值设的过大，因为第二个线程被迫运行会降低JVM的性能。</p>
<p>为什么一些程序频繁发生GC？<br>  有如下原因：<br>  1.程序内调用了System.gc()或Runtime.gc()。<br>  2.一些中间件软件调用自己的GC方法，此时需要设置参数禁止这些GC。<br>  3.Java的Heap太小，一般默认的Heap值都很小。<br>  4.频繁实例化对象，Release对象 此时尽量保存并重用对象，例如使用StringBuffer()和String()。<br>  如果你发现每次GC后，Heap的剩余空间会是总空间的50%，这表示你的Heap处于健康状态,许多Server端的Java程序每次GC后最好能有65%的剩余空间</p>
<p>  <strong>经验之谈：</strong><br>  1．Server端JVM最好将-Xms和-Xmx设为相同值。<strong>为了优化GC，最好让-Xmn值约等于-Xmx的1/3。</strong><br>  2．一个GUI程序最好是每10到20秒间运行一次GC，每次在半秒之内完成。<br>  注意：<br>  1．增加Heap的大小虽然会降低GC的频率，但也增加了每次GC的时间。并且GC运行时，所有的用户线程将暂停，也就是GC期间，Java应用程序不做任何工作。<br>  2．Heap大小并不决定进程的内存使用量。进程的内存使用量要大于-Xmx定义的值，因为Java为其他任务分配内存，例如每个线程的Stack等。</p>
<p>  <strong>Stack的设定</strong><br>  每个线程都有他自己的Stack。<br>  -Xss 表示每个线程的Stack大小<br>  Stack的大小限制着线程的数量。如果Stack过大就好导致内存溢漏。-Xss参数决定Stack大小，例如-Xss1024K。如果Stack太小，也会导致Stack溢漏。<br>  硬件环境<br>  硬件环境也影响GC的效率，例如机器的种类，内存，swap空间，和CPU的数量。<br>  如果你的程序需要频繁创建很多transient对象，会导致JVM频繁GC。这种情况你可以增加机器的内存，来减少Swap空间的使用。</p>
<p>  4种GC<br>  1、第一种为单线程GC，也是默认的GC，该GC适用于单CPU机器。<br>  2、第二种为Throughput GC，是多线程的GC，适用于多CPU，使用大量线程的程序。第二种GC与第一种GC相似，不同在于GC在收集Young区是多线程的，但在Old区和第一种一样，仍然采用单线程。-XX:+UseParallelGC参数启动该GC。<br>  3、第三种为Concurrent Low Pause GC，类似于第一种，适用于多CPU，并要求缩短因GC造成程序停滞的时间。这种GC可以在Old区的回收同时，运行应用程序。-XX:+UseConcMarkSweepGC参数启动该GC。<br>  4、第四种为Incremental Low Pause GC，适用于要求缩短因GC造成程序停滞的时间。这种GC可以在Young区回收的同时，回收一部分Old区对象。-Xincgc参数启动该GC。<br>  单文件的JVM内存进行设置<br>  默认的java虚拟机的大小比较小，在对大数据进行处理时java就会报错：java.lang.OutOfMemoryError。<br>  设置jvm内存的方法，对于单独的.class，可以用下面的方法对Test运行时的jvm内存进行设置。<br>  java -Xms64m -Xmx256m Test<br>  -Xms是设置内存初始化的大小<br>  -Xmx是设置最大能够使用内存的大小（最好不要超过物理内存大小）<br>  tomcat启动jvm内存设置<br>  Linux：<br>  在/usr/local/apache-tomcat-5.5.23/bin目录下的catalina.sh添加：JAVA_OPTS=’-Xms512m -Xmx1024m’要加“m”说明是MB，否则就是KB了，在启动tomcat时会报内存不足。<br>  -Xms：初始值<br>  -Xmx：最大值<br>  -Xmn：最小值Windows<br>  在catalina.bat最前面加入<br>  set JAVA_OPTS=-Xms128m -Xmx350m 如果用startup.bat启动tomcat,OK设置生效.够成功的分配200M内存.但是如果不是执行startup.bat启动tomcat而是利用windows的系统服务启动tomcat服务,上面的设置就不生效了,就是说set JAVA_OPTS=-Xms128m -Xmx350m 没起作用.上面分配200M内存就OOM了</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>









<h2 id="wget最佳用法"><a href="#wget最佳用法" class="headerlink" title="wget最佳用法"></a><a href="https://www.cnblogs.com/SkyForce/p/5883613.html" target="_blank" rel="noopener">wget最佳用法</a></h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> 参数说明： -b 后台下载；-o 输出日志；-c 断点续传；-T 延迟时间(这里是每120s就重新连接一次);-t 尝试无数次 ; -P 输出文件夹</span></span><br><span class="line">for((i=593;i&lt;=600;i++));</span><br><span class="line">do </span><br><span class="line">wget https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos/sra-pub-run-5/SRR1042$i/SRR1042$i.1 -b -c -T 120 -t inf -P $ws/input/sra2 -o $ws/log/wget_$i.log -O name.txt</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line">wget -bc -T 120 -t inf -o *.log html</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下载站点下的所有内容到当前目录中</span></span></span><br><span class="line">wget -m -nH &lt;URL&gt;</span><br><span class="line">wget -m -nH &lt;URL&gt;</span><br><span class="line">wget -c -b -r -np -nH -k -L -p -o $ws/log/wget_$i.log -e robots=off &lt;URL&gt;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># wget断点续传（下载大文件时可选）</span></span></span><br><span class="line">wget -c http://the.url.of/incomplete/file</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下载其它sra文件</span></span></span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_sra=$&#123;ws&#125;/input/sra</span><br><span class="line">path_log=$&#123;ws&#125;/log</span><br><span class="line">mkdir -p $path_sra $path_log</span><br><span class="line">cat $&#123;path_sra2fq_txt&#125; | while read id</span><br><span class="line">do</span><br><span class="line">arr=($&#123;id&#125;)</span><br><span class="line">sample=$&#123;arr[0]&#125; # sample=SRR3182421.sra</span><br><span class="line">if [ ! -f $&#123;path_sra&#125;/$&#123;sample&#125;.sra ]</span><br><span class="line">then</span><br><span class="line">wget https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos/sra-pub-run-5/$&#123;sample&#125;/$&#123;sample&#125;.1 -b -c -T 120 -t inf -P $path_sra -o $path_log/wget_$&#123;sample&#125;.log</span><br><span class="line">else</span><br><span class="line">echo "$&#123;sample&#125;.sra existed."</span><br><span class="line">fi</span><br><span class="line">done</span><br></pre></td></tr></table></figure>



<h2 id="axel多线程下载工具"><a href="#axel多线程下载工具" class="headerlink" title="axel多线程下载工具"></a>axel多线程下载工具</h2><blockquote>
<p> 经过实践，通过Axel下载，网速较之wget而言会快很多</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"> sudo apt-get install axel </span><br><span class="line"> </span><br><span class="line"><span class="meta"> #</span><span class="bash"><span class="comment"># 经典案例</span></span></span><br><span class="line">nohup axel -s 10240  -o ~/Downloads/test https://github.com/broadinstitute/cromwell/releases/download/47/cromwell-47.jar &gt; test.download.log 2&gt;&amp;1 &amp;</span><br><span class="line"> </span><br><span class="line"> </span><br><span class="line"><span class="meta"> #</span><span class="bash"><span class="comment"># axel --help</span></span></span><br><span class="line"> Usage: axel [options] url1 [url2] [url...]</span><br><span class="line">--max-speed=x		-s x	Specify maximum speed (bytes per second)</span><br><span class="line">--num-connections=x	-n x	Specify maximum number of connections</span><br><span class="line">--max-redirect=x		Specify maximum number of redirections</span><br><span class="line">--output=f		-o f	Specify local output file</span><br><span class="line">--search[=n]		-S[n]	Search for mirrors and download from n servers</span><br><span class="line">--ipv4			-4	Use the IPv4 protocol</span><br><span class="line">--ipv6			-6	Use the IPv6 protocol</span><br><span class="line">--header=x		-H x	Add HTTP header string</span><br><span class="line">--user-agent=x		-U x	Set user agent</span><br><span class="line">--no-proxy		-N	Just don't use any proxy server</span><br><span class="line">--insecure		-k	Don't verify the SSL certificate</span><br><span class="line">--no-clobber		-c	Skip download if file already exists</span><br><span class="line">--quiet			-q	Leave stdout alone</span><br><span class="line">--verbose		-v	More status information</span><br><span class="line">--alternate		-a	Alternate progress indicator</span><br><span class="line">--help			-h	This information</span><br><span class="line">--timeout=x		-T x	Set I/O and connection timeout</span><br><span class="line">--version		-V	Version information</span><br></pre></td></tr></table></figure>



<h2 id="查看进程网络占用"><a href="#查看进程网络占用" class="headerlink" title="查看进程网络占用"></a>查看进程网络占用</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sudo apt-get install nethogs</span><br><span class="line">sudo nethogs enp0s25</span><br></pre></td></tr></table></figure>



<h2 id="字符串常用操作"><a href="#字符串常用操作" class="headerlink" title="字符串常用操作"></a>字符串常用操作</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> 去除后缀 .bam</span></span><br><span class="line">sed 's/.bam//g' [object]</span><br><span class="line">sed 's/_marked_bqsr/""/' case3_biorep_B_marked_bqsr.g.vcf.abcde</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> 按空格分割字符串并选择第2个</span></span><br><span class="line">echo "13182122252601" | cut -b 1-2</span><br><span class="line">cut -d' ' -f 2 [object]</span><br><span class="line">cat sra2case.txt |while read id;do cut -d' ' -f 2 -;done</span><br></pre></td></tr></table></figure>



<h2 id="grep常用方法"><a href="#grep常用方法" class="headerlink" title="grep常用方法"></a>grep常用方法</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># grep</span></span></span><br><span class="line">-E ：开启扩展（Extend）的正则表达式。</span><br><span class="line">-i ：忽略大小写（ignore case）。</span><br><span class="line">-v ：反过来（invert），只打印没有匹配的，而匹配的反而不打印。</span><br><span class="line">-n ：显示行号</span><br><span class="line">-w ：被匹配的文本只能是单词，而不能是单词中的某一部分，如文本中有liker，而我搜寻的只是like，就可以使用-w选项来避免匹配liker</span><br><span class="line">-c ：显示总共有多少行被匹配到了，而不是显示被匹配到的内容，注意如果同时使用-cv选项是显示有多少行没有被匹配到。</span><br><span class="line">-o ：只显示被模式匹配到的字符串。</span><br><span class="line">--color :将匹配到的内容以颜色高亮显示。</span><br><span class="line">-A  n：显示匹配到的字符串所在的行及其后n行，after</span><br><span class="line">-B  n：显示匹配到的字符串所在的行及其前n行，before</span><br><span class="line">-C  n：显示匹配到的字符串所在的行及其前后各n行，context</span><br></pre></td></tr></table></figure>

<h2 id="或、与等操作"><a href="#或、与等操作" class="headerlink" title="或、与等操作"></a>或、与等操作</h2><p><strong>1、或操作</strong></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">grep -E '123|abc' filename  // 找出文件（filename）中包含123或者包含abc的行</span><br><span class="line">egrep '123|abc' filename    // 用egrep同样可以实现</span><br><span class="line">awk '/123|abc/' filename   // awk 的实现方式</span><br></pre></td></tr></table></figure>

<p><strong>2、与操作</strong></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">grep pattern1 files | grep pattern2 //显示既匹配 pattern1 又匹配 pattern2 的行。</span><br></pre></td></tr></table></figure>

<p><strong>3、其他操作</strong></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">grep -i pattern files   //不区分大小写地搜索。默认情况区分大小写，</span><br><span class="line">grep -l pattern files   //只列出匹配的文件名，</span><br><span class="line">grep -L pattern files   //列出不匹配的文件名，</span><br><span class="line">grep -w pattern files  //只匹配整个单词，而不是字符串的一部分（如匹配‘magic’，而不是‘magical’），</span><br><span class="line">grep -C number pattern files //匹配的上下文分别显示[number]行，</span><br></pre></td></tr></table></figure>

<h2 id="less的常用操作"><a href="#less的常用操作" class="headerlink" title="less的常用操作"></a>less的常用操作</h2><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">less 在查看之前不会加载整个文件。可以尝试使用 less 和 vi 打开一个很大的文件，你就会看到它们之间在速度上的区别。</span><br><span class="line"></span><br><span class="line">在 less 中导航命令类似于 vi。本文中将介绍一些导航命令以及使用 less 的其它一些技巧。</span><br><span class="line"></span><br><span class="line">## 搜索</span><br><span class="line">当使用命令 less file-name 打开一个文件后，可以使用下面的方式在文件中搜索。搜索时整个文本中匹配的部分会被高亮显示。</span><br><span class="line">向前搜索</span><br><span class="line">&#x2F; - 使用一个模式进行搜索，并定位到下一个匹配的文本</span><br><span class="line">n - 向前查找下一个匹配的文本</span><br><span class="line">N - 向后查找前一个匹配的文本</span><br><span class="line">向后搜索</span><br><span class="line">? - 使用模式进行搜索，并定位到前一个匹配的文本</span><br><span class="line">n - 向后查找下一个匹配的文本</span><br><span class="line">N - 向前查找前一个匹配的文本</span><br><span class="line"></span><br><span class="line">## 全屏导航</span><br><span class="line">ctrl + F - 向前移动一屏</span><br><span class="line">ctrl + B - 向后移动一屏</span><br><span class="line">ctrl + D - 向前移动半屏</span><br><span class="line">ctrl + U - 向后移动半屏</span><br><span class="line"></span><br><span class="line">## 单行导航</span><br><span class="line">j - 向前移动一行</span><br><span class="line">k - 向后移动一行</span><br><span class="line"></span><br><span class="line">## 其它导航</span><br><span class="line">G - 移动到最后一行</span><br><span class="line">g - 移动到第一行</span><br><span class="line">q &#x2F; ZZ - 退出 less 命令</span><br><span class="line"></span><br><span class="line">## 其它有用的命令</span><br><span class="line">v - 使用配置的编辑器编辑当前文件</span><br><span class="line">h - 显示 less 的帮助文档</span><br><span class="line">&amp;pattern - 仅显示匹配模式的行，而不是整个文件</span><br><span class="line"></span><br><span class="line">## 标记导航</span><br><span class="line">当使用 less 查看大文件时，可以在任何一个位置作标记，可以通过命令导航到标有特定标记的文本位置。</span><br><span class="line">ma - 使用 a 标记文本的当前位置</span><br><span class="line">&#39;a - 导航到标记 a 处</span><br><span class="line"></span><br><span class="line">## 浏览多个文件</span><br><span class="line">方式一，传递多个参数给 less，就能浏览多个文件。</span><br><span class="line">less file1 file2</span><br><span class="line"></span><br><span class="line">方式二，正在浏览一个文件时，使用 :e 打开另一个文件。</span><br><span class="line">less file1</span><br><span class="line">:e file2</span><br><span class="line"></span><br><span class="line">当打开多个文件时，使用如下命令在多个文件之间切换</span><br><span class="line">:n - 浏览下一个文件</span><br><span class="line">:p - 浏览前一个文件</span><br></pre></td></tr></table></figure>



<h2 id="文件-文件夹名字的相关操作"><a href="#文件-文件夹名字的相关操作" class="headerlink" title="文件/文件夹名字的相关操作"></a>文件/文件夹名字的相关操作</h2><h3 id="1、提取文件名"><a href="#1、提取文件名" class="headerlink" title="1、提取文件名"></a>1、提取文件名</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost log]# var&#x3D;&#x2F;dir1&#x2F;dir2&#x2F;file.txt</span><br><span class="line">[root@localhost log]# echo $&#123;var##*&#x2F;&#125;</span><br><span class="line">file.txt</span><br><span class="line">123</span><br></pre></td></tr></table></figure>

<h3 id="2、提取后缀"><a href="#2、提取后缀" class="headerlink" title="2、提取后缀"></a>2、提取后缀</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost log]# echo $&#123;var##*.&#125;</span><br><span class="line">txt</span><br><span class="line">12</span><br></pre></td></tr></table></figure>

<h3 id="3、提取不带后缀的文件名，分两步"><a href="#3、提取不带后缀的文件名，分两步" class="headerlink" title="3、提取不带后缀的文件名，分两步"></a>3、提取不带后缀的文件名，分两步</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost log]# tmp&#x3D;$&#123;var##*&#x2F;&#125;</span><br><span class="line">[root@localhost log]# echo $tmp</span><br><span class="line">file.txt</span><br><span class="line">[root@localhost log]# echo $&#123;tmp%.*&#125;</span><br><span class="line">file</span><br><span class="line">12345</span><br></pre></td></tr></table></figure>

<h3 id="4、提取目录"><a href="#4、提取目录" class="headerlink" title="4、提取目录"></a>4、提取目录</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost log]# echo $&#123;var%&#x2F;*&#125;</span><br><span class="line">&#x2F;dir1&#x2F;dir2</span><br><span class="line">12</span><br></pre></td></tr></table></figure>

<p>使用文件目录的专有命令basename和dirname</p>
<h3 id="1、提取文件名，注意：basename是一个命令，使用-而不是-而不是"><a href="#1、提取文件名，注意：basename是一个命令，使用-而不是-而不是" class="headerlink" title="1、提取文件名，注意：basename是一个命令，使用(),而不是(),而不是{}"></a>1、提取文件名，注意：basename是一个命令，使用(),而不是(),而不是{}</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost log]# echo $(basename $var)</span><br><span class="line">file.txt</span><br><span class="line">12</span><br></pre></td></tr></table></figure>

<h3 id="2、提取不带后缀的文件名"><a href="#2、提取不带后缀的文件名" class="headerlink" title="2、提取不带后缀的文件名"></a>2、提取不带后缀的文件名</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost log]# echo $(basename $var .txt)</span><br><span class="line">file</span><br><span class="line">12</span><br></pre></td></tr></table></figure>

<h3 id="3、提取目录"><a href="#3、提取目录" class="headerlink" title="3、提取目录"></a>3、提取目录</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">[root@localhost log]# dirname $var</span><br><span class="line">&#x2F;dir1&#x2F;dir2</span><br><span class="line">[root@localhost log]# echo $(dirname $var)</span><br><span class="line">&#x2F;dir1&#x2F;dir2</span><br></pre></td></tr></table></figure>

<h2 id="cut的用法"><a href="#cut的用法" class="headerlink" title="cut的用法"></a>cut的用法</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">cut -d " " -f1-100,1000-1005 infile.csv &gt; outfile.csv</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> 按空格分割字符串并选择第2个</span></span><br><span class="line">cut -d' ' -f 2 [object]</span><br><span class="line">cat sra2case.txt |while read id;do cut -d' ' -f 2 -;done</span><br></pre></td></tr></table></figure>



<h2 id="xargs的用法"><a href="#xargs的用法" class="headerlink" title="xargs的用法"></a>xargs的用法</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 对某个文件进行操作</span></span></span><br><span class="line">ls *.sam | xargs head -n 1</span><br></pre></td></tr></table></figure>









<h2 id="条件语句：且-或"><a href="#条件语句：且-或" class="headerlink" title="条件语句：且/或"></a>条件语句：且/或</h2><blockquote>
<p> <a href="https://www.cnblogs.com/yizhinantian/p/3540114.html" target="_blank" rel="noopener">linux shell中的条件判断语句</a></p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">1.如果a&gt;b且a&lt;c</span><br><span class="line">   if (( a &gt; b )) &amp;&amp; (( a &lt; c ))   </span><br><span class="line">   if [[ $a &gt; $b ]] &amp;&amp; [[ $a &lt; $c ]]</span><br><span class="line">    或者 if [ $a -gt $b -a $a -lt $c ]</span><br><span class="line"></span><br><span class="line">2.如果a&gt;b或a&lt;c</span><br><span class="line">if (( a &gt; b )) || (( a &lt; c ))</span><br><span class="line">if [[ $a &gt; $b ]] || [[ $a &lt; $c ]]</span><br><span class="line">if [ $a -gt $b -o $a -lt $c ]</span><br><span class="line"></span><br><span class="line">3. -o = or , -a = and , 但我一向只用 || 或者 &amp;&amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 文件是否存在</span></span></span><br><span class="line">if [ ! -f $tbin/$id ] # 空格要空够;if和[]之间的空格是必须的</span><br><span class="line">if [ ! -d "$myPath"]</span><br></pre></td></tr></table></figure>



<h2 id="shell脚本调用的尝试"><a href="#shell脚本调用的尝试" class="headerlink" title="shell脚本调用的尝试"></a>shell脚本调用的尝试</h2><p> 1.sh</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">!/bin/bash</span></span><br><span class="line">A=B</span><br><span class="line">echo "PID for 1.sh before exec/source/fork:$$"</span><br><span class="line">export A</span><br><span class="line"></span><br><span class="line">echo "1.sh: /$A is $A"</span><br><span class="line">case $1 in</span><br><span class="line">       exec)</span><br><span class="line">            echo "using exec..."</span><br><span class="line">            exec ./2.sh ;;</span><br><span class="line">       source)</span><br><span class="line">            echo "using source..."</span><br><span class="line">            . ./2.sh ;;</span><br><span class="line">       *)</span><br><span class="line">            echo "using fork by default..."</span><br><span class="line">            ./2.sh ;;</span><br><span class="line">esac</span><br><span class="line">echo "PID for 1.sh after exec/source/fork:$$"</span><br><span class="line">echo "1.sh: /$A is $A"</span><br></pre></td></tr></table></figure>

<p>2.sh</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">!/bin/bash</span></span><br><span class="line">echo "PID for 2.sh: $$"</span><br><span class="line">echo "2.sh get /$A=$A from 1.sh"</span><br><span class="line">A=C</span><br><span class="line">export A</span><br><span class="line">echo "2.sh: /$A is $A"</span><br></pre></td></tr></table></figure>

<p>运行如下命令调试</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">bash ./1.sh fork</span><br><span class="line">bash ./1.sh source</span><br><span class="line">bash ./1.sh exec</span><br></pre></td></tr></table></figure>



<h2 id="less命令常用方法"><a href="#less命令常用方法" class="headerlink" title="less命令常用方法"></a>less命令常用方法</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">? 6 # 查找字符串</span><br><span class="line">n # 从下向上翻找下一个包含 6 的地方</span><br><span class="line">N # 从上向下翻找下一个包含 6 的地方</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 1.全屏导航</span></span></span><br><span class="line">ctrl + F - 向前移动一屏</span><br><span class="line">ctrl + B - 向后移动一屏</span><br><span class="line">ctrl + D - 向前移动半屏</span><br><span class="line">ctrl + U - 向后移动半屏</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 2.单行导航</span></span></span><br><span class="line">j - 向前移动一行</span><br><span class="line">k - 向后移动一行</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 3.其它导航</span></span></span><br><span class="line">G - 移动到最后一行</span><br><span class="line">g - 移动到第一行</span><br><span class="line">q / ZZ - 退出 less 命令</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 4.其它有用的命令</span></span></span><br><span class="line">v - 使用配置的编辑器编辑当前文件</span><br><span class="line">h - 显示 less 的帮助文档</span><br><span class="line">&amp;pattern - 仅显示匹配模式的行，而不是整个文件</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 5.标记导航</span></span></span><br><span class="line">当使用 less 查看大文件时，可以在任何一个位置作标记，可以通过命令导航到标有特定标记的文本位置：</span><br><span class="line">ma - 使用 a 标记文本的当前位置 </span><br><span class="line">a - #导航到标记 a 处</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 转至下一个文件</span></span></span><br></pre></td></tr></table></figure>





<h2 id="cutadapt基本的用法"><a href="#cutadapt基本的用法" class="headerlink" title="cutadapt基本的用法"></a>cutadapt基本的用法</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"> cutadapt的功能特别强大，相对应的参数真的特别说，有几十个参数，我们平时只会用到很少的几个，我在这里为大家介绍一下。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> 最基本的形式，可以去掉3‘端的adapter序列</span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cutadapt -a AACCGGTT -o output.fastq input.fastq</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> 可以直接输入或者输出压缩文件，不需要修改参数，输出文件的后面加上.gz</span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cutadapt -a AACCGGTT -o output.fastq.gz input.fastq.gz</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> 假如去掉3‘端的adapter AAAAAAA 和5’端的adapter TTTTTTT</span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cutadapt -a AAAAAAA -g TTTTTTT -o output.fastq input.fastq</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cutadapt也可以用来进行reads的cut</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 去掉最前面的5bp</span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cutadapt -u 5 -o trimmed.fastq input_reads.fastq</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> 进行reads测序质量的过滤</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> cutadapt软件可以使用-q参数进行reads质量的过滤。基本原理就是，一般reads头和尾会因为测序仪状态或者是反应时间的问题造成测序质量差，比较粗略的一个过滤办法就是-q进行过滤。需要特别说明的是，这里的-q对应的数字和phred值是不一样的，它是软件根据一定的算法计算出来的。</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 3‘端进行一个简单的过滤,--quality-base=33是指序列使用的是phred33计分系统</span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cutadapt -q 10 --quality-base=33 -o output.fastq input.fastq </span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 3‘端 5’端都进行过滤,3<span class="string">'的阈值是10，5‘的阈值是15</span></span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cutadapt -q 10,15 --quality-base=33 -o output.fastq input.fastq </span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Reads 长度的过滤</span><br><span class="line">[--minimum-length N or -m N]</span><br><span class="line"><span class="meta">#</span><span class="bash"> 当序列长度小于N的时候，reads扔掉</span></span><br><span class="line"></span><br><span class="line">[--too-short-output FILE]</span><br><span class="line"><span class="meta">#</span><span class="bash"> 上面参数获得的这些序列不是直接扔掉，而是输出到一个文件中</span></span><br><span class="line"></span><br><span class="line">[--maximum-length N or -M N]</span><br><span class="line"><span class="meta">#</span><span class="bash"> 当序列长度大于N的时候，reads扔掉</span></span><br><span class="line"></span><br><span class="line">[--too-long-output FILE]</span><br><span class="line"><span class="meta">#</span><span class="bash"> 上面参数获得的这些序列不是直接扔掉，而是输出到一个文件中</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Paired-Reads的裁剪（trim）</span><br><span class="line"><span class="meta">#</span><span class="bash"> 现在很多的测序都是双端测序，那么从测序原理上来说，一对reads来自于1簇反应，所以一起进行adapter的trim可能效果更好。cutadapt自然也提供了这样的功能。</span></span><br><span class="line"><span class="meta">$</span><span class="bash"> cutadapt -a ADAPTER_FWD -A ADAPTER_REV -o out.1.fastq -p out.2.fastq reads.1.fastq reads.2.fastq</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -a是第1个文件的adapter序列</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -A是第2个文件的adapter序列</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -o是第1个输出文件</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -p是第2个输出文件</span></span><br></pre></td></tr></table></figure>








      

      
    </div>
    <div class="article-info article-info-index">
      
      
      

      
        <p class="article-more-link">
          <a class="article-more-a" href="/2020/02/09/Linux%E5%B8%B8%E7%94%A8%E8%AF%AD%E5%8F%A5/">展开全文 >></a>
        </p>
      

      
      <div class="clearfix"></div>
    </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
    <article id="post-bioinf" class="article article-type-post  article-index" itemscope itemprop="blogPost">
  <div class="article-inner">
    
      <header class="article-header">
        
  
    <h1 itemprop="name">
      <a class="article-title" href="/2020/02/09/bioinf/">bioinf</a>
    </h1>
  

        
        <a href="/2020/02/09/bioinf/" class="archive-article-date">
  	<time datetime="2020-02-09T07:24:47.479Z" itemprop="datePublished"><i class="icon-calendar icon"></i>2020-02-09</time>
</a>
        
      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        <hr>
<p>title: bioinformation common codes<br>date: 2019/9/1 20:46:25<br>categories:</p>
<ul>
<li>Bioinformatics<br>tags:</li>
<li>Tutorials</li>
<li>Linux</li>
</ul>
<hr>
<h2 id="文件管理结构"><a href="#文件管理结构" class="headerlink" title="文件管理结构"></a>文件管理结构</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mkdir -p $ws/&#123;bin,log,input/sra,output/&#123;fastq,qc/&#123;pre,post&#125;,align&#125;&#125;</span><br></pre></td></tr></table></figure>



<h2 id="常用定量"><a href="#常用定量" class="headerlink" title="常用定量"></a>常用定量</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 函数常用选项</span></span></span><br><span class="line">ws=$1 # work space</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/fastq</span><br><span class="line">path_bam=$&#123;ws&#125;/output/align</span><br><span class="line">path_log=$&#123;ws&#125;/log/sra2fq</span><br><span class="line">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span><br><span class="line">nthread=48</span><br><span class="line">path_sra=$&#123;ws&#125;/input/sra</span><br><span class="line"></span><br><span class="line">GATK=/home/huangwb8/Downloads/conda/location/envs/bioinf/bin/gatk</span><br><span class="line">path_genome=/data/reference/genome/hg38</span><br><span class="line">path_bwa=/data/reference/index/bwa/hg38</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 终止进程</span></span></span><br><span class="line">kill -9 `ps -ef | grep -E 'java' | awk '&#123;print $2&#125;'`</span><br><span class="line">kill -9 `ps -ef | grep -E 'plot|stats' | awk '&#123;print $2&#125;'`</span><br><span class="line">ps -ef | grep -E 'multiqc|stats'</span><br><span class="line">for((i=1;i&lt;=10;i++)); do sleep 4s ; kill -9 `ps -ef | grep -E 'GenomicsDBImport' | awk '&#123;print $2&#125;'` &gt; test.log 2&gt;&amp;1 ; done &amp;</span><br></pre></td></tr></table></figure>



<h2 id="系统函数"><a href="#系统函数" class="headerlink" title="系统函数"></a>系统函数</h2><h3 id="cp2"><a href="#cp2" class="headerlink" title="cp2"></a>cp2</h3><blockquote>
<p>将~/bin/example中的函数依次复制到新目录中。此时，如果新目录中存在同名的函数，则不覆盖。</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/cp2</span><br><span class="line">rm ~/bin/cp2</span><br><span class="line"><span class="meta">#</span><span class="bash">!/usr/bin/bash</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="keyword">function</span> cp2</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Description</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 将~/bin/example中的函数依次复制到新目录中。此时，如果新目录中存在同名的函数，则不覆盖。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Parameters</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="variable">$1</span> target bin document</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Usage</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> x=cp2;sudo chmod 777 ~/bin/<span class="variable">$x</span>;cat ~/bin/<span class="variable">$x</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> cp2 <span class="variable">$nbin</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Programe</span></span><br><span class="line">fbin=~/bin/example</span><br><span class="line">tbin=$1</span><br><span class="line"></span><br><span class="line">ls $fbin | while read id </span><br><span class="line">do</span><br><span class="line">if [ ! -f $tbin/$id ]</span><br><span class="line">then</span><br><span class="line"> cp $fbin/$id $tbin</span><br><span class="line"> echo "$&#123;id&#125; copied"</span><br><span class="line">else</span><br><span class="line"> echo "$&#123;id&#125; existed"</span><br><span class="line">fi</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 变成可执行</span></span></span><br><span class="line">sudo chmod -R +x $tbin</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> End mv</span></span><br></pre></td></tr></table></figure>



<h3 id="buff-cache-clean"><a href="#buff-cache-clean" class="headerlink" title="buff_cache_clean"></a>buff_cache_clean</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/buff_cache_clean</span><br><span class="line">sync &amp;&amp; su root &amp;&amp; echo 1 &gt; /proc/sys/vm/drop_caches &amp;&amp; echo 2 &gt; /proc/sys/vm/drop_caches &amp;&amp; echo 3 &gt; /proc/sys/vm/drop_caches &amp;&amp; su huangwb8</span><br></pre></td></tr></table></figure>



<h3 id="Fastextra"><a href="#Fastextra" class="headerlink" title="Fastextra"></a>Fastextra</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Split a charactor and output multiple results</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parmeters</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> string=<span class="string">"hello,shell,haha"</span>  </span></span><br><span class="line"><span class="meta">#</span><span class="bash"> split=,</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># example</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Fastextra <span class="string">"hello,shell,haha"</span> ,</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">string=$1</span><br><span class="line">split=$2</span><br><span class="line"></span><br><span class="line">array=($&#123;string//$split/ &#125;)</span><br><span class="line">for var in $&#123;array[@]&#125;; do echo $var; done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End</span></span></span><br></pre></td></tr></table></figure>



<h3 id="fast-log"><a href="#fast-log" class="headerlink" title="fast_log"></a>fast_log</h3><blockquote>
<p>A convenient way to tail the log files with given path of logs.</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">vim ~/bin/fast_log</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Fast way to tail recent <span class="built_in">log</span> files</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> fast_log <span class="variable">$path_log</span> 3</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">path_log=$1</span><br><span class="line">head_num=$2</span><br><span class="line">user=huangwb8</span><br><span class="line"><span class="meta">#</span><span class="bash">head_num=4</span></span><br><span class="line"><span class="meta">#</span><span class="bash">user=huangwb8</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programme</span></span></span><br><span class="line">log=$(ls $path_log/*.log -hlt | head -$&#123;head_num&#125; | grep $user | cut -d ':' -f 2 | cut -d " " -f 2)</span><br><span class="line">cd $path_log; tail -f $log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End</span></span></span><br></pre></td></tr></table></figure>



<h3 id="dl-axel"><a href="#dl-axel" class="headerlink" title="dl_axel"></a>dl_axel</h3><blockquote>
<p>download the contents of the url by <code>axel</code> tool</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Fast way to download via axel</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">url=$1</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -s `expr 100 \* 1024 \* 1024` </span></span><br><span class="line">base=$(basename $url)</span><br><span class="line">nohup axel -vc -s `expr 100 \* 1024 \* 1024`  -T 60 $url &gt; dl_axel-$&#123;base&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End</span></span></span><br></pre></td></tr></table></figure>



<h3 id="dl-aria2c"><a href="#dl-aria2c" class="headerlink" title="dl_aria2c"></a>dl_aria2c</h3><blockquote>
<p>download the contents of the url by <code>aria2c</code> tool</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Descriptions</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Fast way to download a url via aria2c</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">url=$1</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">base=$(basename $url)</span><br><span class="line">nohup aria2c -c -d . -s 10 $url &gt; dl_aria2c-$&#123;base&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End</span></span></span><br></pre></td></tr></table></figure>



<h3 id="dl-wget"><a href="#dl-wget" class="headerlink" title="dl_wget"></a>dl_wget</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/dl_wget</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Descriptions</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Fast way to download a url via aria2c</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">url=$1</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">base=$(basename $url)</span><br><span class="line">wget -bc -o wget_$&#123;base&#125;.log -T 120 -t inf $&#123;url&#125;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End</span></span></span><br></pre></td></tr></table></figure>











<h2 id="sra2fq"><a href="#sra2fq" class="headerlink" title="sra2fq"></a>sra2fq</h2><blockquote>
<p>将*.sra转换成*.fastq.gz。</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/sra2fq</span><br><span class="line">vim $nbin/sra2fq</span><br><span class="line">rm $nbin/sra2fq; cp2 $nbin</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: sra2fq</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Description</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 将sra转换成fastq.gz</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Usage:</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="built_in">cd</span> <span class="variable">$ws</span>/input/sra;nohup bash sra2fq <span class="variable">$ws</span> &amp; <span class="comment"># 由于scratch-path的原因，一定要在.sra的子目录下才可以成功</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Parameters:</span></span><br><span class="line">ws=$1 #wp=/home/huangwb8/Test/wes_1</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_sra=$&#123;ws&#125;/input/sra</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/fastq</span><br><span class="line">path_log=$&#123;ws&#125;/log/sra2fq</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> create space </span></span><br><span class="line">mkdir -p $path_fastq $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> sra2fastq</span></span><br><span class="line">cat $&#123;path_sra2fq_txt&#125; | while read id</span><br><span class="line">do</span><br><span class="line">  arr=($&#123;id&#125;)</span><br><span class="line">  sample=$&#123;arr[0]&#125;.sra # sample=SRR3182421.sra</span><br><span class="line">  case=$&#123;arr[1]&#125; # case=case3_germline</span><br><span class="line">  </span><br><span class="line">  if [ -f $&#123;path_fastq&#125;/$&#123;case&#125;.fastq.gz -o -f $&#123;path_fastq&#125;/$&#123;case&#125;_1.fastq.gz ]; then </span><br><span class="line">  echo "$&#123;case&#125;: fastq.gz data existed. Ignored."</span><br><span class="line">  elif [ -f $&#123;path_sra&#125;/$&#123;sample&#125; ]; then</span><br><span class="line"><span class="meta">  #</span><span class="bash"><span class="comment"># -p show progress; -x print details; -3 writes single reads in special file;-N use row-id as name; -P print read-numbers; -f force to overwrite existing file(s)，这里不加;-e how many thread; --skip-technical skip technical reads</span></span></span><br><span class="line">  time fasterq-dump -p -x -3 -N -P --skip-technical -e $&#123;nthread&#125; $&#123;path_sra&#125;/$&#123;sample&#125; -O $&#123;path_fastq&#125; &gt; $&#123;path_log&#125;/sra2fq_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">  else</span><br><span class="line">  echo "$&#123;path_sra&#125;/$&#123;sample&#125; dosen't existed.Please check it."</span><br><span class="line">  fi</span><br><span class="line">  </span><br><span class="line"><span class="meta">  #</span><span class="bash"><span class="comment"># 检测是双端/单端测序数据</span></span></span><br><span class="line">  if [ -f $&#123;path_fastq&#125;/$&#123;sample&#125;_1.fastq -a -f $&#123;path_fastq&#125;/$&#123;sample&#125;_2.fastq ]; then</span><br><span class="line">     echo "$&#123;case&#125;: paired data"</span><br><span class="line">     mv $&#123;path_fastq&#125;/$&#123;sample&#125;_1.fastq $&#123;path_fastq&#125;/$&#123;case&#125;_1.fastq</span><br><span class="line">     pigz -p $&#123;nthread&#125; $&#123;path_fastq&#125;/$&#123;case&#125;_1.fastq	</span><br><span class="line">     mv $&#123;path_fastq&#125;/$&#123;sample&#125;_2.fastq $&#123;path_fastq&#125;/$&#123;case&#125;_2.fastq</span><br><span class="line">     pigz -p $&#123;nthread&#125; $&#123;path_fastq&#125;/$&#123;case&#125;_2.fastq</span><br><span class="line">  elif [ -f $&#123;path_fastq&#125;/$&#123;sample&#125;.fastq ]; then </span><br><span class="line">     echo "$&#123;case&#125;: unpaired data"</span><br><span class="line">     mv $&#123;path_fastq&#125;/$&#123;sample&#125;.fastq $&#123;path_fastq&#125;/$&#123;case&#125;.fastq</span><br><span class="line">     pigz -p $&#123;nthread&#125; $&#123;path_fastq&#125;/$&#123;case&#125;.fastq</span><br><span class="line">  else</span><br><span class="line">  	echo "$&#123;case&#125;: no fastq data. Ignored."</span><br><span class="line">  fi</span><br><span class="line"></span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 清除多余的文件</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">ls <span class="variable">$path_fastq</span> | grep <span class="string">".fastq$"</span> | <span class="keyword">while</span> <span class="built_in">read</span> i;<span class="keyword">do</span> rm <span class="variable">$path_fastq</span>/<span class="variable">$i</span>;<span class="keyword">done</span></span></span><br><span class="line">echo "All done!"</span><br><span class="line"><span class="meta">#</span><span class="bash"> End sra2fq</span></span><br></pre></td></tr></table></figure>

<h3 id="sra2fq-fasterq-dump"><a href="#sra2fq-fasterq-dump" class="headerlink" title="sra2fq_fasterq-dump"></a>sra2fq_fasterq-dump</h3><blockquote>
<p>将*.sra转换成*.fastq。然而似乎受硬盘存储速度的限制，这种策略其实并不比sra2fq的要快多少。</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/sra2fq_fasterq-dump</span><br><span class="line">vim $nbin/sra2fq_fasterq-dump</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: sra2fq_fasterq-dump</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="built_in">cd</span> <span class="variable">$ws</span>/input/sra;sra2fq_fasterq-dump <span class="variable">$ws</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/sra2fq/*fasterq*.<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 将sra转换成fastq.gz</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters:</span></span></span><br><span class="line">ws=$1 #wp=/home/huangwb8/Test/wes_1</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_sra=$&#123;ws&#125;/input/sra</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/fastq</span><br><span class="line">path_log=$&#123;ws&#125;/log/sra2fq</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Setting</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> x=sra2fq;sudo chmod 777 <span class="variable">$ws</span>/bin/<span class="variable">$x</span>;cat <span class="variable">$ws</span>/bin/<span class="variable">$x</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># create space </span></span></span><br><span class="line">mkdir -p &#123;$&#123;path_fastq&#125;,$&#123;path_log&#125;&#125;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># sra2fastq</span></span></span><br><span class="line">cat $&#123;path_sra2fq_txt&#125; | while read id</span><br><span class="line">do</span><br><span class="line">  arr=($&#123;id&#125;)</span><br><span class="line">  sample=$&#123;arr[0]&#125;.sra # sample=SRR3182421.sra</span><br><span class="line">  case=$&#123;arr[1]&#125; # case=case3_germline</span><br><span class="line">  if [ -f $&#123;path_sra&#125;/$&#123;sample&#125; ]; then</span><br><span class="line">  touch $&#123;path_log&#125;/sra2fq_$&#123;case&#125;.log</span><br><span class="line"><span class="meta">  #</span><span class="bash"><span class="comment"># -p show progress; -x print details; -3 writes single reads in special file;-N use row-id as name; -P print read-numbers; -f force to overwrite existing file(s)，这里不加;-e how many thread; --skip-technical skip technical reads</span></span></span><br><span class="line">  echo "$&#123;path_sra&#125;/$&#123;sample&#125;: do fasterq-dump..."</span><br><span class="line">  nohup time fasterq-dump -p -x -3 -N -P --skip-technical -e $&#123;nthread&#125; $&#123;path_sra&#125;/$&#123;sample&#125; -O $&#123;path_fastq&#125; &gt; $&#123;path_log&#125;/sra2fq_fasterq-dump_$&#123;case&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line">  echo "$&#123;path_sra&#125;/$&#123;sample&#125;: done!"</span><br><span class="line">  else </span><br><span class="line">  echo "$&#123;path_sra&#125;/$&#123;sample&#125; is not existed."</span><br><span class="line">  fi</span><br><span class="line">done</span><br><span class="line">echo "Done!"</span><br><span class="line"><span class="meta">#</span><span class="bash"> End sra2fq_fasterq-dump</span></span><br></pre></td></tr></table></figure>
<h3 id="sra2fq-pigz"><a href="#sra2fq-pigz" class="headerlink" title="sra2fq_pigz"></a>sra2fq_pigz</h3><blockquote>
<p>将*.fastq转换成*.fastq.gz。</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/sra2fq_pigz</span><br><span class="line">vim $nbin/sra2fq_pigz</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: sra2fq_pigz</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Usage:</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup sra2fq_pigz <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/sra2fq/sra2fq_pigz.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/sra2fq/*pigz.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Description</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 将sra转换成fastq.gz</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Parameters:</span></span><br><span class="line">ws=$1 #wp=/home/huangwb8/Test/wes_1</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/fastq</span><br><span class="line">path_log=$&#123;ws&#125;/log/sra2fq</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Setting</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> x=sra2fq;sudo chmod 777 <span class="variable">$ws</span>/bin/<span class="variable">$x</span>;cat <span class="variable">$ws</span>/bin/<span class="variable">$x</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> create space </span></span><br><span class="line">mkdir -p &#123;$&#123;path_fastq&#125;,$&#123;path_log&#125;&#125;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> sra2fastq</span></span><br><span class="line">cat $&#123;path_sra2fq_txt&#125; | while read id</span><br><span class="line">do</span><br><span class="line">  arr=($&#123;id&#125;)</span><br><span class="line">  sample=$&#123;arr[0]&#125;.sra # sample=SRR3182421.sra</span><br><span class="line">  case=$&#123;arr[1]&#125; # case=case3_germline</span><br><span class="line">  </span><br><span class="line"><span class="meta">  #</span><span class="bash"><span class="comment"># 检测是双端/单端测序数据</span></span></span><br><span class="line">  if [ -f $&#123;path_fastq&#125;/$&#123;sample&#125;_1.fastq -a -f $&#123;path_fastq&#125;/$&#123;sample&#125;_2.fastq ]</span><br><span class="line">  then # 双端</span><br><span class="line">  echo "$&#123;path_fastq&#125;/$&#123;sample&#125;: paired data."</span><br><span class="line">  echo "$&#123;path_fastq&#125;/$&#123;sample&#125;: do pigz..."</span><br><span class="line">  mv $&#123;path_fastq&#125;/$&#123;sample&#125;_1.fastq $&#123;path_fastq&#125;/$&#123;case&#125;_1.fastq</span><br><span class="line">  pigz -p $&#123;nthread&#125; $&#123;path_fastq&#125;/$&#123;case&#125;_1.fastq	</span><br><span class="line"><span class="meta">  #</span><span class="bash">sed s/<span class="variable">$&#123;sample&#125;</span>/<span class="variable">$&#123;case&#125;</span>/ <span class="variable">$&#123;path_fastq&#125;</span>/<span class="variable">$&#123;sample&#125;</span>_2.fastq &gt; <span class="variable">$&#123;path_fastq&#125;</span>/<span class="variable">$&#123;case&#125;</span>_2.fastq</span></span><br><span class="line">  mv $&#123;path_fastq&#125;/$&#123;sample&#125;_2.fastq $&#123;path_fastq&#125;/$&#123;case&#125;_2.fastq</span><br><span class="line">  pigz -p $&#123;nthread&#125; $&#123;path_fastq&#125;/$&#123;case&#125;_2.fastq</span><br><span class="line">  echo "$&#123;path_fastq&#125;/$&#123;sample&#125;: done!"</span><br><span class="line">  else # 单端</span><br><span class="line">  echo "$&#123;path_fastq&#125;/$&#123;sample&#125;: single data."</span><br><span class="line">  echo "$&#123;path_fastq&#125;/$&#123;sample&#125;: do pigz..."</span><br><span class="line">  mv $&#123;path_fastq&#125;/$&#123;sample&#125;.fastq $&#123;path_fastq&#125;/$&#123;case&#125;.fastq</span><br><span class="line">  pigz -p $&#123;nthread&#125; $&#123;path_fastq&#125;/$&#123;case&#125;.fastq</span><br><span class="line">  echo "$&#123;path_fastq&#125;/$&#123;sample&#125;: done!"</span><br><span class="line">  fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"> End sra2fq_pigz</span></span><br></pre></td></tr></table></figure>



<h2 id="clean-trim-galore"><a href="#clean-trim-galore" class="headerlink" title="clean_trim_galore"></a>clean_trim_galore</h2><blockquote>
<p>对*.fastq.gz文件去常见adapter</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/clean_trim_galore</span><br><span class="line">rm $nbin/clean_trim_galore; cp2 $nbin</span><br><span class="line">vim $nbin/clean_trim_galore</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: clean_trim_galore</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 对*.fastq.gz文件去常见adapter</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters:</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_sra2case_txt=$ws/input/case/sra2case.txt</span><br><span class="line">path_fastq=$ws/output/fastq</span><br><span class="line">path_log=$ws/log/clean_trim_galore</span><br><span class="line">output=$ws/output/clean-trim_galore</span><br><span class="line">ncore=13 # 48线程时可以设置为13</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup clean_trim_galore <span class="variable">$ws</span> &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> clean_trim_galore <span class="variable">$ws</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/clean_trim_galore/*<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> mkdir </span></span><br><span class="line">mkdir -p $path_log $output</span><br><span class="line">leave_base=`expr $&#123;n_all&#125; - $&#123;cut_end5&#125;`</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="built_in">read</span> id</span></span><br><span class="line">cat $path_sra2case_txt | while read id</span><br><span class="line">do</span><br><span class="line">  arr=($&#123;id&#125;)</span><br><span class="line">  case=$&#123;arr[1]&#125;</span><br><span class="line">  fq1=$&#123;case&#125;_1.fastq.gz</span><br><span class="line">  fq2=$&#123;case&#125;_2.fastq.gz</span><br><span class="line">  fq=$&#123;case&#125;.fastq.gz</span><br><span class="line"><span class="meta">  #</span><span class="bash"> 一般Illumina在1.9以上的都用phred33; length主要看原本测序序列的长度。如果太短（比如RNA-seq中的片段通常只有60多bp），此时length的值应该要偏低些；core是线程，48核时此参数可设置为13。--hardtrim5 切除5‘端的某几个碱基。--gzip在安装pigz时可以不用啦。--stringency Overlap with adapter sequence required to trim a sequence. 一般设置4即可。-e是Maximum allowed error rate，一般默认是0.1</span></span><br><span class="line">  if [ -f $&#123;path_fastq&#125;/$&#123;fq1&#125; -a -f $&#123;path_fastq&#125;/$&#123;fq2&#125; ]; then</span><br><span class="line"><span class="meta">  #</span><span class="bash"> 双端</span></span><br><span class="line">  	echo "find: $&#123;fq1&#125;,$&#123;fq2&#125;. Do trim_galore ..."</span><br><span class="line">    time trim_galore  --paired -q 25 \</span><br><span class="line">        --phred33 --length 35 --stringency 4 \</span><br><span class="line">        --hardtrim3 $&#123;leave_base&#125; \</span><br><span class="line">        -e=0.1 \</span><br><span class="line">        --cores $ncore \</span><br><span class="line">        -o $output $path_fastq/$fq1 $path_fastq/$fq2 \</span><br><span class="line">        &gt; $path_log/trimgalore_paired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">    echo "Done trim_galore of  $&#123;fq1&#125;,$&#123;fq2&#125;!"</span><br><span class="line">  elif [ -f $&#123;path_fastq&#125;/$&#123;fq&#125; ]; then</span><br><span class="line"><span class="meta">  #</span><span class="bash"> 单端</span></span><br><span class="line">      echo "find: $&#123;fq2&#125;"</span><br><span class="line">      time trim_galore -q 25 \</span><br><span class="line">          --phred33 --length 35 --stringency 4 \</span><br><span class="line">          --hardtrim3 $&#123;leave_base&#125; \</span><br><span class="line">          -e=0.1 \</span><br><span class="line">          --cores $ncore \</span><br><span class="line">          -o $output $path_fastq/$fq &gt; \</span><br><span class="line">          $path_log/trimgalore_single_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">  else</span><br><span class="line">  	echo "Files not found: $&#123;fq1&#125;,$&#123;fq2&#125; or $&#123;fq&#125;"</span><br><span class="line">  fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End clean_trim_galore</span></span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">###=================================history========================####</span></span></span><br><span class="line">n_all=48 # 开始时的测序长度。这个要看首次fastqc的结果才可以确定。</span><br><span class="line">cut_end5=4 # 因为测序刚刚开始时都会有些不正常。也就是3'端留下(n_all-cut_end)个碱基，就是间接地，5’端被切掉前4个bp</span><br></pre></td></tr></table></figure>

<h2 id="sm-trim-galore"><a href="#sm-trim-galore" class="headerlink" title="sm_trim_galore"></a>sm_trim_galore</h2><blockquote>
<p>TrimGalore for small RNA-Seq</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/clean_trim_galore</span><br><span class="line">vim $nbin/clean_trim_galore</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: clean_trim_galore</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 对*.fastq.gz文件去常见adapter</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters:</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_sra2case_txt=$ws/input/case/sra2case.txt</span><br><span class="line">path_fastq=$ws/output/fastq</span><br><span class="line">path_log=$ws/log/clean_trim_galore</span><br><span class="line">output=$ws/output/clean-trim_galore</span><br><span class="line">n_end5=4 # 5’端通常要切掉前几bp。因为测序刚刚开始时都会有些不正常。不设置也行</span><br><span class="line">n_end3=12 # 一般都是不需要的</span><br><span class="line">ncore=13 # 48线程时可以设置为13</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup clean_trim_galore <span class="variable">$ws</span> &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> clean_trim_galore <span class="variable">$ws</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/clean_trim_galore/*<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> mkdir </span></span><br><span class="line">mkdir -p $path_log $output</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="built_in">read</span> id</span></span><br><span class="line">cat $path_sra2case_txt | while read id</span><br><span class="line">do</span><br><span class="line">  arr=($&#123;id&#125;)</span><br><span class="line">  case=$&#123;arr[1]&#125;</span><br><span class="line">  fq1=$&#123;case&#125;_1.fastq.gz</span><br><span class="line">  fq2=$&#123;case&#125;_2.fastq.gz</span><br><span class="line">  fq=$&#123;case&#125;.fastq.gz</span><br><span class="line"><span class="meta">  #</span><span class="bash"> 一般Illumina在1.9以上的都用phred33; length主要看原本测序序列的长度。如果太短（比如RNA-seq中的片段通常只有60多bp），此时length的值应该要偏低些；core是线程，48核时此参数可设置为13。--hardtrim5 切除5‘端的某几个碱基。--gzip在安装pigz时可以不用啦。--stringency Overlap with adapter sequence required to trim a sequence. 一般设置4即可。-e是Maximum allowed error rate，一般默认是0.1</span></span><br><span class="line">  if [ -f $&#123;path_fastq&#125;/$&#123;fq1&#125; -a -f $&#123;path_fastq&#125;/$&#123;fq2&#125; ]; then</span><br><span class="line"><span class="meta">  #</span><span class="bash"> 双端</span></span><br><span class="line">  	echo "find: $&#123;fq1&#125;,$&#123;fq2&#125;. Do trim_galore ..."</span><br><span class="line">    time trim_galore  --paired -q 28 \</span><br><span class="line">        --phred33 --length 35 --stringency 4 \</span><br><span class="line">        --hardtrim5 $n_end5 \</span><br><span class="line">        --hardtrim3 $n_end3 \</span><br><span class="line">        -e=0.1 \</span><br><span class="line">        --cores $ncore \</span><br><span class="line">        -o $output $path_fastq/$fq1 $path_fastq/$fq2 \</span><br><span class="line">        &gt; $path_log/trimgalore_paired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">    echo "Done trim_galore of  $&#123;fq1&#125;,$&#123;fq2&#125;!"</span><br><span class="line">  elif [ -f $&#123;path_fastq&#125;/$&#123;fq&#125; ]; then</span><br><span class="line"><span class="meta">  #</span><span class="bash"> 单端</span></span><br><span class="line">      echo "find: $&#123;fq2&#125;"</span><br><span class="line">      time trim_galore -q 28 \</span><br><span class="line">          --phred33 --length 30 --stringency 4 \</span><br><span class="line">          --hardtrim5 $n_end5 \</span><br><span class="line">          --hardtrim3 $n_end3 \</span><br><span class="line">          -e=0.1 \</span><br><span class="line">          --cores $ncore \</span><br><span class="line">          -o $output $path_fastq/$fq &gt; \</span><br><span class="line">          $path_log/trimgalore_single_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">  else</span><br><span class="line">  	echo "Files not found: $&#123;fq1&#125;,$&#123;fq2&#125; or $&#123;fq&#125;"</span><br><span class="line">  fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End clean_trim_galore</span></span></span><br></pre></td></tr></table></figure>




<h2 id="cut5-fastq"><a href="#cut5-fastq" class="headerlink" title="cut5_fastq"></a>cut5_fastq</h2><blockquote>
<p>5’端进行剪切。本人早期脚本</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/cut5_fastq</span><br><span class="line">vim $nbin/cut5_fastq</span><br><span class="line">rm $nbin/cut5_fastq; cp2 $nbin</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function cut5_fastq</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> cut some bps from 5<span class="string">' end</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup cut5_fastq <span class="variable">$ws</span> &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/cut5_fastq/*<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_fastq=$ws/output/fastq</span><br><span class="line">path_res=$ws/output/cut5_fastq</span><br><span class="line">path_log=$ws/log/cut5_fastq</span><br><span class="line">suffix=.fastq.gz</span><br><span class="line">bp=5</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># log</span></span></span><br><span class="line">mkdir -p $path_log $path_res</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># programe</span></span></span><br><span class="line">ls $path_fastq | grep $suffix | while read id</span><br><span class="line">do</span><br><span class="line">echo "$path_fastq/$id"</span><br><span class="line">nohup time cutadapt \</span><br><span class="line">	-u $bp \</span><br><span class="line">    -o $&#123;path_res&#125;/$(basename $&#123;id&#125; $&#123;suffix&#125;)_cut5.fastq.gz \</span><br><span class="line">    $&#123;path_fastq&#125;/$&#123;id&#125; \</span><br><span class="line">    &gt; $path_log/cut5_fastq_$&#123;id&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End cut5_fastq</span></span></span><br></pre></td></tr></table></figure>



<h2 id="bwa-hg38"><a href="#bwa-hg38" class="headerlink" title="bwa_hg38"></a>bwa_hg38</h2><blockquote>
<p>通过bwa的mem算法，根据hg38基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/bwa_hg38</span><br><span class="line">vim $nbin/bwa_hg38</span><br><span class="line">rm $nbin/bwa_hg38; cp2 $nbin</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: bwa_hg38</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过bwa的mem算法，根据hg38基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</span></span><br><span class="line"><span class="meta"> #</span><span class="bash"> bwa mem: -t 线程数；-M mark shorter split hits as secondary？; 管道|后面的-代表管道前的输出;-R @RG对于下一步的GATK碱基质量重校正BQSR最佳实践是需要的。要事先检查一下@RG是否存在</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">index_base=/data/reference/index/bwa/hg38/hg38 # 注意这不是文件路径而是文件前缀</span><br><span class="line">ws=$1</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/clean-trim_galore # 自定义。有时是clean-trim_galore</span><br><span class="line">path_log=$&#123;ws&#125;/log/bwa_hg38</span><br><span class="line">path_align=$&#123;ws&#125;/output/align</span><br><span class="line">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span><br><span class="line">seq_type=WXS</span><br><span class="line">suffix=val</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup bwa_hg38 <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/bwa_hg38/bwa_hg38.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/bwa_hg38/bwa_hg38.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># dir &amp; log</span></span></span><br><span class="line">mkdir -p $path_log $path_align</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle <span class="keyword">for</span> bwa mem</span></span><br><span class="line">cat $path_sra2fq_txt | while read id </span><br><span class="line">do</span><br><span class="line">    # INV_P10_Exon_2_val_2.fq.gz</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    case=$&#123;arr[1]&#125;</span><br><span class="line">    fq1=$path_fastq/$&#123;case&#125;_1_$&#123;suffix&#125;_1.fq.gz</span><br><span class="line">    fq2=$path_fastq/$&#123;case&#125;_2_$&#123;suffix&#125;_2.fq.gz</span><br><span class="line">    fq=$path_fastq/$&#123;case&#125;_$&#123;suffix&#125;.fq.gz</span><br><span class="line">    </span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;.bam ]; then </span><br><span class="line">    	echo `date`"$&#123;case&#125;: bam had existed. Ignore!"</span><br><span class="line">    elif [ -f $fq1 -a -f $fq2 ]; then</span><br><span class="line">    	## bwa/samtools</span><br><span class="line">    	echo `date`"$&#123;case&#125;: Create bam for paired data..."</span><br><span class="line">        echo `date`"$&#123;case&#125;: Do bwa mem | samtools sort ..."</span><br><span class="line">        time bwa mem -M -R "@RG\tID:$&#123;case&#125;\tSM:$&#123;case&#125;\tLB:$&#123;seq_type&#125;\tPL:Illumina" -t $nthread $index_base $fq1 $fq2 | samtools sort -@ $nthread -m 1G --reference $path_genome_fa -o $path_align/$&#123;case&#125;.bam - </span><br><span class="line">        samtools index -@ $nthread $path_align/$&#123;case&#125;.bam</span><br><span class="line">        echo `date`"$&#123;case&#125;: bwa/samtools done!"</span><br><span class="line">    elif [ -f $fq -a ! -f $path_align/$&#123;case&#125;.bam ]; then</span><br><span class="line">    	echo `date`"$&#123;case&#125;: Create bam for unpaired data ..."</span><br><span class="line">        echo `date`"$&#123;case&#125;: Do bwa mem | samtools sort ..."</span><br><span class="line">        time bwa mem -M -R "@RG\tID:$&#123;case&#125;\tSM:$&#123;case&#125;\tLB:$&#123;seq_type&#125;\tPL:Illumina" -t $nthread $index_base $fq | samtools sort -@ $nthread -m 1G --reference $path_genome_fa -o $path_align/$&#123;case&#125;.bam - </span><br><span class="line">        samtools index -@ $nthread $path_align/$&#123;case&#125;.bam</span><br><span class="line">        echo `date`"$&#123;case&#125;: bwa/samtools done!"</span><br><span class="line">    else</span><br><span class="line">    	echo `date`"$&#123;case&#125;: not available fastq.gz file. Ignored!"</span><br><span class="line">    fi  </span><br><span class="line">    </span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"> End bwa_hg38</span></span><br></pre></td></tr></table></figure>



<h2 id="bwa-mem"><a href="#bwa-mem" class="headerlink" title="bwa_mem"></a>bwa_mem</h2><blockquote>
<p>通过bwa的mem算法，根据hg38基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/bwa_mem</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: bwa_mem</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过bwa的mem算法，根据hg38基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</span></span><br><span class="line"><span class="meta"> #</span><span class="bash"> bwa mem: -t 线程数；-M mark shorter split hits as secondary？; 管道|后面的-代表管道前的输出;-R @RG对于下一步的GATK碱基质量重校正BQSR最佳实践是需要的。要事先检查一下@RG是否存在</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">ws=<span class="variable">$1</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">index_base=/data/reference/index/bwa/hg38/hg38 <span class="comment"># 注意这不是文件路径而是文件前缀</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">path_sra2fq_txt=<span class="variable">$&#123;ws&#125;</span>/input/<span class="keyword">case</span>/sra2case.txt</span></span><br><span class="line"><span class="meta">#</span><span class="bash">path_fastq=<span class="variable">$&#123;ws&#125;</span>/output/clean-trim_galore <span class="comment"># 自定义。有时是clean-trim_galore</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">path_log=<span class="variable">$&#123;ws&#125;</span>/<span class="built_in">log</span>/bwa_mem</span></span><br><span class="line"><span class="meta">#</span><span class="bash">path_align=<span class="variable">$&#123;ws&#125;</span>/output/align</span></span><br><span class="line"><span class="meta">#</span><span class="bash">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span></span><br><span class="line"><span class="meta">#</span><span class="bash">seq_type=WXS</span></span><br><span class="line"><span class="meta">#</span><span class="bash">suffix=val</span></span><br><span class="line"><span class="meta">#</span><span class="bash">nthread=48</span></span><br><span class="line">ws=$1</span><br><span class="line">index_base=$2 # 注意这不是文件路径而是文件前缀</span><br><span class="line">path_genome_fa=$3</span><br><span class="line">path_sra2fq_txt=$4</span><br><span class="line">path_fastq=$5 # 自定义。有时是clean-trim_galore</span><br><span class="line">path_align=$6</span><br><span class="line">nthread=$7</span><br><span class="line">seq_type=$8</span><br><span class="line">suffix=$9</span><br><span class="line">path_log=$&#123;10&#125;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">nohup bwa_mem <span class="variable">$ws</span> <span class="variable">$&#123;index_base&#125;</span> <span class="variable">$&#123;path_genome_fa&#125;</span> \</span></span><br><span class="line"><span class="meta">#</span><span class="bash">    <span class="variable">$&#123;path_sra2fq_txt&#125;</span> <span class="variable">$&#123;path_fastq&#125;</span> <span class="variable">$&#123;path_align&#125;</span> <span class="variable">$&#123;nthread&#125;</span> \</span></span><br><span class="line"><span class="meta">#</span><span class="bash">    <span class="variable">$&#123;seq_type&#125;</span> <span class="variable">$&#123;suffix&#125;</span> <span class="variable">$&#123;path_log&#125;</span> \</span></span><br><span class="line"><span class="meta">#</span><span class="bash">    &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/bwa_mem.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/bwa_mem.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># dir &amp; log</span></span></span><br><span class="line">mkdir -p $path_log $path_align</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle <span class="keyword">for</span> bwa mem</span></span><br><span class="line">cat $path_sra2fq_txt | while read id </span><br><span class="line">do</span><br><span class="line">    # INV_P10_Exon_2_val_2.fq.gz</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    case=$&#123;arr[1]&#125; #case=PBS</span><br><span class="line">    fq1=$path_fastq/$&#123;case&#125;_1_$&#123;suffix&#125;_1.fq.gz</span><br><span class="line">    fq2=$path_fastq/$&#123;case&#125;_2_$&#123;suffix&#125;_2.fq.gz</span><br><span class="line">    fq=$path_fastq/$&#123;case&#125;_$&#123;suffix&#125;.fq.gz</span><br><span class="line">    </span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;.bam ]; then </span><br><span class="line">    	echo `date`"$&#123;case&#125;: bam had existed. Ignore!"</span><br><span class="line">    elif [ -f $fq1 -a -f $fq2 ]; then</span><br><span class="line">    	## bwa/samtools</span><br><span class="line">    	echo `date`"$&#123;case&#125;: Create bam for paired data..."</span><br><span class="line">        echo `date`"$&#123;case&#125;: Do bwa mem | samtools sort ..."</span><br><span class="line">        #  bwa mem</span><br><span class="line">        bwa mem -M \</span><br><span class="line">            -R "@RG\tID:$&#123;case&#125;\tSM:$&#123;case&#125;\tLB:$&#123;seq_type&#125;\tPL:Illumina" \</span><br><span class="line">            -t $nthread \</span><br><span class="line">            $index_base $fq1 $fq2 \</span><br><span class="line">            &gt; $path_align/$&#123;case&#125;_unsort.bam</span><br><span class="line">        # samtools</span><br><span class="line">        samtools sort -@ $nthread -m 1G \</span><br><span class="line">            --reference $path_genome_fa \</span><br><span class="line">            -o $path_align/$&#123;case&#125;.bam \</span><br><span class="line">            $path_align/$&#123;case&#125;_unsort.bam </span><br><span class="line">        echo `date`"$&#123;case&#125;: bwa/samtools done!"</span><br><span class="line">        if [ -f $path_align/$&#123;case&#125;.bam ]; then rm $path_align/$&#123;case&#125;_unsort.bam ; fi</span><br><span class="line">    elif [ -f $fq -a ! -f $path_align/$&#123;case&#125;.bam ]; then</span><br><span class="line">    	echo `date`"$&#123;case&#125;: Create bam for unpaired data ..."</span><br><span class="line">        echo `date`"$&#123;case&#125;: Do bwa mem | samtools sort ..."</span><br><span class="line">        # bwa mem</span><br><span class="line">        bwa mem -M \</span><br><span class="line">            -R "@RG\tID:$&#123;case&#125;\tSM:$&#123;case&#125;\tLB:$&#123;seq_type&#125;\tPL:Illumina" \</span><br><span class="line">            -t $nthread $index_base $fq \</span><br><span class="line">            &gt; $path_align/$&#123;case&#125;_unsort.bam</span><br><span class="line">        # sort    </span><br><span class="line">        samtools sort -@ $nthread -m 1G \</span><br><span class="line">            --reference $path_genome_fa \</span><br><span class="line">            -o $path_align/$&#123;case&#125;.bam \</span><br><span class="line">            $path_align/$&#123;case&#125;_unsort.bam</span><br><span class="line">        echo `date`"$&#123;case&#125;: bwa/samtools done!"</span><br><span class="line">        if [ -f $path_align/$&#123;case&#125;.bam ]; then rm $path_align/$&#123;case&#125;_unsort.bam ; fi</span><br><span class="line">    else</span><br><span class="line">    	echo `date`"$&#123;case&#125;: not available fastq.gz file. Ignored!"</span><br><span class="line">    fi  </span><br><span class="line">   </span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"> End bwa_hg38</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">###============================Example======================####</span></span></span><br><span class="line"></span><br><span class="line">index_base=/data/reference/index/bwa/hg38/hg38 # 注意这不是文件路径而是文件前缀</span><br><span class="line">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span><br><span class="line">seq_type=WGS</span><br><span class="line"></span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/clean-trim_galore # 自定义。有时是clean-trim_galore</span><br><span class="line">path_log=$&#123;ws&#125;/log/bwa_hg38</span><br><span class="line">path_align=$&#123;ws&#125;/output/align/bwa</span><br><span class="line">suffix=trimmed; nthread=32</span><br><span class="line"></span><br><span class="line">nohup bwa_mem $ws $&#123;index_base&#125; $&#123;path_genome_fa&#125; \</span><br><span class="line">    $&#123;path_sra2fq_txt&#125; $&#123;path_fastq&#125; $&#123;path_align&#125; $&#123;nthread&#125; \</span><br><span class="line">    $&#123;seq_type&#125; $&#123;suffix&#125; $&#123;path_log&#125; \</span><br><span class="line">    &gt; $ws/log/bwa_mem.log 2&gt;&amp;1 &amp;</span><br><span class="line">tail -f $ws/log/bwa_mem.log</span><br></pre></td></tr></table></figure>

<h2 id="hisat2-hg38"><a href="#hisat2-hg38" class="headerlink" title="hisat2_hg38"></a>hisat2_hg38</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/hisat2_hg38</span><br><span class="line">vim $nbin/hisat2_hg38</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: hisat2_hg38</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过hisat2，根据hg38基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">index_base=/data/reference/index/hisat2/hg38/genome # 注意这不是文件路径而是文件前缀</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/clean-trim_galore</span><br><span class="line">path_log=$&#123;ws&#125;/log/hisat2_hg38</span><br><span class="line">path_align=$&#123;ws&#125;/output/align/hisat2</span><br><span class="line">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span><br><span class="line">seq_type=RNA-Seq # @RG</span><br><span class="line">fq_postfix=val</span><br><span class="line"><span class="meta">#</span><span class="bash">fq_postfix=_trimmed</span></span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup hisat2_hg38 <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/hisat2_hg38/hisat2_hg38.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/hisat2_hg38/hisat2_hg38.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programs: </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> dir &amp; <span class="built_in">log</span></span></span><br><span class="line">mkdir -p $path_log $path_align</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle <span class="keyword">for</span> bwa mem</span></span><br><span class="line">cat $path_sra2fq_txt |while read id </span><br><span class="line"><span class="meta">#</span><span class="bash"> id=<span class="string">"SRR1039508 N61311-Untreated GSM1275862 N61311-Untreated-GSM1275862"</span></span></span><br><span class="line">do</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    case=$&#123;arr[1]&#125;</span><br><span class="line">    fq1=$path_fastq/$&#123;case&#125;_1_$&#123;fq_postfix&#125;_1.fq.gz</span><br><span class="line">    fq2=$path_fastq/$&#123;case&#125;_2_$&#123;fq_postfix&#125;_2.fq.gz</span><br><span class="line">    fq=$path_fastq/$&#123;case&#125;$&#123;fq_postfix&#125;.fq.gz</span><br><span class="line">    </span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;.bam ]; then</span><br><span class="line">    	echo "$path_align/$&#123;case&#125;.bam had existed.Ignore"</span><br><span class="line">    elif [ -f $fq1 -a -f $fq2 ]; then</span><br><span class="line">        echo `date`" $&#123;case&#125;:paired reads data.do hisat2 analysis..."</span><br><span class="line">        # hisat2: -t 输出时间；-p 线程数；-x index前缀；-1/-2 fq文件 -S 输出sam文件;--rg-id 设置read group id. @RG对于下一步的GATK碱基质量重校正BQSR最佳实践是需要的。</span><br><span class="line">        hisat2 -t -p $nthread -x $index_base\</span><br><span class="line">        	-1 $fq1 -2 $fq2 \</span><br><span class="line">            --rg-id "$&#123;case&#125;" \</span><br><span class="line">            --rg "SM:$&#123;case&#125;" --rg "LB:$&#123;seq_type&#125;" --rg "PL:Illumina" \</span><br><span class="line">            -S $path_align/$&#123;case&#125;.sam \</span><br><span class="line">            &gt;$path_log/hisat2_hg38_paired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">         echo `date`" $&#123;case&#125;:hisat2 done!"</span><br><span class="line"></span><br><span class="line">        # samtools view: -q 只纳入质量在此数字以上的reads。转录本需要30吗？</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools view..."</span><br><span class="line">        samtools view -@ $nthread \</span><br><span class="line">            -bhS -q 30 \</span><br><span class="line">            $path_align/$&#123;case&#125;.sam \</span><br><span class="line">            &gt; $path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools view done!"</span><br><span class="line"></span><br><span class="line">        # samtools sort: -m设定每个线程的最大内存</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools sort... "</span><br><span class="line">        samtools sort -@ $nthread -m 1G \</span><br><span class="line">            --reference $path_genome_fa \</span><br><span class="line">            -o $path_align/$&#123;case&#125;.bam \</span><br><span class="line">            $path_align/$&#123;case&#125;_nonsort.bam \</span><br><span class="line">            &gt; $path_log/$&#123;case&#125;.samtools_sort_hg38.log 2&gt;&amp;1</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools sort done!"</span><br><span class="line">        </span><br><span class="line">        # index</span><br><span class="line">        samtools index -@ $nthread $path_align/$&#123;case&#125;.bam</span><br><span class="line">    </span><br><span class="line">    else</span><br><span class="line">        echo `date`" $&#123;case&#125;:unpaired reads data.do hisat2 analysis..."</span><br><span class="line">        hisat2 -t -p $nthread -x $index_base \</span><br><span class="line">            -U $fq \</span><br><span class="line">            --rg-id "$&#123;case&#125;" \</span><br><span class="line">            --rg "SM:$&#123;case&#125;"  --rg "LB:$&#123;seq_type&#125;"  --rg "PL:Illumina" \</span><br><span class="line">            -S $path_align/$&#123;case&#125;.sam \</span><br><span class="line">            &gt;$path_log/hisat2_hg38_unpaired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">        echo `date`" $&#123;case&#125;:hisat2 done!"</span><br><span class="line"></span><br><span class="line">        # samtools view: -q 只纳入质量在此数字以上的reads。转录本需要30吗？</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools view..."</span><br><span class="line">        samtools view -@ $nthread \</span><br><span class="line">            -bhS -q 30 \</span><br><span class="line">            $path_align/$&#123;case&#125;.sam \</span><br><span class="line">            &gt; $path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools view done!"</span><br><span class="line"></span><br><span class="line">        # samtools sort: -m设定每个线程的最大内存</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools sort... "</span><br><span class="line">        samtools sort -@ $nthread \</span><br><span class="line">            -m 1G --reference $path_genome_fa \</span><br><span class="line">            -o $path_align/$&#123;case&#125;.bam \</span><br><span class="line">            $path_align/$&#123;case&#125;_nonsort.bam \</span><br><span class="line">            &gt;$path_log/$&#123;case&#125;.samtools_sort_hg38.log 2&gt;&amp;1</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools sort done!"</span><br><span class="line">        </span><br><span class="line">        # index</span><br><span class="line">        samtools index -@ $nthread $path_align/$&#123;case&#125;.bam</span><br><span class="line">    fi</span><br><span class="line">    </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># file management</span></span></span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;_nonsort.bam ]; then </span><br><span class="line">        rm $path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">    fi</span><br><span class="line"></span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;.sam ]; then</span><br><span class="line">        rm $path_align/$&#123;case&#125;.sam</span><br><span class="line">    fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End hisat2_hg38</span></span></span><br></pre></td></tr></table></figure>



<h2 id="hisat2-mm10"><a href="#hisat2-mm10" class="headerlink" title="hisat2_mm10"></a>hisat2_mm10</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/hisat2_mm10</span><br><span class="line">vim $nbin/hisat2_mm10</span><br><span class="line">rm $nbin/hisat2_mm10;cp2 $nbin</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: hisat2_mm10</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过hisat2，根据mm10基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">index_base=/data/reference/mouse/mm10/index/hisat2/mm10 # 注意这不是文件路径而是文件前缀</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/clean-trim_galore</span><br><span class="line">path_log=$&#123;ws&#125;/log/hisat2_mm10</span><br><span class="line">path_align=$&#123;ws&#125;/output/align</span><br><span class="line">path_genome_fa=/data/reference/mouse/mm10/genome/mm10.fa</span><br><span class="line">seq_type=RNA-Seq # @RG</span><br><span class="line">fq_postfix=trimmed</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup hisat2_mm10 <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/hisat2_mm10.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/hisat2_mm10.log</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/hisat2_mm10/*.<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programs: </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> dir &amp; <span class="built_in">log</span></span></span><br><span class="line">mkdir -p $path_log $path_align</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle <span class="keyword">for</span> bwa mem</span></span><br><span class="line">cat $path_sra2fq_txt |while read id </span><br><span class="line"><span class="meta">#</span><span class="bash"> id=<span class="string">"SRR1039508 N61311-Untreated GSM1275862 N61311-Untreated-GSM1275862"</span></span></span><br><span class="line">do</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    case=$&#123;arr[1]&#125;</span><br><span class="line">    fq1=$path_fastq/$&#123;case&#125;_1_$&#123;fq_postfix&#125;_1.fq.gz</span><br><span class="line">    fq2=$path_fastq/$&#123;case&#125;_2_$&#123;fq_postfix&#125;_2.fq.gz</span><br><span class="line">    fq=$path_fastq/$&#123;case&#125;_$&#123;fq_postfix&#125;.fq.gz</span><br><span class="line">    </span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;.bam ]; then</span><br><span class="line">    	echo "$path_align/$&#123;case&#125;.bam had existed.Ignore"</span><br><span class="line">    elif [ -f $fq1 -a -f $fq2 ]</span><br><span class="line">    then</span><br><span class="line">        echo `date`" $&#123;case&#125;:paired reads data.do hisat2 analysis..."</span><br><span class="line">        # hisat2: -t 输出时间；-p 线程数；-x index前缀；-1/-2 fq文件 -S 输出sam文件;--rg-id 设置read group id. @RG对于下一步的GATK碱基质量重校正BQSR最佳实践是需要的。</span><br><span class="line">        hisat2 -t -p $nthread -x $index_base -1 $fq1 -2 $fq2 \</span><br><span class="line">            --rg-id "$&#123;case&#125;" \</span><br><span class="line">            --rg "SM:$&#123;case&#125;" --rg "LB:$&#123;seq_type&#125;" --rg "PL:Illumina" \</span><br><span class="line">            -S $path_align/$&#123;case&#125;.sam \</span><br><span class="line">            &gt;$path_log/hisat2_mm10_paired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">         echo `date`" $&#123;case&#125;:hisat2 done!"</span><br><span class="line"></span><br><span class="line">        # samtools view: -q 只纳入质量在此数字以上的reads。转录本需要30吗？</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools view..."</span><br><span class="line">        samtools view -@ $nthread -bhS -q 30 $path_align/$&#123;case&#125;.sam &gt;$path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools view done!"</span><br><span class="line"></span><br><span class="line">        # samtools sort: -m设定每个线程的最大内存</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools sort... "</span><br><span class="line">        samtools sort -@ $nthread -m 1G --reference $path_genome_fa -o $path_align/$&#123;case&#125;.bam $path_align/$&#123;case&#125;_nonsort.bam &gt;$path_log/$&#123;case&#125;.samtools_sort_mm10.log 2&gt;&amp;1</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools sort done!"</span><br><span class="line"></span><br><span class="line">        # index</span><br><span class="line">        samtools index -@ $nthread $path_align/$&#123;case&#125;.bam</span><br><span class="line">    </span><br><span class="line">    else</span><br><span class="line">        echo `date`" $&#123;case&#125;:unpaired reads data.do hisat2 analysis..."</span><br><span class="line">        hisat2 -t -p $nthread -x $index_base -U $fq \</span><br><span class="line">        --rg-id "$&#123;case&#125;" \</span><br><span class="line">        --rg "SM:$&#123;case&#125;"  --rg "LB:$&#123;seq_type&#125;"  --rg "PL:Illumina" \</span><br><span class="line">        -S $path_align/$&#123;case&#125;.sam \</span><br><span class="line">        &gt;$path_log/hisat2_mm10_unpaired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">        echo `date`" $&#123;case&#125;:hisat2 done!"</span><br><span class="line"></span><br><span class="line">        # samtools view: -q 只纳入质量在此数字以上的reads。转录本需要30吗？</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools view..."</span><br><span class="line">        samtools view -@ $nthread -bhS -q 30 $path_align/$&#123;case&#125;.sam &gt;$path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools view done!"</span><br><span class="line"></span><br><span class="line">        # samtools sort: -m设定每个线程的最大内存</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools sort... "</span><br><span class="line">        samtools sort -@ $nthread -m 1G --reference $path_genome_fa -o $path_align/$&#123;case&#125;.bam $path_align/$&#123;case&#125;_nonsort.bam &gt;$path_log/$&#123;case&#125;.samtools_sort_mm10.log 2&gt;&amp;1</span><br><span class="line">        echo `date`" $&#123;case&#125;:samtools sort done!"</span><br><span class="line"></span><br><span class="line">        # index</span><br><span class="line">        samtools index -@ $nthread $path_align/$&#123;case&#125;.bam</span><br><span class="line">    fi</span><br><span class="line">    </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># file management</span></span></span><br><span class="line">if [ -f $path_align/$&#123;case&#125;_nonsort.bam ]; then rm $path_align/$&#123;case&#125;_nonsort.bam; fi</span><br><span class="line">if [ -f $path_align/$&#123;case&#125;.sam ]; then rm $path_align/$&#123;case&#125;.sam; fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End hisat2_mm10</span></span></span><br></pre></td></tr></table></figure>


<h2 id="hisat2-miRBase"><a href="#hisat2-miRBase" class="headerlink" title="hisat2_miRBase"></a>hisat2_miRBase</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/hisat2_miRBase</span><br><span class="line">vim $nbin/hisat2_miRBase</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: hisat2_miRBase</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过hisat2，根据miRBase基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">index_base=/data/reference/miRNA/release_22/hisat2 # 注意这不是文件路径而是文件前缀</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/clean-trim_galore</span><br><span class="line">path_log=$&#123;ws&#125;/log/hisat2_miRBase</span><br><span class="line">path_align=$&#123;ws&#125;/output/align</span><br><span class="line">path_genome_fa=/data/reference/miRNA/release_22/hairpin.fa</span><br><span class="line">seq_type=smallRNA-Seq # @RG</span><br><span class="line">fq_postfix=_trimmed</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup hisat2_miRBase <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/hisat2_miRBase.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/hisat2_miRBase.log</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> less <span class="variable">$ws</span>/<span class="built_in">log</span>/hisat2_miRBase/*.<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programs: </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> dir &amp; <span class="built_in">log</span></span></span><br><span class="line">mkdir -p $path_log $path_align</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle <span class="keyword">for</span> bwa mem</span></span><br><span class="line">cat $path_sra2fq_txt |while read id </span><br><span class="line"><span class="meta">#</span><span class="bash"> id=<span class="string">"SRR8560004 PBS GSM3598111"</span></span></span><br><span class="line">do</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    case=$&#123;arr[1]&#125;</span><br><span class="line">    fq1=$path_fastq/$&#123;case&#125;_1_$&#123;fq_postfix&#125;_1.fq.gz</span><br><span class="line">    fq2=$path_fastq/$&#123;case&#125;_2_$&#123;fq_postfix&#125;_2.fq.gz</span><br><span class="line">    fq=$path_fastq/$&#123;case&#125;$&#123;fq_postfix&#125;.fq.gz</span><br><span class="line">    </span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;_mirbase.bam ]; then</span><br><span class="line">    echo "$path_align/$&#123;case&#125;_mirbase.bam had existed.Ignore"</span><br><span class="line">    elif [ -f $fq1 -a -f $fq2 ]</span><br><span class="line">    then</span><br><span class="line">    echo `date`" $&#123;case&#125;_mirbase:paired reads data.do hisat2 analysis..."</span><br><span class="line">    # hisat2: -t 输出时间；-p 线程数；-x index前缀；-1/-2 fq文件 -S 输出sam文件;--rg-id 设置read group id. @RG对于下一步的GATK碱基质量重校正BQSR最佳实践是需要的。</span><br><span class="line">    hisat2 -t -p $nthread -x $index_base -1 $fq1 -2 $fq2 \</span><br><span class="line">    --rg-id "$&#123;case&#125;" \</span><br><span class="line">    --rg "SM:$&#123;case&#125;" --rg "LB:$&#123;seq_type&#125;" --rg "PL:Illumina" \</span><br><span class="line">    -S $path_align/$&#123;case&#125;_mirbase.sam \</span><br><span class="line">    &gt;$path_log/hisat2_mirbase_mirbase_paired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">     echo `date`" $&#123;case&#125;:hisat2 done!"</span><br><span class="line">    </span><br><span class="line">    # samtools view: -q 只纳入质量在此数字以上的reads。转录本需要30吗？</span><br><span class="line">    echo `date`" $&#123;case&#125;_mirbase:samtools view..."</span><br><span class="line">    samtools view -@ $nthread -bhS -q 30 \</span><br><span class="line">    $path_align/$&#123;case&#125;_mirbase.sam \</span><br><span class="line">    &gt;$path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">    echo `date`" $&#123;case&#125;:samtools view done!"</span><br><span class="line">    </span><br><span class="line">    # samtools sort: -m设定每个线程的最大内存</span><br><span class="line">    echo `date`" $&#123;case&#125;:samtools sort... "</span><br><span class="line">    samtools sort -@ $nthread -m 1G \</span><br><span class="line">    --reference $path_genome_fa \</span><br><span class="line">    -o $path_align/$&#123;case&#125;_mirbase.bam \</span><br><span class="line">    $path_align/$&#123;case&#125;_nonsort.bam \</span><br><span class="line">    &gt;$path_log/$&#123;case&#125;.samtools_sort.log 2&gt;&amp;1</span><br><span class="line">    echo `date`" $&#123;case&#125;:samtools sort done!"</span><br><span class="line">    </span><br><span class="line">    else</span><br><span class="line">    echo `date`" $&#123;case&#125;:unpaired reads data.do hisat2 analysis..."</span><br><span class="line">    hisat2 -t -p $nthread \</span><br><span class="line">    -x $index_base -U $fq \</span><br><span class="line">    --rg-id "$&#123;case&#125;" \</span><br><span class="line">    --rg "SM:$&#123;case&#125;"  --rg "LB:$&#123;seq_type&#125;"  --rg "PL:Illumina" \</span><br><span class="line">    -S $path_align/$&#123;case&#125;_mirbase.sam \</span><br><span class="line">    &gt;$path_log/hisat2_mirbase_unpaired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">    echo `date`" $&#123;case&#125;:hisat2 done!"</span><br><span class="line">    </span><br><span class="line">    # samtools view: -q 只纳入质量在此数字以上的reads。转录本需要30吗？</span><br><span class="line">    echo `date`" $&#123;case&#125;:samtools view..."</span><br><span class="line">    samtools view -@ $nthread -bhS \</span><br><span class="line">    -q 30 $path_align/$&#123;case&#125;_mirbase.sam \</span><br><span class="line">    &gt;$path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">    echo `date`" $&#123;case&#125;:samtools view done!"</span><br><span class="line">    </span><br><span class="line">    # samtools sort: -m设定每个线程的最大内存</span><br><span class="line">    echo `date`" $&#123;case&#125;:samtools sort... "</span><br><span class="line">    samtools sort -@ $nthread -m 1G \</span><br><span class="line">    --reference $path_genome_fa \</span><br><span class="line">    -o $path_align/$&#123;case&#125;_mirbase.bam \</span><br><span class="line">    $path_align/$&#123;case&#125;_nonsort.bam \</span><br><span class="line">    &gt;$path_log/$&#123;case&#125;_mirbase_samtools_sort.log 2&gt;&amp;1</span><br><span class="line">    echo `date`" $&#123;case&#125;:samtools sort done!"</span><br><span class="line">    fi</span><br><span class="line">    </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># file management</span></span></span><br><span class="line">if [ -f $path_align/$&#123;case&#125;_nonsort.bam ]; then rm $path_align/$&#123;case&#125;_nonsort.bam; fi</span><br><span class="line">if [ -f $path_align/$&#123;case&#125;_mirbase.sam ]; then rm $path_align/$&#123;case&#125;_mirbase.sam; fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End hisat2_miRBase</span></span></span><br></pre></td></tr></table></figure>



<h2 id="STAR-hg38"><a href="#STAR-hg38" class="headerlink" title="STAR_hg38"></a>STAR_hg38</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/star_hg38</span><br><span class="line">cp2 $nbin; vim $nbin/star_hg38</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: star_hg38</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过star，根据hg38基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">index_base=/data/reference/index/star/hg38 # 注意这不是文件路径而是文件前缀</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/clean-trim_galore</span><br><span class="line">path_log=$&#123;ws&#125;/log/star_hg38</span><br><span class="line">path_align=$&#123;ws&#125;/output/align/STAR</span><br><span class="line">path_gtf=/data/reference/gtf/raw/hg38_gencode.v32.annotation.gtf # gft.gz not supported</span><br><span class="line"><span class="meta">#</span><span class="bash"> path_genome_fa=/data/reference/genome/hg38/hg38.fa</span></span><br><span class="line">seq_type=smallRNA-Seq # @RG</span><br><span class="line"><span class="meta">#</span><span class="bash"> fq_postfix=val</span></span><br><span class="line">fq_postfix=_trimmed</span><br><span class="line">sjdbOverhang=100</span><br><span class="line">nthread=48</span><br><span class="line">raw_ulimit=$(ulimit -n); ulimit -n 10000 # default is 1024, which is not enough for STAR analysis</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup star_hg38 <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/star_hg38.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/star_hg38.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programs: </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> dir &amp; <span class="built_in">log</span></span></span><br><span class="line">mkdir -p $path_log $path_align</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle <span class="keyword">for</span> STAR</span></span><br><span class="line">cat $path_sra2fq_txt | while read id ; do</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    case=$&#123;arr[1]&#125; # case=PBS</span><br><span class="line">    fq1=$path_fastq/$&#123;case&#125;_1_$&#123;fq_postfix&#125;_1.fq.gz</span><br><span class="line">    fq2=$path_fastq/$&#123;case&#125;_2_$&#123;fq_postfix&#125;_2.fq.gz</span><br><span class="line">    fq=$path_fastq/$&#123;case&#125;$&#123;fq_postfix&#125;.fq.gz</span><br><span class="line">    </span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;Aligned.sortedByCoord.out.bam ]; then</span><br><span class="line">    	echo "$&#123;case&#125;: STAR results had existed. Ignore!"</span><br><span class="line">    elif [ -f $fq1 -a -f $fq2 ]; then</span><br><span class="line">        echo `date`" $&#123;case&#125;:paired reads data.do STAR analysis..."</span><br><span class="line">        # hisat2: -t 输出时间；-p 线程数；-x index前缀；-1/-2 fq文件 -S 输出sam文件;--rg-id 设置read group id. @RG对于下一步的GATK碱基质量重校正BQSR最佳实践是需要的。</span><br><span class="line">        STAR \</span><br><span class="line">            --runThreadN $&#123;nthread&#125; \</span><br><span class="line">            --genomeDir $&#123;index_base&#125; \</span><br><span class="line">            --readFilesIn $&#123;fq1&#125; $&#123;fq2&#125; \</span><br><span class="line">            --readFilesCommand  zcat \</span><br><span class="line">            --sjdbGTFfile $&#123;path_gtf&#125; \</span><br><span class="line">            --sjdbOverhang $&#123;sjdbOverhang&#125; \</span><br><span class="line">            --outSAMattrRGline ID:$&#123;case&#125; SM:$&#123;case&#125; LB:$&#123;seq_type&#125; PL:Illumina \</span><br><span class="line">            --outFileNamePrefix $&#123;path_align&#125;/$&#123;case&#125; \</span><br><span class="line">            --outSAMtype BAM SortedByCoordinate \</span><br><span class="line">            --twopassMode Basic \</span><br><span class="line">            &gt; $path_log/STAR_hg38_paired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">         echo `date`" $&#123;case&#125;:STAR done!"</span><br><span class="line"></span><br><span class="line">        ## index</span><br><span class="line">        path_bam=$&#123;path_align&#125;/$&#123;case&#125;Aligned.sortedByCoord.out.bam</span><br><span class="line">        samtools index -@ $nthread $&#123;path_bam&#125;</span><br><span class="line">    </span><br><span class="line">    else</span><br><span class="line">        echo `date`" $&#123;case&#125;:unpaired reads data.do STAR analysis..."</span><br><span class="line">        STAR \</span><br><span class="line">            --runThreadN $&#123;nthread&#125; \</span><br><span class="line">            --genomeDir $&#123;index_base&#125; \</span><br><span class="line">            --readFilesIn $&#123;fq&#125; \</span><br><span class="line">            --readFilesCommand  zcat \</span><br><span class="line">            --sjdbGTFfile $&#123;path_gtf&#125; \</span><br><span class="line">            --sjdbOverhang $&#123;sjdbOverhang&#125; \</span><br><span class="line">            --outSAMattrRGline ID:$&#123;case&#125; SM:$&#123;case&#125; LB:$&#123;seq_type&#125; PL:Illumina \</span><br><span class="line">            --outFileNamePrefix $&#123;path_align&#125;/$&#123;case&#125; \</span><br><span class="line">            --outSAMtype BAM SortedByCoordinate \</span><br><span class="line">            --twopassMode Basic \</span><br><span class="line">            &gt; $path_log/STAR_hg38_unpaired_$&#123;case&#125;.log 2&gt;&amp;1</span><br><span class="line">            # tail -f $path_log/STAR_hg38_unpaired_$&#123;case&#125;.log</span><br><span class="line">            </span><br><span class="line">        echo `date`" $&#123;case&#125;:STAR done!"</span><br><span class="line"></span><br><span class="line">        ## index</span><br><span class="line">        path_bam=$&#123;path_align&#125;/$&#123;case&#125;Aligned.sortedByCoord.out.bam</span><br><span class="line">        samtools index -@ $nthread $&#123;path_bam&#125;</span><br><span class="line">    fi</span><br><span class="line">done    </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># default ulimit</span></span></span><br><span class="line">ulimit -n $&#123;raw_ulimit&#125;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End STAR_hg38</span></span></span><br></pre></td></tr></table></figure>



<h2 id="bowtie2-hg38"><a href="#bowtie2-hg38" class="headerlink" title="bowtie2_hg38"></a>bowtie2_hg38</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/bowtie2_hg38</span><br><span class="line">vim $nbin/bowtie2_hg38</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: bowtie2_hg38</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过bowtie2算法，根据hg38基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">index_base=/data/reference/index/bowtie/hg38/hg38 # 注意这不是文件路径而是文件前缀</span><br><span class="line">ws=$1</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/clean-trim_galore</span><br><span class="line">path_log=$&#123;ws&#125;/log/bowtie2_hg38</span><br><span class="line">path_align=$&#123;ws&#125;/output/align/bowtie2</span><br><span class="line">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span><br><span class="line">seq_type=Chip-Seq</span><br><span class="line">fq_postfix=_trimmed</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup bowtie2_hg38 <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/bowtie2_hg38.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/bwa_hg38.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># dir &amp; log</span></span></span><br><span class="line">mkdir -p $path_log $path_align</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle <span class="keyword">for</span> bwa mem</span></span><br><span class="line">cat $path_sra2fq_txt | while read id </span><br><span class="line">do</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    case=$&#123;arr[1]&#125;</span><br><span class="line">    fq1=$path_fastq/$&#123;case&#125;_1_$&#123;fq_postfix&#125;_1.fq.gz</span><br><span class="line">    fq2=$path_fastq/$&#123;case&#125;_2_$&#123;fq_postfix&#125;_2.fq.gz</span><br><span class="line">    fq=$path_fastq/$&#123;case&#125;$&#123;fq_postfix&#125;.fq.gz</span><br><span class="line">    </span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;.bam ]; then</span><br><span class="line">    	echo `date`"$path_align/$&#123;case&#125;.bam had existed."</span><br><span class="line">    elif [ -f $fq1 -a -f $fq2 ]; then</span><br><span class="line">        echo `date`"$&#123;case&#125;: paired data. Do bowtie2 ..."</span><br><span class="line">        bowtie2 -p $nthread -x $index_base -U $fq1 $fq2 \</span><br><span class="line">            --rg-id "$&#123;case&#125;" \</span><br><span class="line">            --rg "SM:$&#123;case&#125;" \</span><br><span class="line">            --rg "LB:$&#123;seq_type&#125;" \</span><br><span class="line">            --rg "PL:Illumina" \</span><br><span class="line">            -S $path_align/$&#123;case&#125;.sam</span><br><span class="line">        echo `date`"$&#123;case&#125;: bowtie2 done!"</span><br><span class="line"></span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools view ..."</span><br><span class="line">        samtools view -@ $nthread \</span><br><span class="line">            -bhS -q 30 $path_align/$&#123;case&#125;.sam \</span><br><span class="line">            &gt; $path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools view done!"</span><br><span class="line"></span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools sort ..."</span><br><span class="line">        samtools sort -@ $nthread -m 1G \</span><br><span class="line">            --reference $path_genome_fa \</span><br><span class="line">            -o $path_align/$&#123;case&#125;.bam \</span><br><span class="line">            $path_align/$&#123;case&#125;_nonsort.bam \</span><br><span class="line">            &gt; $path_log/$&#123;case&#125;.align_bowtie2_hg38.log 2&gt;&amp;1</span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools sort done!"</span><br><span class="line">        </span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools index..."</span><br><span class="line">        samtools index -@ $nthread $path_align/$&#123;case&#125;.bam</span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools index done!"</span><br><span class="line">    </span><br><span class="line">    else</span><br><span class="line">        echo `date`"$&#123;case&#125;: unpaired data. Do bowtie2 ..."</span><br><span class="line">        bowtie2 -p $nthread -x $index_base -U $fq \</span><br><span class="line">            --rg-id "$&#123;case&#125;" \</span><br><span class="line">            --rg "SM:$&#123;case&#125;" \</span><br><span class="line">            --rg "LB:$&#123;seq_type&#125;" \</span><br><span class="line">            --rg "PL:Illumina" \</span><br><span class="line">            -S $path_align/$&#123;case&#125;.sam</span><br><span class="line">        echo `date`"$&#123;case&#125;: bowtie2 done!"</span><br><span class="line"></span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools view ..."</span><br><span class="line">        samtools view -@ $nthread \</span><br><span class="line">            -bhS -q 30 \</span><br><span class="line">            $path_align/$&#123;case&#125;.sam \</span><br><span class="line">            &gt;$path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools view done!"</span><br><span class="line"></span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools sort ..."</span><br><span class="line">        samtools sort \</span><br><span class="line">            -@ $nthread -m 1G \</span><br><span class="line">            --reference $path_genome_fa \</span><br><span class="line">            -o $path_align/$&#123;case&#125;.bam \</span><br><span class="line">            $path_align/$&#123;case&#125;_nonsort.bam \</span><br><span class="line">            &gt; $path_log/$&#123;case&#125;.align_bowtie2_hg38.log 2&gt;&amp;1</span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools sort done!"</span><br><span class="line">        </span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools index..."</span><br><span class="line">        samtools index -@ $nthread $path_align/$&#123;case&#125;.bam</span><br><span class="line">        echo `date`"$&#123;case&#125;: samtools index done!"</span><br><span class="line">    fi</span><br><span class="line">    </span><br><span class="line">    ## 删除多余的文件</span><br><span class="line">    if [ -f $path_align/*.sam ]; then</span><br><span class="line">    	rm $path_align/*.sam</span><br><span class="line">    elif [ -f $path_align/*_nonsort.bam ]; then </span><br><span class="line">    	rm $path_align/*_nonsort.bam</span><br><span class="line">    else</span><br><span class="line">    	echo "no sam. Ignored."</span><br><span class="line">    fi</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> End bowtie2_hg38</span></span><br></pre></td></tr></table></figure>


<h2 id="bowtie2-mm9"><a href="#bowtie2-mm9" class="headerlink" title="bowtie2_mm9"></a>bowtie2_mm9</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/bowtie2_mm9</span><br><span class="line">vim $nbin/bowtie2_mm9</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: bowtie2_mm9</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过bowtie2算法，根据mm9基因组对*.fastq.gz进行比对。通过samtools sort进行排序。最终生成*.bam文件。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">index_base=/data/reference/mouse/mm9/index/bowtie2/mm9 # 注意这不是文件路径而是文件前缀</span><br><span class="line">ws=$1</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_fastq=$&#123;ws&#125;/output/fastq</span><br><span class="line">path_log=$&#123;ws&#125;/log/bowtie2_mm9</span><br><span class="line">path_align=$&#123;ws&#125;/output/align</span><br><span class="line">path_genome_fa=/data/reference/mouse/mm9/genome/mm9.fa</span><br><span class="line">seq_type=Chip-Seq</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup bowtie2_mm9 <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/bowtie2_mm9/bowtie2_mm9.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/bowtie2_mm9/bowtie2_mm9.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># dir &amp; log</span></span></span><br><span class="line">mkdir -p $path_log $path_align</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle <span class="keyword">for</span> bwa mem</span></span><br><span class="line">cat $path_sra2fq_txt | while read id </span><br><span class="line">do</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    case=$&#123;arr[1]&#125;</span><br><span class="line">    fq1=$path_fastq/$&#123;case&#125;_1.fastq.gz</span><br><span class="line">    fq2=$path_fastq/$&#123;case&#125;_2.fastq.gz</span><br><span class="line">    fq=$path_fastq/$&#123;case&#125;.fastq.gz</span><br><span class="line">    </span><br><span class="line">    if [ -f $path_align/$&#123;case&#125;.bam ]; then</span><br><span class="line">    echo "$path_align/$&#123;case&#125;.bam had existed."</span><br><span class="line">    elif [ -f $fq1 -a -f $fq2 -a]; then</span><br><span class="line">    </span><br><span class="line">    echo `date`"$&#123;case&#125;: paired data. Do bowtie2 ..."</span><br><span class="line">    bowtie2 -p $nthread -x $index_base -U $fq1 $fq2 \</span><br><span class="line">    --rg-id "$&#123;case&#125;" \</span><br><span class="line">    --rg "SM:$&#123;case&#125;" \</span><br><span class="line">    --rg "LB:$&#123;seq_type&#125;" \</span><br><span class="line">    --rg "PL:Illumina" \</span><br><span class="line">    -S $path_align/$&#123;case&#125;.sam</span><br><span class="line">    echo `date`"$&#123;case&#125;: bowtie2 done!"</span><br><span class="line">    </span><br><span class="line">    echo `date`"$&#123;case&#125;: samtools view ..."</span><br><span class="line">    samtools view -@ $nthread -bhS -q 30 $path_align/$&#123;case&#125;.sam &gt;$path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">    echo `date`"$&#123;case&#125;: samtools view done!"</span><br><span class="line">    </span><br><span class="line">    echo `date`"$&#123;case&#125;: samtools sort ..."</span><br><span class="line">    samtools sort -@ $nthread -m 1G --reference $path_genome_fa -o $path_align/$&#123;case&#125;.bam $path_align/$&#123;case&#125;_nonsort.bam &gt;$path_log/$&#123;case&#125;.align_bowtie2_hg38.log 2&gt;&amp;1</span><br><span class="line">    echo `date`"$&#123;case&#125;: samtools sort done!"</span><br><span class="line">    </span><br><span class="line">    else</span><br><span class="line">    echo `date`"$&#123;case&#125;: unpaired data. Do bowtie2 ..."</span><br><span class="line">    bowtie2 -p $nthread -x $index_base -U $fq \</span><br><span class="line">    --rg-id "$&#123;case&#125;" \</span><br><span class="line">    --rg "SM:$&#123;case&#125;" \</span><br><span class="line">    --rg "LB:$&#123;seq_type&#125;" \</span><br><span class="line">    --rg "PL:Illumina" \</span><br><span class="line">    -S $path_align/$&#123;case&#125;.sam</span><br><span class="line">    echo `date`"$&#123;case&#125;: bowtie2 done!"</span><br><span class="line">    </span><br><span class="line">    echo `date`"$&#123;case&#125;: samtools view ..."</span><br><span class="line">    samtools view -@ $nthread -bhS -q 30 $path_align/$&#123;case&#125;.sam &gt;$path_align/$&#123;case&#125;_nonsort.bam</span><br><span class="line">    echo `date`"$&#123;case&#125;: samtools view done!"</span><br><span class="line">    </span><br><span class="line">    echo `date`"$&#123;case&#125;: samtools sort ..."</span><br><span class="line">    samtools sort -@ $nthread -m 1G --reference $path_genome_fa -o $path_align/$&#123;case&#125;.bam $path_align/$&#123;case&#125;_nonsort.bam &gt;$path_log/$&#123;case&#125;.align_bowtie2_hg38.log 2&gt;&amp;1</span><br><span class="line">    echo `date`"$&#123;case&#125;: samtools sort done!"</span><br><span class="line">    fi</span><br><span class="line"></span><br><span class="line">    ## 删除多余的文件</span><br><span class="line">    rm $path_align/*.sam $path_align/*_nonsort.bam</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> End bowtie2_mm9</span></span><br></pre></td></tr></table></figure>




<h2 id="bam-res"><a href="#bam-res" class="headerlink" title="bam_res"></a>bam_res</h2><blockquote>
<p>对.bam结果进行统计计算和质量控制，最后将结果综合于multiqc中展示</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/bam_res</span><br><span class="line">vim $nbin/bam_res</span><br><span class="line">rm $nbin/bam_res; cp2 $nbin</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: bam_res</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Descritiopn</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 对.bam结果进行统计计算和质量控制。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1 #work space</span><br><span class="line"><span class="meta">#</span><span class="bash"> path_genome_fa=/data/reference/genome/hg38/hg38.fa</span></span><br><span class="line">path_genome_fa=/data/reference/mouse/mm10/genome/mm10.fa</span><br><span class="line">path_bam=$&#123;ws&#125;/output/align</span><br><span class="line">path_stat=$&#123;ws&#125;/output/stat;</span><br><span class="line">path_bamstats=$&#123;ws&#125;/output/plot-bamstats</span><br><span class="line">path_log=$&#123;ws&#125;/log/bam_res</span><br><span class="line">project=SCS_2</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup bam_res <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/bam_res.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/output/align/<span class="built_in">stat</span>/*.<span class="built_in">stat</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> dir</span></span><br><span class="line">mkdir -p $path_stat $path_log $path_bamstats</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle: </span></span><br><span class="line">ls $path_bam | grep 'bam$' |while read id</span><br><span class="line">do</span><br><span class="line">case=$(basename $&#123;id&#125; .bam)</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> samtools stats</span></span><br><span class="line">echo `date`" $&#123;case&#125;: Process samtools stats..."</span><br><span class="line">samtools stats -@ $nthread --reference $path_genome_fa $path_bam/$&#123;id&#125; &gt; $path_stat/$&#123;case&#125;.stat</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> plot_bamstats</span></span><br><span class="line">echo `date`" $&#123;case&#125;: Process plot_bamstats..."</span><br><span class="line">id2=$&#123;case&#125;.stat</span><br><span class="line">mkdir -p $path_bamstats/$&#123;case&#125;</span><br><span class="line">plot-bamstats -p $path_bamstats/$&#123;case&#125;/$&#123;case&#125; $path_stat/$&#123;id2&#125; &gt; $path_log/plot-bamstats_$&#123;case&#125;.log</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># multiqc</span></span></span><br><span class="line">echo `date`" Process multiqc..."</span><br><span class="line">mkdir -p $path_stat/multiqc</span><br><span class="line">nohup multiqc $path_stat -n bamQC_$&#123;project&#125; -p -i " Basic bam QC of $&#123;project&#125;" -o $path_stat/multiqc &gt; $path_log/multiqc_samtools-stats.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line">mkdir -p $path_bamstats/multiqc</span><br><span class="line">nohup multiqc $path_bamstats -n plot-bamstats_$&#123;project&#125; -p -i " Bam QC of $&#123;project&#125; by bamstats" -o $path_bamstats/multiqc &gt; $path_log/multiqc_plot-bamstats.log 2&gt;&amp;1 &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End bam_res</span></span></span><br><span class="line">echo `date`" All done!"</span><br></pre></td></tr></table></figure>



<h2 id="bam-qualimap"><a href="#bam-qualimap" class="headerlink" title="bam_qualimap"></a>bam_qualimap</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/bam_qualimap</span><br><span class="line">vim $nbin/bam_qualimap</span><br><span class="line">rm $nbin/bam_qualimap; cp2 $nbin</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function bam_qualimap</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 对bam文件（比如bqsr.bam）进行统计</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup bam_qualimap <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/bam_qualimap.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/bam_qualimap.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">exon_bed=/data/reference/CCDS/human/exon_probe.GRCh38.gene.bed</span><br><span class="line">path_bam=$ws/output/align</span><br><span class="line">path_res=$ws/output/stat/qualimap</span><br><span class="line">path_log=$ws/log/qualimap</span><br><span class="line"><span class="meta">#</span><span class="bash">bam_suffix=<span class="string">"_bqsr.bam"</span> <span class="comment">#按需更改</span></span></span><br><span class="line">bam_suffix=".bam" #按需更改</span><br><span class="line">java_mem=2G #按需更改。一般用不了很多内存</span><br><span class="line">project=wes_2</span><br><span class="line">mkdir -p $path_res/multiqc $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">ls $path_bam | grep $bam_suffix | while read id</span><br><span class="line">do </span><br><span class="line">sample=$(basename $&#123;id&#125; ".bam")</span><br><span class="line">if [ ! -d $path_res/$&#123;sample&#125;_stat ]; then</span><br><span class="line">    mkdir -p $path_res/$&#123;sample&#125;_stat</span><br><span class="line">    echo `date`" $sample: start qualimap..."</span><br><span class="line">    nohup qualimap bamqc --java-mem-size=$java_mem \</span><br><span class="line">        -gff $exon_bed \</span><br><span class="line">        -bam $path_bam/$id \</span><br><span class="line">        -outdir $path_res/$&#123;sample&#125;_stat \</span><br><span class="line">        &gt; $path_log/qualimap_$&#123;sample&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line">    echo `date`" $sample: qualimap done!"</span><br><span class="line"> fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$path_log</span>/qualimap*</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># multiqc</span></span></span><br><span class="line">multiqc $path_res -n qc-$&#123;project&#125; -p -i " QC REPORT OF $&#123;project&#125;" -o $path_res/multiqc</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End function: bam_qualimap</span></span></span><br><span class="line">echo "All done!"</span><br></pre></td></tr></table></figure>



<h2 id="samtools-index"><a href="#samtools-index" class="headerlink" title="samtools_index"></a>samtools_index</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">vim ~/bin/example/samtools_index</span><br><span class="line">vim $nbin/samtools_index</span><br><span class="line">rm $nbin/samtools_index; cp2 $nbin</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: samtools_index</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 为bam文件制作index</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Note:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 如果将samtools index挂起，似乎速度并不会加快多少。从CPU的占用和Mem的占用来看，资源的利用率也比较低。而且硬盘读取占用很频繁，导致整个系统卡顿。如果有其它任务，不建议使用nohup将samtools index挂起。</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 经测，不挂起nohup后，其运行速度将更加的快。10+G的样本只要30s。挂起会让内存爆满</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_bam=$&#123;ws&#125;/output/align</span><br><span class="line">path_log=$&#123;ws&#125;/log/samtools_index</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Example</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup samtools_index <span class="variable">$ws</span> &amp;</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="built_in">log</span> dir</span></span><br><span class="line">mkdir -p $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle</span></span><br><span class="line">ls $path_bam | grep -E ".bam$" | while read id</span><br><span class="line">do</span><br><span class="line">case=$(basename $id ".bam")</span><br><span class="line">echo `date`" $&#123;case&#125;: Start samtools index"</span><br><span class="line">samtools index -@ $nthread $path_bam/$id &gt; $path_log/$&#123;id&#125;_samtools_index.log 2&gt;&amp;1</span><br><span class="line">echo `date`" $&#123;case&#125;: Done samtools index!"</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> End samtools_index</span></span><br><span class="line">echo `date`" All done!"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 调试</span></span></span><br><span class="line">time samtools index -@ 48 DCIS_P2_Exon_marked.bam &amp;</span><br></pre></td></tr></table></figure>

<h2 id="bam-merge"><a href="#bam-merge" class="headerlink" title="bam_merge"></a>bam_merge</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/bam_merge</span><br><span class="line">vim $nbin/bam_merge</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: bam_merge</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> merge .bam via specified rule</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup bam_merge <span class="variable">$ws</span> &amp;</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_sra2fq_txt=$&#123;ws&#125;/input/case/sra2case.txt</span><br><span class="line">path_bam=$&#123;ws&#125;/output/align</span><br><span class="line">path_log=$&#123;ws&#125;/log/bam_merge</span><br><span class="line">nsample=4 # 第几列是用于merge的sample id</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># log</span></span></span><br><span class="line">mkdir -p $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe:</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> create uniuqe sample id </span></span><br><span class="line">cat $path_sra2fq_txt | cut -d' ' -f $nsample - | uniq - &gt; $path_bam/temp_unique.sample</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> merge bam via one sample id</span></span><br><span class="line">cat temp_unique.sample | while read i</span><br><span class="line">do </span><br><span class="line"><span class="meta">  #</span><span class="bash"> create merge bam path</span></span><br><span class="line"><span class="meta">  #</span><span class="bash"> i=GSM850467</span></span><br><span class="line">  touch $path_bam/temp_bamPath</span><br><span class="line">  cat $&#123;path_sra2fq_txt&#125; |grep $i| while read id; </span><br><span class="line">  do </span><br><span class="line">  arr=($&#123;id&#125;)</span><br><span class="line">  bam=$path_bam/$&#123;arr[1]&#125;.rmdup.bam # 按实际情况修改</span><br><span class="line">  echo "$bam" &gt;&gt; $path_bam/temp_bamPath</span><br><span class="line">  echo "Input: the bam files of $i-$bam"</span><br><span class="line">  done</span><br><span class="line"></span><br><span class="line"><span class="meta">  #</span><span class="bash"> merge bam via <span class="string">"temp_bamPath"</span></span></span><br><span class="line">  echo "Merge .bam of $i..."</span><br><span class="line">  cat $path_bam/temp_bamPath | xargs samtools merge -@ $nthread $path_bam/$&#123;i&#125;.merge.bam </span><br><span class="line">  echo "Complete merge .bam of $i! Output: $path_bam/$&#123;i&#125;.merge.bam"</span><br><span class="line">  </span><br><span class="line"><span class="meta">  #</span><span class="bash"> create index <span class="keyword">for</span> merge.bam</span></span><br><span class="line">  echo "samtools index for new merge bam..."</span><br><span class="line">  nohup samtools index -@ $nthread $path_bam/$&#123;i&#125;.merge.bam &gt; $path_log/$&#123;i&#125;.merge_samtool_index.log 2&gt;&amp;1 &amp;</span><br><span class="line">  </span><br><span class="line"><span class="meta">  #</span><span class="bash"> rm temp files</span></span><br><span class="line">  rm $path_bam/temp_bamPath</span><br><span class="line">done  </span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> rm temp files</span></span><br><span class="line">rm $path_bam/temp_unique.sample</span><br><span class="line">echo "Done!"</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End bam_merge</span></span></span><br></pre></td></tr></table></figure>



<h2 id="sm-spar"><a href="#sm-spar" class="headerlink" title="sm_spar"></a>sm_spar</h2><blockquote>
<p>fast way to run SPAR pipeline for small RNA sequencing data</p>
<p>the config file needed prepare first like <code>${ws}/input/config.hg38.sh</code></p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/sm_spar</span><br><span class="line">rm $nbin/sm_spar; cp2 $nbin</span><br><span class="line">vim $nbin/sm_spar</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function sm_spar</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> fast way to run SPAR pipeline <span class="keyword">for</span> small RNA sequencing data</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> input <span class="built_in">type</span> is bam format</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup sm_spar <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/sm_spar.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/sm_spar.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">echo "'STAR;hisat2'" | xargs -i fast_extra &#123;&#125; ";" | while read type; do</span><br><span class="line"></span><br><span class="line"><span class="meta">	#</span><span class="bash"><span class="comment"># dir management</span></span></span><br><span class="line">    path_align=$&#123;ws&#125;/output/align/$&#123;type&#125;</span><br><span class="line">    path_config=$&#123;ws&#125;/input/config.hg38.sh</span><br><span class="line">    path_res=$ws/output/SPAR/$&#123;type&#125;_bam</span><br><span class="line">    path_log=$ws/log/SPAR_$&#123;type&#125;</span><br><span class="line">    mkdir -p $path_res $path_log;</span><br><span class="line">    </span><br><span class="line">    ## SPAR</span><br><span class="line">    ls $&#123;path_align&#125; | grep -E 'bam$' | while read id; do </span><br><span class="line">        base=$(basename $id .bam) </span><br><span class="line">        if [ ! -d $&#123;path_res&#125;/$&#123;base&#125; ]; then</span><br><span class="line">            mkdir -p $&#123;path_res&#125;/$&#123;base&#125;</span><br><span class="line">            nohup SPAR.sh $&#123;path_align&#125;/$&#123;id&#125; \</span><br><span class="line">                $&#123;path_res&#125;/$&#123;base&#125; \</span><br><span class="line">                $&#123;path_config&#125; \</span><br><span class="line">                $&#123;nthread&#125; \</span><br><span class="line">                &gt; $&#123;path_log&#125;/SPAR_$&#123;type&#125;_$&#123;base&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line">            #tail -f $&#123;path_log&#125;/SPAR_$&#123;type&#125;_$&#123;base&#125;.log</span><br><span class="line">        else</span><br><span class="line">            echo "$&#123;path_res&#125;/$&#123;base&#125;: SPAR result exists. Ignored!"</span><br><span class="line">        fi</span><br><span class="line">    done</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End function</span></span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">###================================Config=============================####</span></span></span><br><span class="line"></span><br><span class="line">path_res=$ws/output/SPAR</span><br><span class="line">mkdir -p $path_res; cd $path_res</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Config for hg38</span></span></span><br><span class="line">vim $ws/input/config.hg38.sh</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># parameters</span></span></span><br><span class="line">ws=/home/huangwb8/Test/smallRNA-Seq_2</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># SPAR config file</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="built_in">export</span> HOMEDIR=<span class="string">"<span class="variable">$&#123;ws&#125;</span>"</span> </span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># absolute path to the bin directory</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="built_in">export</span> BINDIR=<span class="string">"usr/bin/"</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># (default) absolute path to the default SPAR output directory</span></span></span><br><span class="line">export SPARDIR="$ws/output/SPAR"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># genome build</span></span></span><br><span class="line">export GENOMEBUILD=hg38</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># data directory for SPAR (DASHR/ENCODE data, etc)</span></span></span><br><span class="line">export SPARDATAPATH="/data/spar/spar_pipeline"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># absolute path to the STAR genome index</span></span></span><br><span class="line">export STAR="/home/huangwb8/Downloads/conda/location/envs/bioinf/bin/STAR" # STAR</span><br><span class="line">export genomeDir="/data/reference/index/star/hg38"  # STAR genome index</span><br><span class="line">export GENOMEFA="/data/reference/genome/hg38/hg38.fa"</span><br><span class="line">export CONSERVATIONTRACK="/data/spar/ConservationTracks/hg38.phastCons100way.bw"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># absolute path to pre-installed STAR, samtools, AWK, etc</span></span></span><br><span class="line">export SAMTOOLS="/home/huangwb8/Downloads/conda/location/bin/samtools" # SAMTOOLS</span><br><span class="line">export BEDTOOLS="/home/huangwb8/Downloads/conda/location/bin/bedtools"</span><br><span class="line">export GAWK="/usr/bin/gawk"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># UCSC tools</span></span></span><br><span class="line">ucsc=/home/huangwb8/Downloads/UCSC-Tools</span><br><span class="line">export BGTOBIGWIG="$&#123;ucsc&#125;/bedGraphToBigWig"</span><br><span class="line">export BEDTOBIGBED="$&#123;ucsc&#125;/bedToBigBed"</span><br><span class="line">export BIGWIGTOBEDGRAPH="$&#123;ucsc&#125;/bigWigToBedGraph"</span><br><span class="line">export BIGWIGAVGOVERBED="$&#123;ucsc&#125;/bigWigAverageOverBed"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Adapter trimming</span></span></span><br><span class="line">export CUTADAPT="/home/huangwb8/.local/bin/cutadapt"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># mapping parameters for STAR</span></span></span><br><span class="line">export maxMismatchCnt=0 # maximum number of genomic mismatches</span><br><span class="line">export maxMapCnt=100 # maximum number of places a read can map to</span><br><span class="line">export minMappedLength=14 # minimum *mapped* length</span><br><span class="line">export maxReadLength=44 # maximum read length</span><br><span class="line">export minReadLength=14 # minimum read length</span><br><span class="line">export minCoverage=10 # minimum read coverage</span><br><span class="line">export minFoldChange=2 # minimum fold change (read coverage(i)/read_coverage(i-1)) </span><br><span class="line">                       # for peak detection</span><br><span class="line">export max5pClip=1 # maximum allowed 5' end clipping ('S' in the CIGAR string) </span><br><span class="line"></span><br><span class="line"></span><br><span class="line">export RSCRIPT="/usr/bin/Rscript"</span><br><span class="line"></span><br><span class="line">function printT</span><br><span class="line">&#123;</span><br><span class="line">  echo "`date +'%b %d %H:%M:%S'` ..... $1"</span><br><span class="line">&#125;</span><br><span class="line">export -f printT</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End config</span></span></span><br></pre></td></tr></table></figure>





<h2 id="count-htseq"><a href="#count-htseq" class="headerlink" title="count_htseq"></a>count_htseq</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/count_htseq</span><br><span class="line">vim $nbin/count_htseq</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: count_htseq</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过htseq-count对.bam文件计算<span class="built_in">read</span> counts，并输出表达矩阵。此法十分消耗内存和时间，不推荐使用。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup count_htseq <span class="variable">$ws</span> &gt; htseq-count.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f *count</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameter</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_align=$&#123;ws&#125;/output/align</span><br><span class="line">path_count=$&#123;ws&#125;/output/count_htseq</span><br><span class="line">path_genome_gtf=/data/reference/gtf/hg38_gencode.v32.annotation.gtf.gz</span><br><span class="line">mkdir -p $path_count</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">ls $path_align | grep '.bam$' | while read id</span><br><span class="line"><span class="meta">#</span><span class="bash">id=N052611-Albuterol.bam</span></span><br><span class="line">do</span><br><span class="line">  case=$(basename $&#123;id&#125; ".bam")</span><br><span class="line">  if [ ! -f $path_count/$&#123;case&#125;.count ]; then </span><br><span class="line">  echo `date`"$&#123;case&#125;: do htseq-count ..."</span><br><span class="line">  time htseq-count -r pos -f bam $path_align/$&#123;id&#125; $path_genome_gtf &gt; $path_count/$&#123;case&#125;.count</span><br><span class="line">  echo `date`"$&#123;case&#125;: done htseq-count!"</span><br><span class="line">  fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End count_htseq</span></span></span><br></pre></td></tr></table></figure>

<h2 id="count-feature"><a href="#count-feature" class="headerlink" title="count_feature"></a>count_feature</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/count_feature</span><br><span class="line">vim $nbin/count_feature; cp2 $nbin</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: count_feature</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过featureCounts对RNA-Seq实验的*.bam文件计算<span class="built_in">read</span> counts，并输出表达矩阵。此法相对较快，推荐使用。一般人的就10来分钟。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup count_feature <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/featureCounts.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/featureCounts.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameter</span></span></span><br><span class="line">ws=$1</span><br><span class="line">nbin=$ws/bin</span><br><span class="line">path_align=$&#123;ws&#125;/output/align</span><br><span class="line">path_log=$&#123;ws&#125;/log/count_feature</span><br><span class="line">path_count=$&#123;ws&#125;/output/count_feature</span><br><span class="line">path_genome_gtf=/data/reference/gtf/hg38_gencode.v32.annotation.gtf.gz</span><br><span class="line"><span class="meta">#</span><span class="bash">path_genome_gtf=/data/reference/gtf/hg38_miRBase_v22_20180305.gff3 <span class="comment"># hg38_miRNA; -t miRNA -g Name</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">path_genome_gtf=/data/reference/gtf/mm10_gencode.vM23.annotation.gtf.gz</span></span><br><span class="line">nthread=48</span><br><span class="line">mkdir -p $path_count $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">ls $path_align | grep '.bam$' | while read id</span><br><span class="line"><span class="meta">#</span><span class="bash">id=N052611-Albuterol.bam</span></span><br><span class="line">do</span><br><span class="line">  case=$(basename $&#123;id&#125; ".bam")</span><br><span class="line">  if [ ! -f $path_count/$&#123;case&#125;.count ]; then   </span><br><span class="line"><span class="meta">  #</span><span class="bash"> -p If specified, fragments (or templates) will be counted instead of reads. This option is only applicable <span class="keyword">for</span> paired-end reads.</span></span><br><span class="line"><span class="meta">  #</span><span class="bash"> exon保证了<span class="built_in">read</span>是比对到外显子中的，这些<span class="built_in">read</span>对于RNA-Seq而言才是有意义的。</span></span><br><span class="line">  echo `date`"$&#123;case&#125;: do count_feature ..."  </span><br><span class="line">  nohup featureCounts -T $nthread -p \</span><br><span class="line">      -a $path_genome_gtf \</span><br><span class="line">      --tmpDir /data/ \</span><br><span class="line">      --verbose \</span><br><span class="line">      -t exon -g gene_id  \</span><br><span class="line">      -o $path_count/$&#123;case&#125;.count \</span><br><span class="line">      $path_align/$&#123;id&#125; \</span><br><span class="line">      &gt; $path_log/featureCounts_$&#123;case&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line">  echo `date`"$&#123;case&#125;: done count_feature!"</span><br><span class="line">  fi</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># make count matrix</span></span></span><br><span class="line">echo "path_count='$path_count'" &gt; $path_count/mergeFeatureCounts.R</span><br><span class="line">cat $nbin/mergeFeatureCounts.R &gt;&gt; $path_count/mergeFeatureCounts.R</span><br><span class="line">Rscript $path_count/mergeFeatureCounts.R &gt; $path_log/mergeFeatureCounts.log 2&gt;&amp;1  &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End count_feature</span></span></span><br></pre></td></tr></table></figure>

<h2 id="fpkm-cufflinks"><a href="#fpkm-cufflinks" class="headerlink" title="fpkm_cufflinks"></a>fpkm_cufflinks</h2><blockquote>
<p>Cufflinks的使用: <a href="https://blog.csdn.net/hugolee123/article/details/44814469" target="_blank" rel="noopener">https://blog.csdn.net/hugolee123/article/details/44814469</a></p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/fpkm_cufflinks</span><br><span class="line">vim $nbin/fpkm_cufflinks</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: fpkm_cufflinks</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过cufflinks对RNA-Seq实验的*.bam文件计算fpkm值。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup fpkm_cufflinks <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/fpkm_cufflinks.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/fpkm_cufflinks.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameter</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_align=$&#123;ws&#125;/output/align</span><br><span class="line">path_log=$&#123;ws&#125;/log/fpkm_cufflinks</span><br><span class="line">path_res=$&#123;ws&#125;/output/fpkm_cufflinks</span><br><span class="line">path_genome_gtf=/data/reference/gtf/raw/hg38_gencode.v32.annotation.gtf</span><br><span class="line">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span><br><span class="line">nthread=48</span><br><span class="line">mkdir -p $path_res $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">ls $path_align | grep '.bam$' | while read id</span><br><span class="line"><span class="meta">#</span><span class="bash">id=N052611-Albuterol.bam</span></span><br><span class="line">do</span><br><span class="line">  case=$(basename $&#123;id&#125; ".bam")</span><br><span class="line">  if [ ! -f $path_res/$&#123;case&#125;/transcripts.gtf ]; then   </span><br><span class="line">  mkdir -p $path_res/$&#123;case&#125;</span><br><span class="line">  echo `date`" $&#123;case&#125;: do fpkm_cufflinks ..."  </span><br><span class="line">  cufflinks -p $nthread \</span><br><span class="line">  -G $path_genome_gtf \</span><br><span class="line">  -o $path_res/$&#123;case&#125; \</span><br><span class="line"><span class="meta">  $</span><span class="bash">path_align/<span class="variable">$id</span> \</span></span><br><span class="line"><span class="meta">  &gt;</span><span class="bash"> <span class="variable">$path_log</span>/fpkm_cufflinks_<span class="variable">$&#123;case&#125;</span>.<span class="built_in">log</span> 2&gt;&amp;1 </span></span><br><span class="line">  echo `date`" $&#123;case&#125;: done fpkm_cufflinks!"</span><br><span class="line">  else </span><br><span class="line">  echo `date`" FPKM data of $&#123;case&#125; existed.Ignore!"</span><br><span class="line">  fi</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># merge gtf</span></span></span><br><span class="line">echo "cuffmerge: merge multiple gtf files of every sample..."</span><br><span class="line">ls $path_res | while read i; do echo "$&#123;path_res&#125;/$&#123;i&#125;/transcripts.gtf" &gt;&gt; $&#123;path_res&#125;/temp.fpkm_cufflink.list; done</span><br><span class="line">cuffmerge -p $nthread \</span><br><span class="line">    -s $path_genome_fa \</span><br><span class="line">    -g $path_genome_gtf \</span><br><span class="line">    -o $path_res \</span><br><span class="line">    temp.fpkm_cufflink.list \</span><br><span class="line">    &gt; $path_log/cuffmerge.log 2&gt;&amp;1</span><br><span class="line">echo "cuffmerge: done!"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># make FPKM matrix</span></span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End fpkm_cufflinks</span></span></span><br><span class="line">echo "All done! "</span><br></pre></td></tr></table></figure>



<h2 id="bam2rpkm"><a href="#bam2rpkm" class="headerlink" title="bam2rpkm"></a>bam2rpkm</h2><blockquote>
<p><a href="https://mp.weixin.qq.com/s/SF9HRGaBKIchYTHbEI6Nog" target="_blank" rel="noopener">根据比对后的bam文件来计算外显子的RPKM</a></p>
</blockquote>
<p>假设gene的exon一共有<strong>n</strong>个。此bam文件中一共mapped到<strong>a</strong>百万个reads. 第i个外显子的mapped reads的个数为<strong>e<sub>i</sub></strong>，外显子的长度为<strong>L<sub>i</sub></strong>(kb)，对应的rpkm值为<strong>r<sub>i</sub></strong>，则该exon的RPKM值为：<br>$$<br>r_i = \frac{e_i} {a \times L_i}<br>$$<br>因而有：<br>$$<br>e_i=a \times r_i \times L_i<br>$$<br>同理，对于gene（拥有多个exon）则有：<br>$$<br>RPKM_{gene} = \frac{\sum^{n}<em>{i=1} e_i} {a \times \sum^{n}</em>{i=1}L_i}<br>    = \frac{a \times  \sum^{n}<em>{i=1} (r_i \times L_i)} {a \times \sum^{n}</em>{i=1}L_i}<br>    = \frac{\sum^{n}<em>{i=1} (r_i \times L_i)} {\sum^{n}</em>{i=1}L_i}<br>$$<br>因此，只要有每个外显子的RPKM值<strong>r<sub>i</sub></strong>和该外显子的长度<strong>L<sub>i</sub></strong>，即可求整个基因的RPKM值。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/bam2rpkm</span><br><span class="line">vim $nbin/bam2rpkm</span><br><span class="line">rm $nbin/&#123;bam2rpkm,mergeRPKM.R&#125;; cp2 $nbin</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function bam2rpkm</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 将bam文件生成rpkm值</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage：</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup bam2rpkm <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/bam2rpkm.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/bam2rpkm.log </span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters:</span></span></span><br><span class="line">ws=$1</span><br><span class="line">nbin=$&#123;ws&#125;/bin</span><br><span class="line">path_align=$&#123;ws&#125;/output/align</span><br><span class="line">path_log=$&#123;ws&#125;/log/bam2rpkm</span><br><span class="line">path_res=$&#123;ws&#125;/output/bam2rpkm</span><br><span class="line">path_genome_gtf=/data/reference/gtf/mm10_gencode.vM23.annotation.gtf.gz</span><br><span class="line">path_genome_fa=/data/reference/mouse/mm10/genome/mm10.fa</span><br><span class="line">bed=/data/reference/CCDS/mouse/mm10.exon_sub1.bed</span><br><span class="line"><span class="meta">#</span><span class="bash">path_genome_gtf=/data/reference/gtf/raw/hg38_gencode.v32.annotation.gtf</span></span><br><span class="line"><span class="meta">#</span><span class="bash">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span></span><br><span class="line"><span class="meta">#</span><span class="bash">bed=</span></span><br><span class="line">mkdir -p $path_res $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Get RPKM value <span class="keyword">for</span> every sample</span></span><br><span class="line">for bam in $path_align/*.bam; do</span><br><span class="line"><span class="meta">	#</span><span class="bash">bam=<span class="variable">$path_align</span>/Mammary_tumor_fibroblast-GSM3026607.bam</span></span><br><span class="line">    file=$(basename $bam )</span><br><span class="line">    sample=$&#123;file%%.*&#125;</span><br><span class="line">    echo $sample</span><br><span class="line">    export total_reads=$(samtools idxstats  $bam|awk -F '\t' '&#123;s+=$3&#125;END&#123;print s&#125;')</span><br><span class="line">    echo The number of reads is $total_reads</span><br><span class="line">    # 这个脚本是为了弥补conifer这个软件的bug，所以按照了conifer的output格式来输出的，输出3列，第一列就是纯粹的递增行数，对应于exon_probe.hg38.gene.bed 文件的每一行，第二列是该对应的外显子上面的覆盖的reads，第三列就是算出的rpkm值啦，这里并没有输出该外显子长度。</span><br><span class="line">    bedtools multicov \</span><br><span class="line">    -bams $bam \</span><br><span class="line">    -bed $bed \</span><br><span class="line">    | perl -alne '&#123;$len=$F[2]-$F[1];if($len &lt;1 )&#123;print "$.\t$F[4]\t0" &#125;else&#123;$rpkm=(1000000000*$F[4]/($len* $ENV&#123;total_reads&#125;));print "$.\t$F[4]\t$rpkm"&#125;&#125;' \</span><br><span class="line">    &gt; $&#123;path_res&#125;/$&#123;sample&#125;.rpkm.txt &amp;</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Merge indivitual data to a expression matrix </span></span></span><br><span class="line">echo "path_rpkm='$path_res'" &gt; $path_res/mergeRPKM.R</span><br><span class="line">echo "path_bed='$bed'" &gt;&gt; $path_res/mergeRPKM.R</span><br><span class="line">cat $nbin/mergeRPKM.R &gt;&gt; $path_res/mergeRPKM.R</span><br><span class="line">Rscript $path_res/mergeRPKM.R &amp;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End bam2rpkm</span></span></span><br></pre></td></tr></table></figure>









<h2 id="gatk-checkDepth"><a href="#gatk-checkDepth" class="headerlink" title="gatk_checkDepth"></a>gatk_checkDepth</h2><blockquote>
<p>根据hg38的信息来检测测序深度</p>
<ul>
<li><p>-R需要输入你的referencegenome</p>
</li>
<li><p>bam.list为待分析的样本，可将多个bam放在一个bam.list里</p>
</li>
<li><p>-L 需要提供捕获的区域，为bed格式，如果是agilent的全外显子可在其官方网站下载</p>
</li>
<li><p>-ct1代表至少覆盖1X的区域，</p>
</li>
<li><p>-ct10代表至少覆盖10X的区域，你可以自己添加自己想要的深度</p>
</li>
</ul>
<p>参考文献</p>
<ul>
<li><a href="http://mp.weixin.qq.com/s?__biz=MzAxMDkxODM1Ng==&mid=2247483807&idx=1&sn=f092fa39003f6e66a501dc4e42ec0fa2&chksm=9b484124ac3fc832dc496b623b64b95aef88758dd9036401dc4c4aa74d40ceff8426a2c5907f&scene=21#wechat_redirect" target="_blank" rel="noopener"><strong>直播我的基因组20：覆盖度详细探究</strong></a></li>
<li><a href="http://www.biotrainee.com/thread-1376-1-1.html" target="_blank" rel="noopener"><strong>直播我的基因组分析-目录-持续更新</strong></a></li>
</ul>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"></span><br><span class="line">path_genome_fa=/data/reference/genome/hg38/hg38.fa</span><br><span class="line">gatk DepthOfCoverage -R $path_genome_fa -o result_name -I bam.list -L target.bed --omitDepthOutputAtEachBase --omitIntervalStatistics -ct1 -ct 10 -ct 20</span><br></pre></td></tr></table></figure>

<h2 id="gatk-internal-list"><a href="#gatk-internal-list" class="headerlink" title="gatk: internal list"></a>gatk: internal list</h2><blockquote>
<p><a href="https://gatkforums.broadinstitute.org/gatk/discussion/comment/46329#Comment_46329" target="_blank" rel="noopener">gatk4: Intervals and interval lists</a></p>
<p>有关bed的使用情况：We also accept the widely-used BED format, where intervals are in the form <chr> <start> <stop>, with fields separated by tabs. However, you should be aware that this file format is 0-based for the start coordinates, so coordinates taken from 1-based formats (e.g. if you’re cooking up a custom interval list derived from a file in a 1-based format) should be offset by 1. The GATK engine recognizes the .bed extension and interprets the coordinate system accordingly<br> 以CCDS数据库为例，首先需要先下载个最新的CCDS文件 (CCDS.20180614.txt) <a href="ftp://ftp.ncbi.nlm.nih.gov/pub/CCDS/current_human" target="_blank" rel="noopener">ftp://ftp.ncbi.nlm.nih.gov/pub/CCDS/current_human</a><br> 然后提取其外显子的起始/终止碱基位置，然后以150bp测序话，可以再加减150bp区间，由于bed格式是以0-based for the start coordinates，所以还要再减去1，最终做成一个bed格式文件，bed文件需要有3列信息chr，start位置和stop位置<br> <a href="http://www.bioinfo-scrounger.com/archives/642" target="_blank" rel="noopener">GATK 4.0 全外显子call variant</a>有关于bed的讨论<br><a href="https://mp.weixin.qq.com/s/JwxmheyRWH5aW4gvJU4vNQ" target="_blank" rel="noopener">一起来了解一下基因区间RangeData</a></p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 制作1-base bed</span></span></span><br><span class="line">wget ftp://ftp.ncbi.nlm.nih.gov/pub/CCDS//archive/21/CCDS.20161208.txt</span><br><span class="line">cd /data/reference/CCDS/human</span><br><span class="line">cat CCDS.current.txt | perl -alne '&#123;/\[(.*?)\]/;next unless $1;$gene=$F[2];$exons=$1;$exons=~s/\s//g;$exons=~s/-/\t/g;print "$F[0]\t$_\t$gene" foreach split/,/,$exons;&#125;'| sort -u | bedtools sort -i | awk '&#123;print "chr"$0"\t0\t+"&#125;'&gt; hg38.exon.bed</span><br></pre></td></tr></table></figure>

<p>然后，通过R来制作0-base bed</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 制作gatk interval lists(*.bed)</span></span><br><span class="line">setwd(<span class="string">"/data/reference/CCDS/human/"</span>)</span><br><span class="line">a &lt;- read.table(<span class="string">"./hg38.exon.bed"</span>,header = <span class="literal">F</span>,check.names = <span class="literal">F</span>)</span><br><span class="line">a$V2 &lt;- a$V2 - <span class="number">1</span></span><br><span class="line">dir.create(<span class="string">"./interval_lists"</span>,showWarnings = <span class="literal">F</span>,recursive = <span class="literal">T</span>)</span><br><span class="line">write.table(a,<span class="string">"./interval_lists/hg38.exon.bed"</span>,col.names = <span class="literal">F</span>,row.names = <span class="literal">F</span>,quote = <span class="literal">F</span>)</span><br></pre></td></tr></table></figure>

<p>或者用picard工具来制作坐标文件</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 用picard制作外显子坐标文件</span></span></span><br><span class="line">conda activate wes</span><br><span class="line">conda install -y picard</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># 下载CCDS外显子坐标并处理</span></span></span><br><span class="line">cd /data/reference/CCDS/human</span><br><span class="line">cat CCDS.current.txt | perl -alne '&#123;/\[(.*?)\]/;next unless $1;$gene=$F[2];$exons=$1;$exons=~s/\s//g;$exons=~s/-/\t/g;print "$F[0]\t$_\t$gene" foreach split/,/,$exons;&#125;'| sort -u | bedtools sort -i | awk '&#123;print "chr"$0"\t0\t+"&#125;'&gt; hg38.chr.bed</span><br></pre></td></tr></table></figure>

<p>其中，hg38.chr.bed的内容如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">chr1    935771  935895  SAMD11  0       +</span><br><span class="line">chr1    939039  939128  SAMD11  0       +</span><br><span class="line">chr1    939274  939459  SAMD11  0       +</span><br><span class="line">chr1    941143  941305  SAMD11  0       +</span><br><span class="line">chr1    942135  942250  SAMD11  0       +</span><br><span class="line">chr1    942409  942487  SAMD11  0       +</span><br><span class="line">chr1    942558  943057  SAMD11  0       +</span><br><span class="line">chr1    943252  943376  SAMD11  0       +</span><br></pre></td></tr></table></figure>

<p>picard对输入的bed的格式 要求是：</p>
<blockquote>
<p>In each annotation track, the number of fields per line must be consistent throughout a data set. For additional information regarding BED files and the annotation field options, please see:<br><a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1" target="_blank" rel="noopener">http://genome.ucsc.edu/FAQ/FAQformat.html#format1</a>.</p>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">BED files contain sequence data displayed in a flexible format that includes nine optional fields, in addition to three required fields within the annotation tracks. The required fields of a BED file include:</span><br><span class="line">chrom - The name of the chromosome (e.g. chr20) or scaffold (e.g. scaffold10671) </span><br><span class="line">chromStart - The starting position of the feature in the chromosome or scaffold. The first base in a chromosome is numbered &quot;0&quot; </span><br><span class="line">chromEnd - The ending position of the feature in the chromosome or scaffold.  The chromEnd base is not included in the display of the feature. For example, the first 100 bases of a chromosome are defined as chromStart&#x3D;0, chromEnd&#x3D;100, and span the bases numbered 0-99.</span><br></pre></td></tr></table></figure>

<p>通过picard生成bed: </p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># picard工具来制作坐标文件</span></span></span><br><span class="line">picard_jar=/home/huangwb8/Downloads/conda/location/envs/wes/share/picard-2.21.1-0/picard.jar</span><br><span class="line">java -jar $picard_jar BedToIntervalList \</span><br><span class="line">I=hg38.chr.bed \</span><br><span class="line">O=hg38.interval_list \</span><br><span class="line">SD=/data/reference/index/gatk/hg38/Homo_sapiens_assembly38.dict</span><br><span class="line"></span><br><span class="line">picard BedToIntervalList \</span><br><span class="line">I=/data/reference/bed/human/hg38_refSeq.bed \</span><br><span class="line">O=/data/reference/bed/human/hg38_picard.interval_list \</span><br><span class="line">SD=/data/reference/index/gatk/hg38/Homo_sapiens_assembly38.dict</span><br></pre></td></tr></table></figure>

<h2 id="gatk-markDuplicates"><a href="#gatk-markDuplicates" class="headerlink" title="gatk_markDuplicates"></a>gatk_markDuplicates</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/gatk_markDuplicates</span><br><span class="line">vim $nbin/gatk_markDuplicates</span><br><span class="line">rm $nbin/gatk_markDuplicates; cp2 $nbin</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: gatk_markDuplicates</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 去除重复序列。常用于DNA测序双端数据。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters: </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="variable">$1</span>: the path of .bam</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="variable">$2</span>：the number of threads</span></span><br><span class="line">ws=$1</span><br><span class="line">path_bam=$&#123;ws&#125;/output/align</span><br><span class="line">path_log=$&#123;ws&#125;/log/gatk_markDuplicates</span><br><span class="line">GATK=/home/huangwb8/Downloads/conda/location/bin/gatk</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup gatk_markDuplicates <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/gatk_markDuplicates.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/gatk_markDuplicates.log</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/gatk_markDuplicates/*.<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> dir</span></span><br><span class="line">mkdir -p $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle</span></span><br><span class="line">ls $path_bam | grep "bam$" | sed "s/.bam//g" | while read id</span><br><span class="line">do</span><br><span class="line">    BAM=$path_bam/$&#123;id&#125;.bam</span><br><span class="line"></span><br><span class="line">    if [ ! -f $path_bam/$&#123;id&#125;_marked.bam ]</span><br><span class="line">    then</span><br><span class="line">    echo "start MarkDuplicates for $&#123;id&#125;" `date`</span><br><span class="line">    nohup $GATK --java-options "-Xmx60G -Djava.io.tmpdir=/data/tmp " \</span><br><span class="line">    MarkDuplicates \</span><br><span class="line">    	-I $&#123;BAM&#125; \</span><br><span class="line">    	--REMOVE_DUPLICATES=true \</span><br><span class="line">    	-O $path_bam/$&#123;id&#125;_marked.bam \</span><br><span class="line">    	-M $path_bam/$&#123;id&#125;.metrics \</span><br><span class="line">    	&gt; $path_log/$&#123;id&#125;_marked.log 2&gt;&amp;1 &amp;</span><br><span class="line">    echo "end MarkDuplicates for $&#123;id&#125;" `date`</span><br><span class="line">    #samtools index -@ $nthread -m 5G -b $path_bam/$&#123;id&#125;_marked.bam</span><br><span class="line">    fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End function</span></span></span><br></pre></td></tr></table></figure>

<h2 id="after-gatk-markDup"><a href="#after-gatk-markDup" class="headerlink" title="after_gatk_markDup"></a>after_gatk_markDup</h2><blockquote>
<p>用于比较去重复序列前和去重复序列后的read数的差异；</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/after_gatk_markDup</span><br><span class="line">vim $nbin/after_gatk_markDup</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function:after_gatk_markDup</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 用于比较去重复序列前和去重复序列后的<span class="built_in">read</span>数的差异；</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> <span class="built_in">cd</span> <span class="variable">$ws</span>/output/align;nohup after_gatk_markDup <span class="variable">$ws</span> &amp;</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_bam=$&#123;ws&#125;/output/align</span><br><span class="line">touch $path_bam/temp_rawRead $path_bam/temp_markRead $path_bam/colnames $path_bam/temp_case</span><br><span class="line">echo "cases rawReads markReads" &gt; $path_bam/after_gatk_markDup.report</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Program</span></span></span><br><span class="line">ls $path_bam | grep "_marked.bam$" | while read id</span><br><span class="line">do</span><br><span class="line">case=$(basename $&#123;id&#125; _marked.bam)</span><br><span class="line">echo "$case" &gt;&gt; $path_bam/temp_case</span><br><span class="line">rawbam=$&#123;case&#125;.bam</span><br><span class="line">samtools view -@ 48 $path_bam/$&#123;rawbam&#125; |wc -l &gt;&gt; $path_bam/temp_rawRead</span><br><span class="line">samtools view -@ 48 $path_bam/$&#123;id&#125; | wc -l &gt;&gt; $path_bam/temp_markRead</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># paste</span></span></span><br><span class="line">paste $path_bam/temp_case $path_bam/temp_rawRead $path_bam/temp_markRead &gt; $path_bam/temp_compare</span><br><span class="line">cat $path_bam/temp_compare &gt;&gt; $path_bam/after_gatk_markDup.report</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># rm temprature values</span></span></span><br><span class="line">rm -r $path_bam/temp_*</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End after_gatk_markDup</span></span></span><br></pre></td></tr></table></figure>

<h2 id="gatk-BQSR-hg38"><a href="#gatk-BQSR-hg38" class="headerlink" title="gatk_BQSR_hg38"></a>gatk_BQSR_hg38</h2><blockquote>
<p>利用GATK4, 对bam文件进行碱基质量重校正BQSR（Base Quality Score Recalibration）。</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/gatk_BQSR_hg38</span><br><span class="line">vim $nbin/gatk_BQSR_hg38</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function gatk_BQSR_hg38</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 利用GATK4, 对bam文件进行碱基质量重校正BQSR（Base Quality Score Recalibration）。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup gatk_BQSR_hg38 <span class="variable">$ws</span> &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/gatk_BQSR_hg38/*<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters:</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_sra2case_txt=$ws/input/case/sra2case.txt</span><br><span class="line">path_bam=$ws/output/align</span><br><span class="line">path_log=$ws/log/gatk_BQSR_hg38</span><br><span class="line">GATK=/home/huangwb8/Downloads/conda/location/bin/gatk #按需改动</span><br><span class="line">snp=/data/reference/index/gatk/hg38/dbsnp_146.hg38.vcf.gz</span><br><span class="line">indel=/data/reference/index/gatk/hg38/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz</span><br><span class="line">ref=/data/reference/index/gatk/hg38/Homo_sapiens_assembly38.fasta</span><br><span class="line">bs=_marked # bam suffix..请按需要更改</span><br><span class="line">nthread=48</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># log</span></span></span><br><span class="line">mkdir -p $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> create <span class="keyword">case</span>(选择第2行的<span class="keyword">case</span> id)</span></span><br><span class="line">cat $path_sra2case_txt | while read id;do cut -d' ' -f 2 -;done &gt; $path_bam/temp_case</span><br><span class="line">echo "The case is about:" </span><br><span class="line">cat $path_bam/temp_case</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle: 碱基质量重校正BQSR</span></span><br><span class="line">cat $path_bam/temp_case | while read id</span><br><span class="line"><span class="meta">#</span><span class="bash">id=case3_biorep_B</span></span><br><span class="line">do</span><br><span class="line">    if [ ! -f $path_bam/$&#123;id&#125;$&#123;bs&#125;_bqsr.bam ]; then</span><br><span class="line">    </span><br><span class="line">    # BaseRecalibrator</span><br><span class="line">    echo "start BaseRecalibrator for $&#123;id&#125;..." `date`</span><br><span class="line">    $GATK --java-options "-Xmx60G -Djava.io.tmpdir=./" \</span><br><span class="line">    	BaseRecalibrator \</span><br><span class="line">    	-R $ref \</span><br><span class="line">    	-I $path_bam/$&#123;id&#125;$&#123;bs&#125;.bam \</span><br><span class="line">    	--known-sites $snp \</span><br><span class="line">    	--known-sites $indel \</span><br><span class="line">    	-O $path_bam/$&#123;id&#125;$&#123;bs&#125;_recal.table \</span><br><span class="line">    	&gt; $path_log/BaseRecalibrator_$&#123;id&#125;.log 2&gt;&amp;1 </span><br><span class="line">    echo "end BaseRecalibrator for $&#123;id&#125;!" `date`</span><br><span class="line">    </span><br><span class="line">    # ApplyBQSR</span><br><span class="line">    echo "start ApplyBQSR for $&#123;id&#125;..." `date`</span><br><span class="line">    $GATK --java-options "-Xmx60G -Djava.io.tmpdir=./" \</span><br><span class="line">    	ApplyBQSR \</span><br><span class="line">    	-R $ref \</span><br><span class="line">    	-I $path_bam/$&#123;id&#125;$&#123;bs&#125;.bam \</span><br><span class="line">    	-bqsr $path_bam/$&#123;id&#125;$&#123;bs&#125;_recal.table \</span><br><span class="line">    	-O $path_bam/$&#123;id&#125;$&#123;bs&#125;_bqsr.bam \</span><br><span class="line">    	&gt; $path_log/ApplyBQSR_$&#123;id&#125;.log 2&gt;&amp;1 </span><br><span class="line">    echo "end ApplyBQSR for $&#123;id&#125;!" `date`</span><br><span class="line">    </span><br><span class="line">    # samtools index</span><br><span class="line">    nohup samtools index -@ $nthread $path_bam/$&#123;id&#125;_marked_addRG_bqsr.bam &gt; $path_log/$&#123;id&#125;_marked_addRG_bqsr_samtools_index.log 2&gt;&amp;1 &amp;</span><br><span class="line">    echo "end samtools index for $&#123;id&#125;!"</span><br><span class="line">    else </span><br><span class="line">    echo "$path_bam/$&#123;id&#125;$&#123;bs&#125;_bqsr.bam had been existed."</span><br><span class="line">    fi</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> file management</span></span><br><span class="line"><span class="meta">#</span><span class="bash">ls <span class="variable">$path_log</span> | grep <span class="string">"BaseRecalibrator_"</span> | <span class="keyword">while</span> <span class="built_in">read</span> i;<span class="keyword">do</span> cat <span class="variable">$path_log</span>/<span class="variable">$i</span> &gt;&gt; <span class="variable">$path_log</span>/Whole_BaseRecalibrator.log;<span class="keyword">done</span>; <span class="comment"># rm -r BaseRecalibrator_*</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash">ls <span class="variable">$path_log</span> | grep <span class="string">"ApplyBQSR_"</span> | <span class="keyword">while</span> <span class="built_in">read</span> i;<span class="keyword">do</span> cat <span class="variable">$path_log</span>/<span class="variable">$i</span> &gt;&gt; <span class="variable">$path_log</span>/Whole_ApplyBQSR.log;<span class="keyword">done</span>; <span class="comment"># rm -r ApplyBQSR_*</span></span></span><br><span class="line">rm $path_bam/temp_case</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> Report</span></span><br><span class="line">echo "Done!"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> End <span class="keyword">function</span> gatk_BQSR</span></span><br></pre></td></tr></table></figure>

<h2 id="gatk-HaplotypeCaller"><a href="#gatk-HaplotypeCaller" class="headerlink" title="gatk_HaplotypeCaller"></a>gatk_HaplotypeCaller</h2><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/gatk_HaplotypeCaller</span><br><span class="line">vim $nbin/gatk_HaplotypeCaller</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function gatk_HaplotypeCaller</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> value</span></span><br><span class="line">ws=$1</span><br><span class="line">path_bam=$ws/output/align</span><br><span class="line">path_log=$ws/log/gatk_HaplotypeCaller</span><br><span class="line">GATK=/home/huangwb8/Downloads/conda/location/bin/gatk # which GATK4</span><br><span class="line">path_genome=/data/reference/genome/hg38</span><br><span class="line">bam_postfix="_marked_bqsr.bam"</span><br><span class="line">mkdir -p $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> mkdir -p <span class="variable">$ws</span>/<span class="built_in">log</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup gatk_HaplotypeCaller <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/gatk_HaplotypeCaller.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/gatk_HaplotypeCaller/*.<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> cycle: 生成 vcf</span></span><br><span class="line">ls $path_bam | grep $&#123;bam_postfix&#125; | while read id </span><br><span class="line">do </span><br><span class="line"><span class="meta">  #</span><span class="bash"> id=case3_biorep_B_marked_bqsr.bam</span></span><br><span class="line">  base=$(basename $&#123;id&#125; ".bam")</span><br><span class="line"><span class="meta">  #</span><span class="bash"> HaplotypeCaller: 生成gvcf中间文件。 210.55 minutes.</span></span><br><span class="line">  if [ ! -f $path_bam/$base.vcf ]; then </span><br><span class="line">    echo `date`" $&#123;base&#125;: do HaplotypeCaller ..."</span><br><span class="line">    time $GATK HaplotypeCaller \</span><br><span class="line">    -R $path_genome/hg38.fa \</span><br><span class="line">    --emit-ref-confidence GVCF \</span><br><span class="line">    -I $path_bam/$id \</span><br><span class="line">    -O $path_bam/$&#123;base&#125;.g.vcf \</span><br><span class="line">    &gt; $path_log/$&#123;base&#125;_HaplotypeCaller.log 2&gt;&amp;1 </span><br><span class="line">    echo `date`" $&#123;base&#125;: done HaplotypeCaller!" </span><br><span class="line"></span><br><span class="line">    # GenotypeGVCFs: 通过gvcf检测变异。5.98 minutes</span><br><span class="line">    echo `date`"$&#123;base&#125;: Note: GenotypeGVCFs"</span><br><span class="line">    time $GATK GenotypeGVCFs \</span><br><span class="line">    -R $path_genome/hg38.fa \</span><br><span class="line">    -V $path_bam/$base.g.vcf \</span><br><span class="line">    -O $path_bam/$&#123;base&#125;.vcf \</span><br><span class="line">    &gt; $path_log/$&#123;base&#125;_GenotypeGVCFs.log 2&gt;&amp;1</span><br><span class="line">    echo `date`" $&#123;base&#125;: done GenotypeGVCFs!"</span><br><span class="line">    </span><br><span class="line">  else</span><br><span class="line">    echo `date`" $&#123;base&#125;.vcf existed. Ignore!"</span><br><span class="line">  fi</span><br><span class="line">    </span><br><span class="line">    # 用bgzip对VCF进行压缩，并用tabix为它构建索引(tbi)</span><br><span class="line">    echo `date`" $&#123;base&#125;: index..."</span><br><span class="line">    time bgzip -f $path_bam/$base.vcf</span><br><span class="line">    time tabix -p vcf $path_bam/$base.vcf.gz</span><br><span class="line">    echo `date`" $&#123;base&#125;: index done!"</span><br><span class="line">  </span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># file management</span></span></span><br><span class="line">echo `date`"rm all *.g.vcf file."</span><br><span class="line">ls $path_bam/*.g.vcf | xagus -i rm &#123;&#125;</span><br><span class="line">echo "All done!"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End function: gatk_HaplotypeCaller</span></span></span><br></pre></td></tr></table></figure>

<h2 id="gatk-Mutect2"><a href="#gatk-Mutect2" class="headerlink" title="gatk_Mutect2"></a>gatk_Mutect2</h2><blockquote>
<p><code>${tumorsampletag}</code> <code>tumorsampletag</code> 分别对应normalsample和tumorsample bam文件header里的sample记录信息。</p>
<p>Mutect2: <code>${germline_resource}</code> 来自 gnomAD resource，包含了～200K exomes测序的 allele-sepcific population frequency信息。作为参考的population germline AF。</p>
<p><code>${Interval}</code> 是WES interval 的bed文件。</p>
<p>和germline过程一样，需要过滤。</p>
<p>首先 GetPileupSummaries 收集每个突变位点上ref 和 alt 的支持reads 数量。</p>
<p>calculatecontamination 估计样本cross contaminated的程度。这里的样本交叉是指不同sample间的污染不是normal和tumor间的污染。</p>
</blockquote>
<p><img src="https://picture.hwb0307.top/images/2019/09/27/OOS3rFVovVx1lYt4.jpg" alt="gatk_肿瘤组织突变挖掘.jpg"></p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/gatk_Mutect2</span><br><span class="line">vim $nbin/gatk_Mutect2</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: gatk_Mutect2</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> wes类函数</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过GATK Mutect2对两组（比如germline vs. tumor）进行比较，从而寻找tumor-specified基因变异。</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> mutect2_config文件的内容：第1列sample对应的分组名（如患者id）；第2列是germline样本的bam的id。第2列是tumor样本的bam的id。</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 肿瘤/正常样本可否为多个？</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup gatk_Mutect2 <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/gatk_Mutect2.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/gatk_Mutect2.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Introduction</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> Mutect2中的-tumor和-normal一般要用比较原始的bam header名才行。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters:</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_bam=$ws/output/align</span><br><span class="line">path_res=$ws/output/gatk_Mutect2</span><br><span class="line">path_log=$ws/log/gatk_Mutect2</span><br><span class="line">config=$ws/input/mutect2_config # 按要求改</span><br><span class="line">GATK=/home/huangwb8/Downloads/conda/location/bin/gatk # which gatk</span><br><span class="line">ref=/data/reference/index/gatk/hg38/Homo_sapiens_assembly38.fasta</span><br><span class="line">bed=/data/reference/index/gatk/hg38/gatk_internal.list_hg38.exon.bed</span><br><span class="line">mkdir -p $path_log $path_res</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe: </span></span></span><br><span class="line">cat $config |while read id</span><br><span class="line">do</span><br><span class="line">    # id="case6 case6_germline_marked_bqsr.bam case6_biorep_A_techrep_1_marked_bqsr.bam"</span><br><span class="line">    arr=($&#123;id&#125;)</span><br><span class="line">    sample=$&#123;arr[0]&#125;</span><br><span class="line">    T=$path_bam/$&#123;arr[2]&#125;</span><br><span class="line">    N=$path_bam/$&#123;arr[1]&#125;</span><br><span class="line">    </span><br><span class="line">    # Call candidate variants</span><br><span class="line">    echo "$&#123;sample&#125;: start Mutect2..." `date`</span><br><span class="line">    time $GATK  --java-options "-Xmx60G -Djava.io.tmpdir=./"  Mutect2 \</span><br><span class="line">    -R $ref \</span><br><span class="line">    -I $&#123;T&#125; -tumor $(basename $T "_marked_bqsr.bam") \</span><br><span class="line">    -I $&#123;N&#125; -normal $(basename $N "_marked_bqsr.bam") \</span><br><span class="line">    -L $bed  \</span><br><span class="line">    -O $path_res/$&#123;sample&#125;_mutect2.vcf \</span><br><span class="line">    1&gt; $path_log/$&#123;sample&#125;_gatk-Mutect2.log 2&gt;&amp;1</span><br><span class="line">    echo "$&#123;sample&#125;: Mutect2 Done! " `date`</span><br><span class="line">    </span><br><span class="line">    ## 要改动</span><br><span class="line">    $gatk GetPileupSummaries \</span><br><span class="line">    -I $&#123;work_dir&#125;/output/sm_$&#123;sm&#125;/*.ApplyBQSR.bam \</span><br><span class="line">    -V $&#123;resource&#125; \</span><br><span class="line">    -L $&#123;Interval&#125; \ </span><br><span class="line">    -O $&#123;filter_dir&#125;/table/$&#123;sm&#125;.getpileupsummaries.table</span><br><span class="line"></span><br><span class="line">    $gatk CalculateContamination \</span><br><span class="line">    -I $&#123;filter_dir&#125;/table/$&#123;sm&#125;.getpileupsummaries.table \</span><br><span class="line">    -O $&#123;filter_dir&#125;/table/$&#123;sm&#125;.calculatecontamination.table</span><br><span class="line">    </span><br><span class="line">    sm1=$&#123;sm%%_*&#125;.$&#123;sm##*_&#125;</span><br><span class="line"> </span><br><span class="line">    # gatk FilterMutectCalls</span><br><span class="line">    echo "$&#123;sample&#125;: do FilterMutectCalls... " `date` </span><br><span class="line">    time $GATK  FilterMutectCalls \</span><br><span class="line">    -V $path_res/$&#123;sample&#125;_mutect2.vcf \</span><br><span class="line">    -O $path_res/$&#123;sample&#125;_somatic.vcf \</span><br><span class="line">    1&gt; $path_log/$&#123;sample&#125;_gatk-FilterMutectCalls.log 2&gt;&amp;1</span><br><span class="line">    echo "$&#123;sample&#125;: FilterMutectCalls done! " `date` </span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># result filtering</span></span></span><br><span class="line">for i in $path_res/*_somatic.vcf</span><br><span class="line">do</span><br><span class="line">    j=$(basename $&#123;i&#125; "_somatic.vcf")</span><br><span class="line">    echo "$j: create filter.vcf files ..."`date` </span><br><span class="line">    cat $i | perl -alne '&#123;if(/^#/)&#123;print&#125;else&#123;next unless $F[6] eq "PASS";next if $F[0] =~/_/;print &#125; &#125;' &gt; $path_res/$&#123;j&#125;_filter.vcf</span><br><span class="line">    echo "$j: create filter.vcf files completed!"`date` </span><br><span class="line">done</span><br><span class="line">echo "gatk_Mutect2: All done!"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End gatk_Mutect2</span></span></span><br></pre></td></tr></table></figure>



<h2 id="Chiq-Seq-RIP-Seq-series"><a href="#Chiq-Seq-RIP-Seq-series" class="headerlink" title="Chiq-Seq/RIP-Seq series"></a>Chiq-Seq/RIP-Seq series</h2><h3 id="Code-find-peaks"><a href="#Code-find-peaks" class="headerlink" title="Code: find peaks"></a>Code: find peaks</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">mkdir -p $ws/output/peaks $ws/log/peaks</span><br><span class="line">cd $ws/output/align</span><br><span class="line"><span class="meta">#</span><span class="bash"> 参数：</span></span><br><span class="line"><span class="meta">#</span><span class="bash">-B: save extended fragment pileup, and <span class="built_in">local</span> lambda tracks (two files) at every bp into a bedGraph file. DEFAULT: False </span></span><br><span class="line"><span class="meta">#</span><span class="bash">-g: genome sizes。<span class="string">'mm'</span> <span class="keyword">for</span> mouse(1.87e9)</span></span><br><span class="line"><span class="meta">#</span><span class="bash">-n: Experiment name, <span class="built_in">which</span> will be used to generate output file names</span></span><br><span class="line"><span class="meta">#</span><span class="bash">--tempdir: 临时文件的保存位置</span></span><br><span class="line">ls *merge.rmdup.bam | grep -v Control | while read id; do </span><br><span class="line">    nohup macs2 callpeak \</span><br><span class="line">        --tempdir /data/tmp \</span><br><span class="line">        -c Control*bam -t $id \</span><br><span class="line">        -f BAM -B -g mm \</span><br><span class="line">        -n $(basename $id ".bam") \</span><br><span class="line">        --outdir $ws/output/peaks \</span><br><span class="line">        &gt; $ws/log/peaks/$(basename $id ".bam").log 2&gt;&amp;1 &amp;</span><br><span class="line">done</span><br><span class="line">tail -f $ws/log/peaks/*.log</span><br></pre></td></tr></table></figure>



<h3 id="macs2-peak"><a href="#macs2-peak" class="headerlink" title="macs2_peak"></a>macs2_peak</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/macs2_peak</span><br><span class="line">vim $nbin/macs2_peak</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function: macs2_peak</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 为macs2制作expression和names,并最后运行macs2 callpeak</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -B: save extended fragment pileup, and <span class="built_in">local</span> lambda tracks (two files) at every bp into a bedGraph file. DEFAULT: False </span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -g: genome sizes。<span class="string">'mm'</span> <span class="keyword">for</span> mouse(1.87e9); hs</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> -n: Experiment name, <span class="built_in">which</span> will be used to generate output file names</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> --tempdir: 临时文件的保存位置</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup macs2_peak <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/macs2_peak.log &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/macs2_peak.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_case=$ws/input/case</span><br><span class="line">peakSampleID=$&#123;path_case&#125;/peakSampleID.txt # 自定义。第一列是每个组的组名，其他列为样本（包括control和treatment。个数不限）。</span><br><span class="line">bamtype=hisat2 # 看情况 </span><br><span class="line">path_bam=$ws/output/align/$&#123;bamtype&#125;</span><br><span class="line">control_symbol="null" # 自定义。不是control_symbol的被认为全是treatment</span><br><span class="line">species=hs # mm是鼠,hs是人</span><br><span class="line">MACS2=/home/huangwb8/Downloads/conda/location/envs/chip-seq/bin/macs2</span><br><span class="line">path_res=$ws/output/peaks/$&#123;bamtype&#125;</span><br><span class="line">path_log=$ws/log/peaks/$&#123;bamtype&#125;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># file management</span></span></span><br><span class="line">mkdir -p $path_res $path_log</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> get macs2_expression</span></span><br><span class="line">/usr/bin/Rscript \</span><br><span class="line">    /home/huangwb8/bin/example/macs2_expression.R \</span><br><span class="line">    $&#123;peakSampleID&#125; \</span><br><span class="line">    $&#123;path_bam&#125; \</span><br><span class="line">    $&#123;control_symbol&#125;</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> find peaks by macs2</span></span><br><span class="line">cat $&#123;path_case&#125;/macs2_expression | while read id; do</span><br><span class="line">    name=$(echo $id | cut -d' ' -f 1)</span><br><span class="line">    expr=$(echo "'$id'" | xargs -i fast_extra &#123;&#125; "$&#123;name&#125; ")</span><br><span class="line">    </span><br><span class="line">    if [ ! -f $path_res/$&#123;name&#125;_control_lambda.bdg ]; then</span><br><span class="line">        ## find peaks by macs2</span><br><span class="line">        nohup $MACS2 callpeak \</span><br><span class="line">            --tempdir /data/tmp \</span><br><span class="line">            $&#123;expr&#125; \</span><br><span class="line">            -f BAM -B -g $&#123;species&#125; \</span><br><span class="line">            -n $&#123;name&#125; \</span><br><span class="line">            --outdir $path_res \</span><br><span class="line">            &gt; $path_log/$&#123;names&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line">    else </span><br><span class="line">    	echo "$&#123;name&#125;: MACS2 callpeak had been done. Ignored!"</span><br><span class="line">    fi</span><br><span class="line">done</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End</span></span></span><br></pre></td></tr></table></figure>



<h3 id="signal-tss"><a href="#signal-tss" class="headerlink" title="signal_tss"></a>signal_tss</h3><blockquote>
<p>chip-seq时查看TSS附近的信号强度</p>
</blockquote>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/signal_tss</span><br><span class="line">vim $nbin/signal_tss</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function:signal_tss</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 做chip-seq分析时, 基于基因组bed，样本bw文件，观察整体的TSS（transcription start sites）附近的信号强度</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 还可以画其它模式: computeMatrix scale-regions</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup signal_tss <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/signal_tss.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/signal_tss.log</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_tss=$ws/output/signal_tss</span><br><span class="line"><span class="meta">#</span><span class="bash">path_bed=/data/reference/bed/ChipSeq/mm9_RefSeq.bed <span class="comment"># 按需要选择。这里是mm9</span></span></span><br><span class="line">path_bed=/data/reference/bed/ChipSeq/hg38_refSeq_Chip.bed</span><br><span class="line">path_log=$ws/log/signal_tss</span><br><span class="line">path_bw=$ws/output/peaks/bw</span><br><span class="line">path_plot=$path_tss/plot</span><br><span class="line">nthread=48</span><br><span class="line">nD=10000 # 一般要画10k和2k的</span><br><span class="line">ComputeMatrix=/home/huangwb8/Downloads/conda/location/envs/chip-seq/bin/computeMatrix</span><br><span class="line">mkdir -p $path_tss $path_log $path_plot</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># report</span></span></span><br><span class="line">echo "Parameters:</span><br><span class="line">workspace:$ws</span><br><span class="line">path_tss: $path_tss</span><br><span class="line">path_bed: $path_bed</span><br><span class="line">path_bw: $path_bw</span><br><span class="line">path_plot: $path_plot</span><br><span class="line">nD: $nD"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe:</span></span></span><br><span class="line">ls $path_bw | grep 'bw$' |while read id </span><br><span class="line"><span class="meta">#</span><span class="bash"> id=H2Aub1.ChIPSeq-GSM850471.merge.rmdup.bw</span></span><br><span class="line">do</span><br><span class="line">    base=$(basename $&#123;id&#125; ".bw")</span><br><span class="line">    # get TSS.gz files</span><br><span class="line">    #computeMatrix reference-point --referencePoint TSS -p $nthread -b 10000 -a 10000 -R $path_bed -S $&#123;path_bw&#125;/$&#123;id&#125; --skipZeros -o $&#123;path_tss&#125;/$&#123;base&#125;_TSS_$&#123;nD&#125;.gz --outFileSortedRegions $&#123;path_tss&#125;/$&#123;base&#125;_TSS_$&#123;nD&#125;.bed </span><br><span class="line">    echo "$&#123;base&#125;: computeMatrix reference-point process..."</span><br><span class="line">    time $ComputeMatrix reference-point \</span><br><span class="line">    	--referencePoint TSS \</span><br><span class="line">        -p $nthread -b $nD -a $nD \</span><br><span class="line">        -R $path_bed \</span><br><span class="line">        -S $&#123;path_bw&#125;/$&#123;id&#125; \</span><br><span class="line">        --skipZeros -o $&#123;path_tss&#125;/$&#123;base&#125;_TSS_$&#123;nD&#125;.gz \</span><br><span class="line">        --outFileSortedRegions $&#123;path_tss&#125;/$&#123;base&#125;_TSS_$&#123;nD&#125;.bed \</span><br><span class="line">        &gt; $path_log/$&#123;base&#125;_TSS_$&#123;nD&#125;.log 2&gt;&amp;1</span><br><span class="line">    echo "$&#123;base&#125;: computeMatrix reference-point done!"</span><br><span class="line"></span><br><span class="line">    # Plots</span><br><span class="line">    echo "$&#123;base&#125;: plot heatmaps..."</span><br><span class="line">    plotHeatmap -m $&#123;path_tss&#125;/$&#123;base&#125;_TSS_$&#123;nD&#125;.gz -out $&#123;path_plot&#125;/$&#123;base&#125;_Heatmap_$&#123;nD&#125;.png</span><br><span class="line">    plotHeatmap -m $&#123;path_tss&#125;/$&#123;base&#125;_TSS_$&#123;nD&#125;.gz -out $&#123;path_plot&#125;/$&#123;base&#125;_Heatmap_$&#123;nD&#125;.pdf --plotFileFormat pdf --dpi 720</span><br><span class="line">    plotProfile -m $&#123;path_tss&#125;/$&#123;base&#125;_TSS_$&#123;nD&#125;.gz -out $&#123;path_plot&#125;/$&#123;base&#125;_Profile_$&#123;nD&#125;.png</span><br><span class="line">    plotProfile -m $&#123;path_tss&#125;/$&#123;base&#125;_TSS_$&#123;nD&#125;.gz -out $&#123;path_plot&#125;/$&#123;base&#125;_Profile_$&#123;nD&#125;.pdf --plotFileFormat pdf --dpi 720</span><br><span class="line">    echo "$&#123;base&#125;: plot heatmaps done!"</span><br><span class="line">done</span><br><span class="line">echo "Done!"</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End signal_tss</span></span></span><br></pre></td></tr></table></figure>



<h3 id="homer-motif"><a href="#homer-motif" class="headerlink" title="homer_motif"></a>homer_motif</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br></pre></td><td class="code"><pre><span class="line">vim ~/bin/example/homer_motif</span><br><span class="line">vim $nbin/homer_motif</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># function : homer_motif</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Description:</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> 通过homer软件，对.bed文件进行peak motif的寻找及peak注释。此程序中，findMotifsGenome步骤是限速步骤。</span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Usage</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"> nohup homer_motif <span class="variable">$ws</span> &gt; <span class="variable">$ws</span>/<span class="built_in">log</span>/homer_motif.log 2&gt;&amp;1 &amp;</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> cat <span class="variable">$ws</span>/<span class="built_in">log</span>/homer_motif.log</span></span><br><span class="line"><span class="meta">#</span><span class="bash"> tail -f <span class="variable">$ws</span>/<span class="built_in">log</span>/homer_motif/*.<span class="built_in">log</span></span></span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Parameters</span></span></span><br><span class="line">ws=$1</span><br><span class="line">path_log=$ws/log/homer_motif</span><br><span class="line">path_bed=$ws/output/peaks</span><br><span class="line">path_homer=/data/reference/homer/bin #homer安装位置</span><br><span class="line">path_res=$ws/output/homer_motif</span><br><span class="line">findMotifsGenome=$path_homer/findMotifsGenome.pl</span><br><span class="line">annotatePeaks=$path_homer/annotatePeaks.pl</span><br><span class="line">genome=/data/reference/homer/data/genomes/mm9 #homer基因组所在文件夹。</span><br><span class="line">mkdir -p $path_log $path_res</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"> parameter report</span></span><br><span class="line">echo "homer_motif by Weibin Huang" `date`</span><br><span class="line">echo "path_log=$path_log"</span><br><span class="line">echo "path_bed=$path_bed"</span><br><span class="line">echo "genome=$genome"</span><br><span class="line"></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Programe</span></span></span><br><span class="line">ls $path_bed | grep "bed$" | while read id</span><br><span class="line">do</span><br><span class="line"><span class="meta">  #</span><span class="bash"> id=H2Aub1.ChIPSeq-GSM850471.merge.rmdup_summits.bed</span></span><br><span class="line">  echo "$&#123;id&#125;"</span><br><span class="line">  case=$(basename $&#123;id&#125; ".bed")</span><br><span class="line">  awk '&#123;print $4"\t"$1"\t"$2"\t"$3"\t+"&#125;' $path_bed/$id &gt; $path_bed/homer_peaks.tmp</span><br><span class="line">  </span><br><span class="line"><span class="meta">  #</span><span class="bash"> findMotifsGenome寻找富集Motifs</span></span><br><span class="line">  if [ ! -d $path_res/$&#123;case&#125;_motifDir ]; then</span><br><span class="line">    echo "findMotifsGenome: $&#123;case&#125;..."</span><br><span class="line">    nohup time $findMotifsGenome \</span><br><span class="line">        $path_bed/homer_peaks.tmp \</span><br><span class="line">        $genome \</span><br><span class="line">        $path_res/$&#123;case&#125;_motifDir \</span><br><span class="line">        -len 8,10,12 \ # 这个要自己选好。想看哪些length的motif</span><br><span class="line">        &gt; $path_log/findMotifsGenome_$&#123;case&#125;.log 2&gt;&amp;1 &amp;</span><br><span class="line">    echo "findMotifsGenome: done!"</span><br><span class="line">  else </span><br><span class="line">    echo "homer_motif findMotifsGenome: $&#123;case&#125; had existed! Ignore it"</span><br><span class="line">  fi</span><br><span class="line"></span><br><span class="line"><span class="meta">  #</span><span class="bash"> annotatePeaks对Peaks进行注释</span></span><br><span class="line">  if [ ! -f $path_res/$&#123;case&#125;_peakAnn.xls ]; then </span><br><span class="line">    echo "annotatePeaks: $&#123;case&#125;..."</span><br><span class="line">    nohup time $annotatePeaks \</span><br><span class="line">    $path_bed/homer_peaks.tmp \</span><br><span class="line">    $genome \</span><br><span class="line">    1&gt;$path_res/$&#123;case&#125;_peakAnn.xls \</span><br><span class="line">    2&gt;$path_res/$&#123;case&#125;_annLog.txt &amp;</span><br><span class="line">    echo "annotatePeaks: done!"</span><br><span class="line">  else </span><br><span class="line">    echo "homer_motif annotatePeaks: $&#123;case&#125; had existed! Ignore it"</span><br><span class="line">  fi</span><br><span class="line">done</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># file management</span></span></span><br><span class="line">rm $path_bed/homer_peaks.tmp</span><br><span class="line">echo "All done!"</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># End homer_motif</span></span></span><br></pre></td></tr></table></figure>




















      

      
    </div>
    <div class="article-info article-info-index">
      
      
      

      
        <p class="article-more-link">
          <a class="article-more-a" href="/2020/02/09/bioinf/">展开全文 >></a>
        </p>
      

      
      <div class="clearfix"></div>
    </div>
  </div>
</article>

<aside class="wrap-side-operation">
    <div class="mod-side-operation">
        
        <div class="jump-container" id="js-jump-container" style="display:none;">
            <a href="javascript:void(0)" class="mod-side-operation__jump-to-top">
                <i class="icon-font icon-back"></i>
            </a>
            <div id="js-jump-plan-container" class="jump-plan-container" style="top: -11px;">
                <i class="icon-font icon-plane jump-plane"></i>
            </div>
        </div>
        
        
    </div>
</aside>




  
  


          </div>
        </div>
      </div>
      <footer id="footer">
  <div class="outer">
    <div id="footer-info">
    	<div class="footer-left">
    		&copy; 2021 Weber Huang
    	</div>
      	<div class="footer-right">
      		<a href="http://hexo.io/" target="_blank">Hexo</a>  Theme <a href="https://github.com/litten/hexo-theme-yilia" target="_blank">Yilia</a> by Litten
      	</div>
    </div>
  </div>
</footer>
    </div>
    <script>
	var yiliaConfig = {
		mathjax: false,
		isHome: true,
		isPost: false,
		isArchive: false,
		isTag: false,
		isCategory: false,
		open_in_new: false,
		toc_hide_index: true,
		root: "/",
		innerArchive: true,
		showTags: false
	}
</script>

<script>!function(t){function n(e){if(r[e])return r[e].exports;var i=r[e]={exports:{},id:e,loaded:!1};return t[e].call(i.exports,i,i.exports,n),i.loaded=!0,i.exports}var r={};n.m=t,n.c=r,n.p="./",n(0)}([function(t,n,r){r(195),t.exports=r(191)},function(t,n,r){var e=r(3),i=r(52),o=r(27),u=r(28),c=r(53),f="prototype",a=function(t,n,r){var s,l,h,v,p=t&a.F,d=t&a.G,y=t&a.S,g=t&a.P,b=t&a.B,m=d?e:y?e[n]||(e[n]={}):(e[n]||{})[f],x=d?i:i[n]||(i[n]={}),w=x[f]||(x[f]={});d&&(r=n);for(s in r)l=!p&&m&&void 0!==m[s],h=(l?m:r)[s],v=b&&l?c(h,e):g&&"function"==typeof h?c(Function.call,h):h,m&&u(m,s,h,t&a.U),x[s]!=h&&o(x,s,v),g&&w[s]!=h&&(w[s]=h)};e.core=i,a.F=1,a.G=2,a.S=4,a.P=8,a.B=16,a.W=32,a.U=64,a.R=128,t.exports=a},function(t,n,r){var e=r(6);t.exports=function(t){if(!e(t))throw TypeError(t+" is not an object!");return t}},function(t,n){var r=t.exports="undefined"!=typeof window&&window.Math==Math?window:"undefined"!=typeof self&&self.Math==Math?self:Function("return this")();"number"==typeof __g&&(__g=r)},function(t,n){t.exports=function(t){try{return!!t()}catch(t){return!0}}},function(t,n){var r=t.exports="undefined"!=typeof window&&window.Math==Math?window:"undefined"!=typeof self&&self.Math==Math?self:Function("return this")();"number"==typeof __g&&(__g=r)},function(t,n){t.exports=function(t){return"object"==typeof t?null!==t:"function"==typeof t}},function(t,n,r){var e=r(126)("wks"),i=r(76),o=r(3).Symbol,u="function"==typeof o;(t.exports=function(t){return e[t]||(e[t]=u&&o[t]||(u?o:i)("Symbol."+t))}).store=e},function(t,n){var r={}.hasOwnProperty;t.exports=function(t,n){return r.call(t,n)}},function(t,n,r){var e=r(94),i=r(33);t.exports=function(t){return e(i(t))}},function(t,n,r){t.exports=!r(4)(function(){return 7!=Object.defineProperty({},"a",{get:function(){return 7}}).a})},function(t,n,r){var e=r(2),i=r(167),o=r(50),u=Object.defineProperty;n.f=r(10)?Object.defineProperty:function(t,n,r){if(e(t),n=o(n,!0),e(r),i)try{return u(t,n,r)}catch(t){}if("get"in r||"set"in r)throw TypeError("Accessors not supported!");return"value"in r&&(t[n]=r.value),t}},function(t,n,r){t.exports=!r(18)(function(){return 7!=Object.defineProperty({},"a",{get:function(){return 7}}).a})},function(t,n,r){var e=r(14),i=r(22);t.exports=r(12)?function(t,n,r){return e.f(t,n,i(1,r))}:function(t,n,r){return t[n]=r,t}},function(t,n,r){var e=r(20),i=r(58),o=r(42),u=Object.defineProperty;n.f=r(12)?Object.defineProperty:function(t,n,r){if(e(t),n=o(n,!0),e(r),i)try{return u(t,n,r)}catch(t){}if("get"in r||"set"in r)throw TypeError("Accessors not supported!");return"value"in r&&(t[n]=r.value),t}},function(t,n,r){var e=r(40)("wks"),i=r(23),o=r(5).Symbol,u="function"==typeof o;(t.exports=function(t){return e[t]||(e[t]=u&&o[t]||(u?o:i)("Symbol."+t))}).store=e},function(t,n,r){var e=r(67),i=Math.min;t.exports=function(t){return t>0?i(e(t),9007199254740991):0}},function(t,n,r){var e=r(46);t.exports=function(t){return Object(e(t))}},function(t,n){t.exports=function(t){try{return!!t()}catch(t){return!0}}},function(t,n,r){var e=r(63),i=r(34);t.exports=Object.keys||function(t){return e(t,i)}},function(t,n,r){var e=r(21);t.exports=function(t){if(!e(t))throw TypeError(t+" is not an object!");return t}},function(t,n){t.exports=function(t){return"object"==typeof t?null!==t:"function"==typeof t}},function(t,n){t.exports=function(t,n){return{enumerable:!(1&t),configurable:!(2&t),writable:!(4&t),value:n}}},function(t,n){var r=0,e=Math.random();t.exports=function(t){return"Symbol(".concat(void 0===t?"":t,")_",(++r+e).toString(36))}},function(t,n){var r={}.hasOwnProperty;t.exports=function(t,n){return r.call(t,n)}},function(t,n){var r=t.exports={version:"2.4.0"};"number"==typeof __e&&(__e=r)},function(t,n){t.exports=function(t){if("function"!=typeof t)throw TypeError(t+" is not a function!");return t}},function(t,n,r){var e=r(11),i=r(66);t.exports=r(10)?function(t,n,r){return e.f(t,n,i(1,r))}:function(t,n,r){return t[n]=r,t}},function(t,n,r){var e=r(3),i=r(27),o=r(24),u=r(76)("src"),c="toString",f=Function[c],a=(""+f).split(c);r(52).inspectSource=function(t){return f.call(t)},(t.exports=function(t,n,r,c){var f="function"==typeof r;f&&(o(r,"name")||i(r,"name",n)),t[n]!==r&&(f&&(o(r,u)||i(r,u,t[n]?""+t[n]:a.join(String(n)))),t===e?t[n]=r:c?t[n]?t[n]=r:i(t,n,r):(delete t[n],i(t,n,r)))})(Function.prototype,c,function(){return"function"==typeof this&&this[u]||f.call(this)})},function(t,n,r){var e=r(1),i=r(4),o=r(46),u=function(t,n,r,e){var i=String(o(t)),u="<"+n;return""!==r&&(u+=" "+r+'="'+String(e).replace(/"/g,"&quot;")+'"'),u+">"+i+"</"+n+">"};t.exports=function(t,n){var r={};r[t]=n(u),e(e.P+e.F*i(function(){var n=""[t]('"');return n!==n.toLowerCase()||n.split('"').length>3}),"String",r)}},function(t,n,r){var e=r(115),i=r(46);t.exports=function(t){return e(i(t))}},function(t,n,r){var e=r(116),i=r(66),o=r(30),u=r(50),c=r(24),f=r(167),a=Object.getOwnPropertyDescriptor;n.f=r(10)?a:function(t,n){if(t=o(t),n=u(n,!0),f)try{return a(t,n)}catch(t){}if(c(t,n))return i(!e.f.call(t,n),t[n])}},function(t,n,r){var e=r(24),i=r(17),o=r(145)("IE_PROTO"),u=Object.prototype;t.exports=Object.getPrototypeOf||function(t){return t=i(t),e(t,o)?t[o]:"function"==typeof t.constructor&&t instanceof t.constructor?t.constructor.prototype:t instanceof Object?u:null}},function(t,n){t.exports=function(t){if(void 0==t)throw TypeError("Can't call method on  "+t);return t}},function(t,n){t.exports="constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf".split(",")},function(t,n){t.exports={}},function(t,n){t.exports=!0},function(t,n){n.f={}.propertyIsEnumerable},function(t,n,r){var e=r(14).f,i=r(8),o=r(15)("toStringTag");t.exports=function(t,n,r){t&&!i(t=r?t:t.prototype,o)&&e(t,o,{configurable:!0,value:n})}},function(t,n,r){var e=r(40)("keys"),i=r(23);t.exports=function(t){return e[t]||(e[t]=i(t))}},function(t,n,r){var e=r(5),i="__core-js_shared__",o=e[i]||(e[i]={});t.exports=function(t){return o[t]||(o[t]={})}},function(t,n){var r=Math.ceil,e=Math.floor;t.exports=function(t){return isNaN(t=+t)?0:(t>0?e:r)(t)}},function(t,n,r){var e=r(21);t.exports=function(t,n){if(!e(t))return t;var r,i;if(n&&"function"==typeof(r=t.toString)&&!e(i=r.call(t)))return i;if("function"==typeof(r=t.valueOf)&&!e(i=r.call(t)))return i;if(!n&&"function"==typeof(r=t.toString)&&!e(i=r.call(t)))return i;throw TypeError("Can't convert object to primitive value")}},function(t,n,r){var e=r(5),i=r(25),o=r(36),u=r(44),c=r(14).f;t.exports=function(t){var n=i.Symbol||(i.Symbol=o?{}:e.Symbol||{});"_"==t.charAt(0)||t in n||c(n,t,{value:u.f(t)})}},function(t,n,r){n.f=r(15)},function(t,n){var r={}.toString;t.exports=function(t){return r.call(t).slice(8,-1)}},function(t,n){t.exports=function(t){if(void 0==t)throw TypeError("Can't call method on  "+t);return t}},function(t,n,r){var e=r(4);t.exports=function(t,n){return!!t&&e(function(){n?t.call(null,function(){},1):t.call(null)})}},function(t,n,r){var e=r(53),i=r(115),o=r(17),u=r(16),c=r(203);t.exports=function(t,n){var r=1==t,f=2==t,a=3==t,s=4==t,l=6==t,h=5==t||l,v=n||c;return function(n,c,p){for(var d,y,g=o(n),b=i(g),m=e(c,p,3),x=u(b.length),w=0,S=r?v(n,x):f?v(n,0):void 0;x>w;w++)if((h||w in b)&&(d=b[w],y=m(d,w,g),t))if(r)S[w]=y;else if(y)switch(t){case 3:return!0;case 5:return d;case 6:return w;case 2:S.push(d)}else if(s)return!1;return l?-1:a||s?s:S}}},function(t,n,r){var e=r(1),i=r(52),o=r(4);t.exports=function(t,n){var r=(i.Object||{})[t]||Object[t],u={};u[t]=n(r),e(e.S+e.F*o(function(){r(1)}),"Object",u)}},function(t,n,r){var e=r(6);t.exports=function(t,n){if(!e(t))return t;var r,i;if(n&&"function"==typeof(r=t.toString)&&!e(i=r.call(t)))return i;if("function"==typeof(r=t.valueOf)&&!e(i=r.call(t)))return i;if(!n&&"function"==typeof(r=t.toString)&&!e(i=r.call(t)))return i;throw TypeError("Can't convert object to primitive value")}},function(t,n,r){var e=r(5),i=r(25),o=r(91),u=r(13),c="prototype",f=function(t,n,r){var a,s,l,h=t&f.F,v=t&f.G,p=t&f.S,d=t&f.P,y=t&f.B,g=t&f.W,b=v?i:i[n]||(i[n]={}),m=b[c],x=v?e:p?e[n]:(e[n]||{})[c];v&&(r=n);for(a in r)(s=!h&&x&&void 0!==x[a])&&a in b||(l=s?x[a]:r[a],b[a]=v&&"function"!=typeof x[a]?r[a]:y&&s?o(l,e):g&&x[a]==l?function(t){var n=function(n,r,e){if(this instanceof t){switch(arguments.length){case 0:return new t;case 1:return new t(n);case 2:return new t(n,r)}return new t(n,r,e)}return t.apply(this,arguments)};return n[c]=t[c],n}(l):d&&"function"==typeof l?o(Function.call,l):l,d&&((b.virtual||(b.virtual={}))[a]=l,t&f.R&&m&&!m[a]&&u(m,a,l)))};f.F=1,f.G=2,f.S=4,f.P=8,f.B=16,f.W=32,f.U=64,f.R=128,t.exports=f},function(t,n){var r=t.exports={version:"2.4.0"};"number"==typeof __e&&(__e=r)},function(t,n,r){var e=r(26);t.exports=function(t,n,r){if(e(t),void 0===n)return t;switch(r){case 1:return function(r){return t.call(n,r)};case 2:return function(r,e){return t.call(n,r,e)};case 3:return function(r,e,i){return t.call(n,r,e,i)}}return function(){return t.apply(n,arguments)}}},function(t,n,r){var e=r(183),i=r(1),o=r(126)("metadata"),u=o.store||(o.store=new(r(186))),c=function(t,n,r){var i=u.get(t);if(!i){if(!r)return;u.set(t,i=new e)}var o=i.get(n);if(!o){if(!r)return;i.set(n,o=new e)}return o},f=function(t,n,r){var e=c(n,r,!1);return void 0!==e&&e.has(t)},a=function(t,n,r){var e=c(n,r,!1);return void 0===e?void 0:e.get(t)},s=function(t,n,r,e){c(r,e,!0).set(t,n)},l=function(t,n){var r=c(t,n,!1),e=[];return r&&r.forEach(function(t,n){e.push(n)}),e},h=function(t){return void 0===t||"symbol"==typeof t?t:String(t)},v=function(t){i(i.S,"Reflect",t)};t.exports={store:u,map:c,has:f,get:a,set:s,keys:l,key:h,exp:v}},function(t,n,r){"use strict";if(r(10)){var e=r(69),i=r(3),o=r(4),u=r(1),c=r(127),f=r(152),a=r(53),s=r(68),l=r(66),h=r(27),v=r(73),p=r(67),d=r(16),y=r(75),g=r(50),b=r(24),m=r(180),x=r(114),w=r(6),S=r(17),_=r(137),O=r(70),E=r(32),P=r(71).f,j=r(154),F=r(76),M=r(7),A=r(48),N=r(117),T=r(146),I=r(155),k=r(80),L=r(123),R=r(74),C=r(130),D=r(160),U=r(11),W=r(31),G=U.f,B=W.f,V=i.RangeError,z=i.TypeError,q=i.Uint8Array,K="ArrayBuffer",J="Shared"+K,Y="BYTES_PER_ELEMENT",H="prototype",$=Array[H],X=f.ArrayBuffer,Q=f.DataView,Z=A(0),tt=A(2),nt=A(3),rt=A(4),et=A(5),it=A(6),ot=N(!0),ut=N(!1),ct=I.values,ft=I.keys,at=I.entries,st=$.lastIndexOf,lt=$.reduce,ht=$.reduceRight,vt=$.join,pt=$.sort,dt=$.slice,yt=$.toString,gt=$.toLocaleString,bt=M("iterator"),mt=M("toStringTag"),xt=F("typed_constructor"),wt=F("def_constructor"),St=c.CONSTR,_t=c.TYPED,Ot=c.VIEW,Et="Wrong length!",Pt=A(1,function(t,n){return Tt(T(t,t[wt]),n)}),jt=o(function(){return 1===new q(new Uint16Array([1]).buffer)[0]}),Ft=!!q&&!!q[H].set&&o(function(){new q(1).set({})}),Mt=function(t,n){if(void 0===t)throw z(Et);var r=+t,e=d(t);if(n&&!m(r,e))throw V(Et);return e},At=function(t,n){var r=p(t);if(r<0||r%n)throw V("Wrong offset!");return r},Nt=function(t){if(w(t)&&_t in t)return t;throw z(t+" is not a typed array!")},Tt=function(t,n){if(!(w(t)&&xt in t))throw z("It is not a typed array constructor!");return new t(n)},It=function(t,n){return kt(T(t,t[wt]),n)},kt=function(t,n){for(var r=0,e=n.length,i=Tt(t,e);e>r;)i[r]=n[r++];return i},Lt=function(t,n,r){G(t,n,{get:function(){return this._d[r]}})},Rt=function(t){var n,r,e,i,o,u,c=S(t),f=arguments.length,s=f>1?arguments[1]:void 0,l=void 0!==s,h=j(c);if(void 0!=h&&!_(h)){for(u=h.call(c),e=[],n=0;!(o=u.next()).done;n++)e.push(o.value);c=e}for(l&&f>2&&(s=a(s,arguments[2],2)),n=0,r=d(c.length),i=Tt(this,r);r>n;n++)i[n]=l?s(c[n],n):c[n];return i},Ct=function(){for(var t=0,n=arguments.length,r=Tt(this,n);n>t;)r[t]=arguments[t++];return r},Dt=!!q&&o(function(){gt.call(new q(1))}),Ut=function(){return gt.apply(Dt?dt.call(Nt(this)):Nt(this),arguments)},Wt={copyWithin:function(t,n){return D.call(Nt(this),t,n,arguments.length>2?arguments[2]:void 0)},every:function(t){return rt(Nt(this),t,arguments.length>1?arguments[1]:void 0)},fill:function(t){return C.apply(Nt(this),arguments)},filter:function(t){return It(this,tt(Nt(this),t,arguments.length>1?arguments[1]:void 0))},find:function(t){return et(Nt(this),t,arguments.length>1?arguments[1]:void 0)},findIndex:function(t){return it(Nt(this),t,arguments.length>1?arguments[1]:void 0)},forEach:function(t){Z(Nt(this),t,arguments.length>1?arguments[1]:void 0)},indexOf:function(t){return ut(Nt(this),t,arguments.length>1?arguments[1]:void 0)},includes:function(t){return ot(Nt(this),t,arguments.length>1?arguments[1]:void 0)},join:function(t){return vt.apply(Nt(this),arguments)},lastIndexOf:function(t){return st.apply(Nt(this),arguments)},map:function(t){return Pt(Nt(this),t,arguments.length>1?arguments[1]:void 0)},reduce:function(t){return lt.apply(Nt(this),arguments)},reduceRight:function(t){return ht.apply(Nt(this),arguments)},reverse:function(){for(var t,n=this,r=Nt(n).length,e=Math.floor(r/2),i=0;i<e;)t=n[i],n[i++]=n[--r],n[r]=t;return n},some:function(t){return nt(Nt(this),t,arguments.length>1?arguments[1]:void 0)},sort:function(t){return pt.call(Nt(this),t)},subarray:function(t,n){var r=Nt(this),e=r.length,i=y(t,e);return new(T(r,r[wt]))(r.buffer,r.byteOffset+i*r.BYTES_PER_ELEMENT,d((void 0===n?e:y(n,e))-i))}},Gt=function(t,n){return It(this,dt.call(Nt(this),t,n))},Bt=function(t){Nt(this);var n=At(arguments[1],1),r=this.length,e=S(t),i=d(e.length),o=0;if(i+n>r)throw V(Et);for(;o<i;)this[n+o]=e[o++]},Vt={entries:function(){return at.call(Nt(this))},keys:function(){return ft.call(Nt(this))},values:function(){return ct.call(Nt(this))}},zt=function(t,n){return w(t)&&t[_t]&&"symbol"!=typeof n&&n in t&&String(+n)==String(n)},qt=function(t,n){return zt(t,n=g(n,!0))?l(2,t[n]):B(t,n)},Kt=function(t,n,r){return!(zt(t,n=g(n,!0))&&w(r)&&b(r,"value"))||b(r,"get")||b(r,"set")||r.configurable||b(r,"writable")&&!r.writable||b(r,"enumerable")&&!r.enumerable?G(t,n,r):(t[n]=r.value,t)};St||(W.f=qt,U.f=Kt),u(u.S+u.F*!St,"Object",{getOwnPropertyDescriptor:qt,defineProperty:Kt}),o(function(){yt.call({})})&&(yt=gt=function(){return vt.call(this)});var Jt=v({},Wt);v(Jt,Vt),h(Jt,bt,Vt.values),v(Jt,{slice:Gt,set:Bt,constructor:function(){},toString:yt,toLocaleString:Ut}),Lt(Jt,"buffer","b"),Lt(Jt,"byteOffset","o"),Lt(Jt,"byteLength","l"),Lt(Jt,"length","e"),G(Jt,mt,{get:function(){return this[_t]}}),t.exports=function(t,n,r,f){f=!!f;var a=t+(f?"Clamped":"")+"Array",l="Uint8Array"!=a,v="get"+t,p="set"+t,y=i[a],g=y||{},b=y&&E(y),m=!y||!c.ABV,S={},_=y&&y[H],j=function(t,r){var e=t._d;return e.v[v](r*n+e.o,jt)},F=function(t,r,e){var i=t._d;f&&(e=(e=Math.round(e))<0?0:e>255?255:255&e),i.v[p](r*n+i.o,e,jt)},M=function(t,n){G(t,n,{get:function(){return j(this,n)},set:function(t){return F(this,n,t)},enumerable:!0})};m?(y=r(function(t,r,e,i){s(t,y,a,"_d");var o,u,c,f,l=0,v=0;if(w(r)){if(!(r instanceof X||(f=x(r))==K||f==J))return _t in r?kt(y,r):Rt.call(y,r);o=r,v=At(e,n);var p=r.byteLength;if(void 0===i){if(p%n)throw V(Et);if((u=p-v)<0)throw V(Et)}else if((u=d(i)*n)+v>p)throw V(Et);c=u/n}else c=Mt(r,!0),u=c*n,o=new X(u);for(h(t,"_d",{b:o,o:v,l:u,e:c,v:new Q(o)});l<c;)M(t,l++)}),_=y[H]=O(Jt),h(_,"constructor",y)):L(function(t){new y(null),new y(t)},!0)||(y=r(function(t,r,e,i){s(t,y,a);var o;return w(r)?r instanceof X||(o=x(r))==K||o==J?void 0!==i?new g(r,At(e,n),i):void 0!==e?new g(r,At(e,n)):new g(r):_t in r?kt(y,r):Rt.call(y,r):new g(Mt(r,l))}),Z(b!==Function.prototype?P(g).concat(P(b)):P(g),function(t){t in y||h(y,t,g[t])}),y[H]=_,e||(_.constructor=y));var A=_[bt],N=!!A&&("values"==A.name||void 0==A.name),T=Vt.values;h(y,xt,!0),h(_,_t,a),h(_,Ot,!0),h(_,wt,y),(f?new y(1)[mt]==a:mt in _)||G(_,mt,{get:function(){return a}}),S[a]=y,u(u.G+u.W+u.F*(y!=g),S),u(u.S,a,{BYTES_PER_ELEMENT:n,from:Rt,of:Ct}),Y in _||h(_,Y,n),u(u.P,a,Wt),R(a),u(u.P+u.F*Ft,a,{set:Bt}),u(u.P+u.F*!N,a,Vt),u(u.P+u.F*(_.toString!=yt),a,{toString:yt}),u(u.P+u.F*o(function(){new y(1).slice()}),a,{slice:Gt}),u(u.P+u.F*(o(function(){return[1,2].toLocaleString()!=new y([1,2]).toLocaleString()})||!o(function(){_.toLocaleString.call([1,2])})),a,{toLocaleString:Ut}),k[a]=N?A:T,e||N||h(_,bt,T)}}else t.exports=function(){}},function(t,n){var r={}.toString;t.exports=function(t){return r.call(t).slice(8,-1)}},function(t,n,r){var e=r(21),i=r(5).document,o=e(i)&&e(i.createElement);t.exports=function(t){return o?i.createElement(t):{}}},function(t,n,r){t.exports=!r(12)&&!r(18)(function(){return 7!=Object.defineProperty(r(57)("div"),"a",{get:function(){return 7}}).a})},function(t,n,r){"use strict";var e=r(36),i=r(51),o=r(64),u=r(13),c=r(8),f=r(35),a=r(96),s=r(38),l=r(103),h=r(15)("iterator"),v=!([].keys&&"next"in[].keys()),p="keys",d="values",y=function(){return this};t.exports=function(t,n,r,g,b,m,x){a(r,n,g);var w,S,_,O=function(t){if(!v&&t in F)return F[t];switch(t){case p:case d:return function(){return new r(this,t)}}return function(){return new r(this,t)}},E=n+" Iterator",P=b==d,j=!1,F=t.prototype,M=F[h]||F["@@iterator"]||b&&F[b],A=M||O(b),N=b?P?O("entries"):A:void 0,T="Array"==n?F.entries||M:M;if(T&&(_=l(T.call(new t)))!==Object.prototype&&(s(_,E,!0),e||c(_,h)||u(_,h,y)),P&&M&&M.name!==d&&(j=!0,A=function(){return M.call(this)}),e&&!x||!v&&!j&&F[h]||u(F,h,A),f[n]=A,f[E]=y,b)if(w={values:P?A:O(d),keys:m?A:O(p),entries:N},x)for(S in w)S in F||o(F,S,w[S]);else i(i.P+i.F*(v||j),n,w);return w}},function(t,n,r){var e=r(20),i=r(100),o=r(34),u=r(39)("IE_PROTO"),c=function(){},f="prototype",a=function(){var t,n=r(57)("iframe"),e=o.length;for(n.style.display="none",r(93).appendChild(n),n.src="javascript:",t=n.contentWindow.document,t.open(),t.write("<script>document.F=Object<\/script>"),t.close(),a=t.F;e--;)delete a[f][o[e]];return a()};t.exports=Object.create||function(t,n){var r;return null!==t?(c[f]=e(t),r=new c,c[f]=null,r[u]=t):r=a(),void 0===n?r:i(r,n)}},function(t,n,r){var e=r(63),i=r(34).concat("length","prototype");n.f=Object.getOwnPropertyNames||function(t){return e(t,i)}},function(t,n){n.f=Object.getOwnPropertySymbols},function(t,n,r){var e=r(8),i=r(9),o=r(90)(!1),u=r(39)("IE_PROTO");t.exports=function(t,n){var r,c=i(t),f=0,a=[];for(r in c)r!=u&&e(c,r)&&a.push(r);for(;n.length>f;)e(c,r=n[f++])&&(~o(a,r)||a.push(r));return a}},function(t,n,r){t.exports=r(13)},function(t,n,r){var e=r(76)("meta"),i=r(6),o=r(24),u=r(11).f,c=0,f=Object.isExtensible||function(){return!0},a=!r(4)(function(){return f(Object.preventExtensions({}))}),s=function(t){u(t,e,{value:{i:"O"+ ++c,w:{}}})},l=function(t,n){if(!i(t))return"symbol"==typeof t?t:("string"==typeof t?"S":"P")+t;if(!o(t,e)){if(!f(t))return"F";if(!n)return"E";s(t)}return t[e].i},h=function(t,n){if(!o(t,e)){if(!f(t))return!0;if(!n)return!1;s(t)}return t[e].w},v=function(t){return a&&p.NEED&&f(t)&&!o(t,e)&&s(t),t},p=t.exports={KEY:e,NEED:!1,fastKey:l,getWeak:h,onFreeze:v}},function(t,n){t.exports=function(t,n){return{enumerable:!(1&t),configurable:!(2&t),writable:!(4&t),value:n}}},function(t,n){var r=Math.ceil,e=Math.floor;t.exports=function(t){return isNaN(t=+t)?0:(t>0?e:r)(t)}},function(t,n){t.exports=function(t,n,r,e){if(!(t instanceof n)||void 0!==e&&e in t)throw TypeError(r+": incorrect invocation!");return t}},function(t,n){t.exports=!1},function(t,n,r){var e=r(2),i=r(173),o=r(133),u=r(145)("IE_PROTO"),c=function(){},f="prototype",a=function(){var t,n=r(132)("iframe"),e=o.length;for(n.style.display="none",r(135).appendChild(n),n.src="javascript:",t=n.contentWindow.document,t.open(),t.write("<script>document.F=Object<\/script>"),t.close(),a=t.F;e--;)delete a[f][o[e]];return a()};t.exports=Object.create||function(t,n){var r;return null!==t?(c[f]=e(t),r=new c,c[f]=null,r[u]=t):r=a(),void 0===n?r:i(r,n)}},function(t,n,r){var e=r(175),i=r(133).concat("length","prototype");n.f=Object.getOwnPropertyNames||function(t){return e(t,i)}},function(t,n,r){var e=r(175),i=r(133);t.exports=Object.keys||function(t){return e(t,i)}},function(t,n,r){var e=r(28);t.exports=function(t,n,r){for(var i in n)e(t,i,n[i],r);return t}},function(t,n,r){"use strict";var e=r(3),i=r(11),o=r(10),u=r(7)("species");t.exports=function(t){var n=e[t];o&&n&&!n[u]&&i.f(n,u,{configurable:!0,get:function(){return this}})}},function(t,n,r){var e=r(67),i=Math.max,o=Math.min;t.exports=function(t,n){return t=e(t),t<0?i(t+n,0):o(t,n)}},function(t,n){var r=0,e=Math.random();t.exports=function(t){return"Symbol(".concat(void 0===t?"":t,")_",(++r+e).toString(36))}},function(t,n,r){var e=r(33);t.exports=function(t){return Object(e(t))}},function(t,n,r){var e=r(7)("unscopables"),i=Array.prototype;void 0==i[e]&&r(27)(i,e,{}),t.exports=function(t){i[e][t]=!0}},function(t,n,r){var e=r(53),i=r(169),o=r(137),u=r(2),c=r(16),f=r(154),a={},s={},n=t.exports=function(t,n,r,l,h){var v,p,d,y,g=h?function(){return t}:f(t),b=e(r,l,n?2:1),m=0;if("function"!=typeof g)throw TypeError(t+" is not iterable!");if(o(g)){for(v=c(t.length);v>m;m++)if((y=n?b(u(p=t[m])[0],p[1]):b(t[m]))===a||y===s)return y}else for(d=g.call(t);!(p=d.next()).done;)if((y=i(d,b,p.value,n))===a||y===s)return y};n.BREAK=a,n.RETURN=s},function(t,n){t.exports={}},function(t,n,r){var e=r(11).f,i=r(24),o=r(7)("toStringTag");t.exports=function(t,n,r){t&&!i(t=r?t:t.prototype,o)&&e(t,o,{configurable:!0,value:n})}},function(t,n,r){var e=r(1),i=r(46),o=r(4),u=r(150),c="["+u+"]",f="​",a=RegExp("^"+c+c+"*"),s=RegExp(c+c+"*$"),l=function(t,n,r){var i={},c=o(function(){return!!u[t]()||f[t]()!=f}),a=i[t]=c?n(h):u[t];r&&(i[r]=a),e(e.P+e.F*c,"String",i)},h=l.trim=function(t,n){return t=String(i(t)),1&n&&(t=t.replace(a,"")),2&n&&(t=t.replace(s,"")),t};t.exports=l},function(t,n,r){t.exports={default:r(86),__esModule:!0}},function(t,n,r){t.exports={default:r(87),__esModule:!0}},function(t,n,r){"use strict";function e(t){return t&&t.__esModule?t:{default:t}}n.__esModule=!0;var i=r(84),o=e(i),u=r(83),c=e(u),f="function"==typeof c.default&&"symbol"==typeof o.default?function(t){return typeof t}:function(t){return t&&"function"==typeof c.default&&t.constructor===c.default&&t!==c.default.prototype?"symbol":typeof t};n.default="function"==typeof c.default&&"symbol"===f(o.default)?function(t){return void 0===t?"undefined":f(t)}:function(t){return t&&"function"==typeof c.default&&t.constructor===c.default&&t!==c.default.prototype?"symbol":void 0===t?"undefined":f(t)}},function(t,n,r){r(110),r(108),r(111),r(112),t.exports=r(25).Symbol},function(t,n,r){r(109),r(113),t.exports=r(44).f("iterator")},function(t,n){t.exports=function(t){if("function"!=typeof t)throw TypeError(t+" is not a function!");return t}},function(t,n){t.exports=function(){}},function(t,n,r){var e=r(9),i=r(106),o=r(105);t.exports=function(t){return function(n,r,u){var c,f=e(n),a=i(f.length),s=o(u,a);if(t&&r!=r){for(;a>s;)if((c=f[s++])!=c)return!0}else for(;a>s;s++)if((t||s in f)&&f[s]===r)return t||s||0;return!t&&-1}}},function(t,n,r){var e=r(88);t.exports=function(t,n,r){if(e(t),void 0===n)return t;switch(r){case 1:return function(r){return t.call(n,r)};case 2:return function(r,e){return t.call(n,r,e)};case 3:return function(r,e,i){return t.call(n,r,e,i)}}return function(){return t.apply(n,arguments)}}},function(t,n,r){var e=r(19),i=r(62),o=r(37);t.exports=function(t){var n=e(t),r=i.f;if(r)for(var u,c=r(t),f=o.f,a=0;c.length>a;)f.call(t,u=c[a++])&&n.push(u);return n}},function(t,n,r){t.exports=r(5).document&&document.documentElement},function(t,n,r){var e=r(56);t.exports=Object("z").propertyIsEnumerable(0)?Object:function(t){return"String"==e(t)?t.split(""):Object(t)}},function(t,n,r){var e=r(56);t.exports=Array.isArray||function(t){return"Array"==e(t)}},function(t,n,r){"use strict";var e=r(60),i=r(22),o=r(38),u={};r(13)(u,r(15)("iterator"),function(){return this}),t.exports=function(t,n,r){t.prototype=e(u,{next:i(1,r)}),o(t,n+" Iterator")}},function(t,n){t.exports=function(t,n){return{value:n,done:!!t}}},function(t,n,r){var e=r(19),i=r(9);t.exports=function(t,n){for(var r,o=i(t),u=e(o),c=u.length,f=0;c>f;)if(o[r=u[f++]]===n)return r}},function(t,n,r){var e=r(23)("meta"),i=r(21),o=r(8),u=r(14).f,c=0,f=Object.isExtensible||function(){return!0},a=!r(18)(function(){return f(Object.preventExtensions({}))}),s=function(t){u(t,e,{value:{i:"O"+ ++c,w:{}}})},l=function(t,n){if(!i(t))return"symbol"==typeof t?t:("string"==typeof t?"S":"P")+t;if(!o(t,e)){if(!f(t))return"F";if(!n)return"E";s(t)}return t[e].i},h=function(t,n){if(!o(t,e)){if(!f(t))return!0;if(!n)return!1;s(t)}return t[e].w},v=function(t){return a&&p.NEED&&f(t)&&!o(t,e)&&s(t),t},p=t.exports={KEY:e,NEED:!1,fastKey:l,getWeak:h,onFreeze:v}},function(t,n,r){var e=r(14),i=r(20),o=r(19);t.exports=r(12)?Object.defineProperties:function(t,n){i(t);for(var r,u=o(n),c=u.length,f=0;c>f;)e.f(t,r=u[f++],n[r]);return t}},function(t,n,r){var e=r(37),i=r(22),o=r(9),u=r(42),c=r(8),f=r(58),a=Object.getOwnPropertyDescriptor;n.f=r(12)?a:function(t,n){if(t=o(t),n=u(n,!0),f)try{return a(t,n)}catch(t){}if(c(t,n))return i(!e.f.call(t,n),t[n])}},function(t,n,r){var e=r(9),i=r(61).f,o={}.toString,u="object"==typeof window&&window&&Object.getOwnPropertyNames?Object.getOwnPropertyNames(window):[],c=function(t){try{return i(t)}catch(t){return u.slice()}};t.exports.f=function(t){return u&&"[object Window]"==o.call(t)?c(t):i(e(t))}},function(t,n,r){var e=r(8),i=r(77),o=r(39)("IE_PROTO"),u=Object.prototype;t.exports=Object.getPrototypeOf||function(t){return t=i(t),e(t,o)?t[o]:"function"==typeof t.constructor&&t instanceof t.constructor?t.constructor.prototype:t instanceof Object?u:null}},function(t,n,r){var e=r(41),i=r(33);t.exports=function(t){return function(n,r){var o,u,c=String(i(n)),f=e(r),a=c.length;return f<0||f>=a?t?"":void 0:(o=c.charCodeAt(f),o<55296||o>56319||f+1===a||(u=c.charCodeAt(f+1))<56320||u>57343?t?c.charAt(f):o:t?c.slice(f,f+2):u-56320+(o-55296<<10)+65536)}}},function(t,n,r){var e=r(41),i=Math.max,o=Math.min;t.exports=function(t,n){return t=e(t),t<0?i(t+n,0):o(t,n)}},function(t,n,r){var e=r(41),i=Math.min;t.exports=function(t){return t>0?i(e(t),9007199254740991):0}},function(t,n,r){"use strict";var e=r(89),i=r(97),o=r(35),u=r(9);t.exports=r(59)(Array,"Array",function(t,n){this._t=u(t),this._i=0,this._k=n},function(){var t=this._t,n=this._k,r=this._i++;return!t||r>=t.length?(this._t=void 0,i(1)):"keys"==n?i(0,r):"values"==n?i(0,t[r]):i(0,[r,t[r]])},"values"),o.Arguments=o.Array,e("keys"),e("values"),e("entries")},function(t,n){},function(t,n,r){"use strict";var e=r(104)(!0);r(59)(String,"String",function(t){this._t=String(t),this._i=0},function(){var t,n=this._t,r=this._i;return r>=n.length?{value:void 0,done:!0}:(t=e(n,r),this._i+=t.length,{value:t,done:!1})})},function(t,n,r){"use strict";var e=r(5),i=r(8),o=r(12),u=r(51),c=r(64),f=r(99).KEY,a=r(18),s=r(40),l=r(38),h=r(23),v=r(15),p=r(44),d=r(43),y=r(98),g=r(92),b=r(95),m=r(20),x=r(9),w=r(42),S=r(22),_=r(60),O=r(102),E=r(101),P=r(14),j=r(19),F=E.f,M=P.f,A=O.f,N=e.Symbol,T=e.JSON,I=T&&T.stringify,k="prototype",L=v("_hidden"),R=v("toPrimitive"),C={}.propertyIsEnumerable,D=s("symbol-registry"),U=s("symbols"),W=s("op-symbols"),G=Object[k],B="function"==typeof N,V=e.QObject,z=!V||!V[k]||!V[k].findChild,q=o&&a(function(){return 7!=_(M({},"a",{get:function(){return M(this,"a",{value:7}).a}})).a})?function(t,n,r){var e=F(G,n);e&&delete G[n],M(t,n,r),e&&t!==G&&M(G,n,e)}:M,K=function(t){var n=U[t]=_(N[k]);return n._k=t,n},J=B&&"symbol"==typeof N.iterator?function(t){return"symbol"==typeof t}:function(t){return t instanceof N},Y=function(t,n,r){return t===G&&Y(W,n,r),m(t),n=w(n,!0),m(r),i(U,n)?(r.enumerable?(i(t,L)&&t[L][n]&&(t[L][n]=!1),r=_(r,{enumerable:S(0,!1)})):(i(t,L)||M(t,L,S(1,{})),t[L][n]=!0),q(t,n,r)):M(t,n,r)},H=function(t,n){m(t);for(var r,e=g(n=x(n)),i=0,o=e.length;o>i;)Y(t,r=e[i++],n[r]);return t},$=function(t,n){return void 0===n?_(t):H(_(t),n)},X=function(t){var n=C.call(this,t=w(t,!0));return!(this===G&&i(U,t)&&!i(W,t))&&(!(n||!i(this,t)||!i(U,t)||i(this,L)&&this[L][t])||n)},Q=function(t,n){if(t=x(t),n=w(n,!0),t!==G||!i(U,n)||i(W,n)){var r=F(t,n);return!r||!i(U,n)||i(t,L)&&t[L][n]||(r.enumerable=!0),r}},Z=function(t){for(var n,r=A(x(t)),e=[],o=0;r.length>o;)i(U,n=r[o++])||n==L||n==f||e.push(n);return e},tt=function(t){for(var n,r=t===G,e=A(r?W:x(t)),o=[],u=0;e.length>u;)!i(U,n=e[u++])||r&&!i(G,n)||o.push(U[n]);return o};B||(N=function(){if(this instanceof N)throw TypeError("Symbol is not a constructor!");var t=h(arguments.length>0?arguments[0]:void 0),n=function(r){this===G&&n.call(W,r),i(this,L)&&i(this[L],t)&&(this[L][t]=!1),q(this,t,S(1,r))};return o&&z&&q(G,t,{configurable:!0,set:n}),K(t)},c(N[k],"toString",function(){return this._k}),E.f=Q,P.f=Y,r(61).f=O.f=Z,r(37).f=X,r(62).f=tt,o&&!r(36)&&c(G,"propertyIsEnumerable",X,!0),p.f=function(t){return K(v(t))}),u(u.G+u.W+u.F*!B,{Symbol:N});for(var nt="hasInstance,isConcatSpreadable,iterator,match,replace,search,species,split,toPrimitive,toStringTag,unscopables".split(","),rt=0;nt.length>rt;)v(nt[rt++]);for(var nt=j(v.store),rt=0;nt.length>rt;)d(nt[rt++]);u(u.S+u.F*!B,"Symbol",{for:function(t){return i(D,t+="")?D[t]:D[t]=N(t)},keyFor:function(t){if(J(t))return y(D,t);throw TypeError(t+" is not a symbol!")},useSetter:function(){z=!0},useSimple:function(){z=!1}}),u(u.S+u.F*!B,"Object",{create:$,defineProperty:Y,defineProperties:H,getOwnPropertyDescriptor:Q,getOwnPropertyNames:Z,getOwnPropertySymbols:tt}),T&&u(u.S+u.F*(!B||a(function(){var t=N();return"[null]"!=I([t])||"{}"!=I({a:t})||"{}"!=I(Object(t))})),"JSON",{stringify:function(t){if(void 0!==t&&!J(t)){for(var n,r,e=[t],i=1;arguments.length>i;)e.push(arguments[i++]);return n=e[1],"function"==typeof n&&(r=n),!r&&b(n)||(n=function(t,n){if(r&&(n=r.call(this,t,n)),!J(n))return n}),e[1]=n,I.apply(T,e)}}}),N[k][R]||r(13)(N[k],R,N[k].valueOf),l(N,"Symbol"),l(Math,"Math",!0),l(e.JSON,"JSON",!0)},function(t,n,r){r(43)("asyncIterator")},function(t,n,r){r(43)("observable")},function(t,n,r){r(107);for(var e=r(5),i=r(13),o=r(35),u=r(15)("toStringTag"),c=["NodeList","DOMTokenList","MediaList","StyleSheetList","CSSRuleList"],f=0;f<5;f++){var a=c[f],s=e[a],l=s&&s.prototype;l&&!l[u]&&i(l,u,a),o[a]=o.Array}},function(t,n,r){var e=r(45),i=r(7)("toStringTag"),o="Arguments"==e(function(){return arguments}()),u=function(t,n){try{return t[n]}catch(t){}};t.exports=function(t){var n,r,c;return void 0===t?"Undefined":null===t?"Null":"string"==typeof(r=u(n=Object(t),i))?r:o?e(n):"Object"==(c=e(n))&&"function"==typeof n.callee?"Arguments":c}},function(t,n,r){var e=r(45);t.exports=Object("z").propertyIsEnumerable(0)?Object:function(t){return"String"==e(t)?t.split(""):Object(t)}},function(t,n){n.f={}.propertyIsEnumerable},function(t,n,r){var e=r(30),i=r(16),o=r(75);t.exports=function(t){return function(n,r,u){var c,f=e(n),a=i(f.length),s=o(u,a);if(t&&r!=r){for(;a>s;)if((c=f[s++])!=c)return!0}else for(;a>s;s++)if((t||s in f)&&f[s]===r)return t||s||0;return!t&&-1}}},function(t,n,r){"use strict";var e=r(3),i=r(1),o=r(28),u=r(73),c=r(65),f=r(79),a=r(68),s=r(6),l=r(4),h=r(123),v=r(81),p=r(136);t.exports=function(t,n,r,d,y,g){var b=e[t],m=b,x=y?"set":"add",w=m&&m.prototype,S={},_=function(t){var n=w[t];o(w,t,"delete"==t?function(t){return!(g&&!s(t))&&n.call(this,0===t?0:t)}:"has"==t?function(t){return!(g&&!s(t))&&n.call(this,0===t?0:t)}:"get"==t?function(t){return g&&!s(t)?void 0:n.call(this,0===t?0:t)}:"add"==t?function(t){return n.call(this,0===t?0:t),this}:function(t,r){return n.call(this,0===t?0:t,r),this})};if("function"==typeof m&&(g||w.forEach&&!l(function(){(new m).entries().next()}))){var O=new m,E=O[x](g?{}:-0,1)!=O,P=l(function(){O.has(1)}),j=h(function(t){new m(t)}),F=!g&&l(function(){for(var t=new m,n=5;n--;)t[x](n,n);return!t.has(-0)});j||(m=n(function(n,r){a(n,m,t);var e=p(new b,n,m);return void 0!=r&&f(r,y,e[x],e),e}),m.prototype=w,w.constructor=m),(P||F)&&(_("delete"),_("has"),y&&_("get")),(F||E)&&_(x),g&&w.clear&&delete w.clear}else m=d.getConstructor(n,t,y,x),u(m.prototype,r),c.NEED=!0;return v(m,t),S[t]=m,i(i.G+i.W+i.F*(m!=b),S),g||d.setStrong(m,t,y),m}},function(t,n,r){"use strict";var e=r(27),i=r(28),o=r(4),u=r(46),c=r(7);t.exports=function(t,n,r){var f=c(t),a=r(u,f,""[t]),s=a[0],l=a[1];o(function(){var n={};return n[f]=function(){return 7},7!=""[t](n)})&&(i(String.prototype,t,s),e(RegExp.prototype,f,2==n?function(t,n){return l.call(t,this,n)}:function(t){return l.call(t,this)}))}
},function(t,n,r){"use strict";var e=r(2);t.exports=function(){var t=e(this),n="";return t.global&&(n+="g"),t.ignoreCase&&(n+="i"),t.multiline&&(n+="m"),t.unicode&&(n+="u"),t.sticky&&(n+="y"),n}},function(t,n){t.exports=function(t,n,r){var e=void 0===r;switch(n.length){case 0:return e?t():t.call(r);case 1:return e?t(n[0]):t.call(r,n[0]);case 2:return e?t(n[0],n[1]):t.call(r,n[0],n[1]);case 3:return e?t(n[0],n[1],n[2]):t.call(r,n[0],n[1],n[2]);case 4:return e?t(n[0],n[1],n[2],n[3]):t.call(r,n[0],n[1],n[2],n[3])}return t.apply(r,n)}},function(t,n,r){var e=r(6),i=r(45),o=r(7)("match");t.exports=function(t){var n;return e(t)&&(void 0!==(n=t[o])?!!n:"RegExp"==i(t))}},function(t,n,r){var e=r(7)("iterator"),i=!1;try{var o=[7][e]();o.return=function(){i=!0},Array.from(o,function(){throw 2})}catch(t){}t.exports=function(t,n){if(!n&&!i)return!1;var r=!1;try{var o=[7],u=o[e]();u.next=function(){return{done:r=!0}},o[e]=function(){return u},t(o)}catch(t){}return r}},function(t,n,r){t.exports=r(69)||!r(4)(function(){var t=Math.random();__defineSetter__.call(null,t,function(){}),delete r(3)[t]})},function(t,n){n.f=Object.getOwnPropertySymbols},function(t,n,r){var e=r(3),i="__core-js_shared__",o=e[i]||(e[i]={});t.exports=function(t){return o[t]||(o[t]={})}},function(t,n,r){for(var e,i=r(3),o=r(27),u=r(76),c=u("typed_array"),f=u("view"),a=!(!i.ArrayBuffer||!i.DataView),s=a,l=0,h="Int8Array,Uint8Array,Uint8ClampedArray,Int16Array,Uint16Array,Int32Array,Uint32Array,Float32Array,Float64Array".split(",");l<9;)(e=i[h[l++]])?(o(e.prototype,c,!0),o(e.prototype,f,!0)):s=!1;t.exports={ABV:a,CONSTR:s,TYPED:c,VIEW:f}},function(t,n){"use strict";var r={versions:function(){var t=window.navigator.userAgent;return{trident:t.indexOf("Trident")>-1,presto:t.indexOf("Presto")>-1,webKit:t.indexOf("AppleWebKit")>-1,gecko:t.indexOf("Gecko")>-1&&-1==t.indexOf("KHTML"),mobile:!!t.match(/AppleWebKit.*Mobile.*/),ios:!!t.match(/\(i[^;]+;( U;)? CPU.+Mac OS X/),android:t.indexOf("Android")>-1||t.indexOf("Linux")>-1,iPhone:t.indexOf("iPhone")>-1||t.indexOf("Mac")>-1,iPad:t.indexOf("iPad")>-1,webApp:-1==t.indexOf("Safari"),weixin:-1==t.indexOf("MicroMessenger")}}()};t.exports=r},function(t,n,r){"use strict";var e=r(85),i=function(t){return t&&t.__esModule?t:{default:t}}(e),o=function(){function t(t,n,e){return n||e?String.fromCharCode(n||e):r[t]||t}function n(t){return e[t]}var r={"&quot;":'"',"&lt;":"<","&gt;":">","&amp;":"&","&nbsp;":" "},e={};for(var u in r)e[r[u]]=u;return r["&apos;"]="'",e["'"]="&#39;",{encode:function(t){return t?(""+t).replace(/['<> "&]/g,n).replace(/\r?\n/g,"<br/>").replace(/\s/g,"&nbsp;"):""},decode:function(n){return n?(""+n).replace(/<br\s*\/?>/gi,"\n").replace(/&quot;|&lt;|&gt;|&amp;|&nbsp;|&apos;|&#(\d+);|&#(\d+)/g,t).replace(/\u00a0/g," "):""},encodeBase16:function(t){if(!t)return t;t+="";for(var n=[],r=0,e=t.length;e>r;r++)n.push(t.charCodeAt(r).toString(16).toUpperCase());return n.join("")},encodeBase16forJSON:function(t){if(!t)return t;t=t.replace(/[\u4E00-\u9FBF]/gi,function(t){return escape(t).replace("%u","\\u")});for(var n=[],r=0,e=t.length;e>r;r++)n.push(t.charCodeAt(r).toString(16).toUpperCase());return n.join("")},decodeBase16:function(t){if(!t)return t;t+="";for(var n=[],r=0,e=t.length;e>r;r+=2)n.push(String.fromCharCode("0x"+t.slice(r,r+2)));return n.join("")},encodeObject:function(t){if(t instanceof Array)for(var n=0,r=t.length;r>n;n++)t[n]=o.encodeObject(t[n]);else if("object"==(void 0===t?"undefined":(0,i.default)(t)))for(var e in t)t[e]=o.encodeObject(t[e]);else if("string"==typeof t)return o.encode(t);return t},loadScript:function(t){var n=document.createElement("script");document.getElementsByTagName("body")[0].appendChild(n),n.setAttribute("src",t)},addLoadEvent:function(t){var n=window.onload;"function"!=typeof window.onload?window.onload=t:window.onload=function(){n(),t()}}}}();t.exports=o},function(t,n,r){"use strict";var e=r(17),i=r(75),o=r(16);t.exports=function(t){for(var n=e(this),r=o(n.length),u=arguments.length,c=i(u>1?arguments[1]:void 0,r),f=u>2?arguments[2]:void 0,a=void 0===f?r:i(f,r);a>c;)n[c++]=t;return n}},function(t,n,r){"use strict";var e=r(11),i=r(66);t.exports=function(t,n,r){n in t?e.f(t,n,i(0,r)):t[n]=r}},function(t,n,r){var e=r(6),i=r(3).document,o=e(i)&&e(i.createElement);t.exports=function(t){return o?i.createElement(t):{}}},function(t,n){t.exports="constructor,hasOwnProperty,isPrototypeOf,propertyIsEnumerable,toLocaleString,toString,valueOf".split(",")},function(t,n,r){var e=r(7)("match");t.exports=function(t){var n=/./;try{"/./"[t](n)}catch(r){try{return n[e]=!1,!"/./"[t](n)}catch(t){}}return!0}},function(t,n,r){t.exports=r(3).document&&document.documentElement},function(t,n,r){var e=r(6),i=r(144).set;t.exports=function(t,n,r){var o,u=n.constructor;return u!==r&&"function"==typeof u&&(o=u.prototype)!==r.prototype&&e(o)&&i&&i(t,o),t}},function(t,n,r){var e=r(80),i=r(7)("iterator"),o=Array.prototype;t.exports=function(t){return void 0!==t&&(e.Array===t||o[i]===t)}},function(t,n,r){var e=r(45);t.exports=Array.isArray||function(t){return"Array"==e(t)}},function(t,n,r){"use strict";var e=r(70),i=r(66),o=r(81),u={};r(27)(u,r(7)("iterator"),function(){return this}),t.exports=function(t,n,r){t.prototype=e(u,{next:i(1,r)}),o(t,n+" Iterator")}},function(t,n,r){"use strict";var e=r(69),i=r(1),o=r(28),u=r(27),c=r(24),f=r(80),a=r(139),s=r(81),l=r(32),h=r(7)("iterator"),v=!([].keys&&"next"in[].keys()),p="keys",d="values",y=function(){return this};t.exports=function(t,n,r,g,b,m,x){a(r,n,g);var w,S,_,O=function(t){if(!v&&t in F)return F[t];switch(t){case p:case d:return function(){return new r(this,t)}}return function(){return new r(this,t)}},E=n+" Iterator",P=b==d,j=!1,F=t.prototype,M=F[h]||F["@@iterator"]||b&&F[b],A=M||O(b),N=b?P?O("entries"):A:void 0,T="Array"==n?F.entries||M:M;if(T&&(_=l(T.call(new t)))!==Object.prototype&&(s(_,E,!0),e||c(_,h)||u(_,h,y)),P&&M&&M.name!==d&&(j=!0,A=function(){return M.call(this)}),e&&!x||!v&&!j&&F[h]||u(F,h,A),f[n]=A,f[E]=y,b)if(w={values:P?A:O(d),keys:m?A:O(p),entries:N},x)for(S in w)S in F||o(F,S,w[S]);else i(i.P+i.F*(v||j),n,w);return w}},function(t,n){var r=Math.expm1;t.exports=!r||r(10)>22025.465794806718||r(10)<22025.465794806718||-2e-17!=r(-2e-17)?function(t){return 0==(t=+t)?t:t>-1e-6&&t<1e-6?t+t*t/2:Math.exp(t)-1}:r},function(t,n){t.exports=Math.sign||function(t){return 0==(t=+t)||t!=t?t:t<0?-1:1}},function(t,n,r){var e=r(3),i=r(151).set,o=e.MutationObserver||e.WebKitMutationObserver,u=e.process,c=e.Promise,f="process"==r(45)(u);t.exports=function(){var t,n,r,a=function(){var e,i;for(f&&(e=u.domain)&&e.exit();t;){i=t.fn,t=t.next;try{i()}catch(e){throw t?r():n=void 0,e}}n=void 0,e&&e.enter()};if(f)r=function(){u.nextTick(a)};else if(o){var s=!0,l=document.createTextNode("");new o(a).observe(l,{characterData:!0}),r=function(){l.data=s=!s}}else if(c&&c.resolve){var h=c.resolve();r=function(){h.then(a)}}else r=function(){i.call(e,a)};return function(e){var i={fn:e,next:void 0};n&&(n.next=i),t||(t=i,r()),n=i}}},function(t,n,r){var e=r(6),i=r(2),o=function(t,n){if(i(t),!e(n)&&null!==n)throw TypeError(n+": can't set as prototype!")};t.exports={set:Object.setPrototypeOf||("__proto__"in{}?function(t,n,e){try{e=r(53)(Function.call,r(31).f(Object.prototype,"__proto__").set,2),e(t,[]),n=!(t instanceof Array)}catch(t){n=!0}return function(t,r){return o(t,r),n?t.__proto__=r:e(t,r),t}}({},!1):void 0),check:o}},function(t,n,r){var e=r(126)("keys"),i=r(76);t.exports=function(t){return e[t]||(e[t]=i(t))}},function(t,n,r){var e=r(2),i=r(26),o=r(7)("species");t.exports=function(t,n){var r,u=e(t).constructor;return void 0===u||void 0==(r=e(u)[o])?n:i(r)}},function(t,n,r){var e=r(67),i=r(46);t.exports=function(t){return function(n,r){var o,u,c=String(i(n)),f=e(r),a=c.length;return f<0||f>=a?t?"":void 0:(o=c.charCodeAt(f),o<55296||o>56319||f+1===a||(u=c.charCodeAt(f+1))<56320||u>57343?t?c.charAt(f):o:t?c.slice(f,f+2):u-56320+(o-55296<<10)+65536)}}},function(t,n,r){var e=r(122),i=r(46);t.exports=function(t,n,r){if(e(n))throw TypeError("String#"+r+" doesn't accept regex!");return String(i(t))}},function(t,n,r){"use strict";var e=r(67),i=r(46);t.exports=function(t){var n=String(i(this)),r="",o=e(t);if(o<0||o==1/0)throw RangeError("Count can't be negative");for(;o>0;(o>>>=1)&&(n+=n))1&o&&(r+=n);return r}},function(t,n){t.exports="\t\n\v\f\r   ᠎             　\u2028\u2029\ufeff"},function(t,n,r){var e,i,o,u=r(53),c=r(121),f=r(135),a=r(132),s=r(3),l=s.process,h=s.setImmediate,v=s.clearImmediate,p=s.MessageChannel,d=0,y={},g="onreadystatechange",b=function(){var t=+this;if(y.hasOwnProperty(t)){var n=y[t];delete y[t],n()}},m=function(t){b.call(t.data)};h&&v||(h=function(t){for(var n=[],r=1;arguments.length>r;)n.push(arguments[r++]);return y[++d]=function(){c("function"==typeof t?t:Function(t),n)},e(d),d},v=function(t){delete y[t]},"process"==r(45)(l)?e=function(t){l.nextTick(u(b,t,1))}:p?(i=new p,o=i.port2,i.port1.onmessage=m,e=u(o.postMessage,o,1)):s.addEventListener&&"function"==typeof postMessage&&!s.importScripts?(e=function(t){s.postMessage(t+"","*")},s.addEventListener("message",m,!1)):e=g in a("script")?function(t){f.appendChild(a("script"))[g]=function(){f.removeChild(this),b.call(t)}}:function(t){setTimeout(u(b,t,1),0)}),t.exports={set:h,clear:v}},function(t,n,r){"use strict";var e=r(3),i=r(10),o=r(69),u=r(127),c=r(27),f=r(73),a=r(4),s=r(68),l=r(67),h=r(16),v=r(71).f,p=r(11).f,d=r(130),y=r(81),g="ArrayBuffer",b="DataView",m="prototype",x="Wrong length!",w="Wrong index!",S=e[g],_=e[b],O=e.Math,E=e.RangeError,P=e.Infinity,j=S,F=O.abs,M=O.pow,A=O.floor,N=O.log,T=O.LN2,I="buffer",k="byteLength",L="byteOffset",R=i?"_b":I,C=i?"_l":k,D=i?"_o":L,U=function(t,n,r){var e,i,o,u=Array(r),c=8*r-n-1,f=(1<<c)-1,a=f>>1,s=23===n?M(2,-24)-M(2,-77):0,l=0,h=t<0||0===t&&1/t<0?1:0;for(t=F(t),t!=t||t===P?(i=t!=t?1:0,e=f):(e=A(N(t)/T),t*(o=M(2,-e))<1&&(e--,o*=2),t+=e+a>=1?s/o:s*M(2,1-a),t*o>=2&&(e++,o/=2),e+a>=f?(i=0,e=f):e+a>=1?(i=(t*o-1)*M(2,n),e+=a):(i=t*M(2,a-1)*M(2,n),e=0));n>=8;u[l++]=255&i,i/=256,n-=8);for(e=e<<n|i,c+=n;c>0;u[l++]=255&e,e/=256,c-=8);return u[--l]|=128*h,u},W=function(t,n,r){var e,i=8*r-n-1,o=(1<<i)-1,u=o>>1,c=i-7,f=r-1,a=t[f--],s=127&a;for(a>>=7;c>0;s=256*s+t[f],f--,c-=8);for(e=s&(1<<-c)-1,s>>=-c,c+=n;c>0;e=256*e+t[f],f--,c-=8);if(0===s)s=1-u;else{if(s===o)return e?NaN:a?-P:P;e+=M(2,n),s-=u}return(a?-1:1)*e*M(2,s-n)},G=function(t){return t[3]<<24|t[2]<<16|t[1]<<8|t[0]},B=function(t){return[255&t]},V=function(t){return[255&t,t>>8&255]},z=function(t){return[255&t,t>>8&255,t>>16&255,t>>24&255]},q=function(t){return U(t,52,8)},K=function(t){return U(t,23,4)},J=function(t,n,r){p(t[m],n,{get:function(){return this[r]}})},Y=function(t,n,r,e){var i=+r,o=l(i);if(i!=o||o<0||o+n>t[C])throw E(w);var u=t[R]._b,c=o+t[D],f=u.slice(c,c+n);return e?f:f.reverse()},H=function(t,n,r,e,i,o){var u=+r,c=l(u);if(u!=c||c<0||c+n>t[C])throw E(w);for(var f=t[R]._b,a=c+t[D],s=e(+i),h=0;h<n;h++)f[a+h]=s[o?h:n-h-1]},$=function(t,n){s(t,S,g);var r=+n,e=h(r);if(r!=e)throw E(x);return e};if(u.ABV){if(!a(function(){new S})||!a(function(){new S(.5)})){S=function(t){return new j($(this,t))};for(var X,Q=S[m]=j[m],Z=v(j),tt=0;Z.length>tt;)(X=Z[tt++])in S||c(S,X,j[X]);o||(Q.constructor=S)}var nt=new _(new S(2)),rt=_[m].setInt8;nt.setInt8(0,2147483648),nt.setInt8(1,2147483649),!nt.getInt8(0)&&nt.getInt8(1)||f(_[m],{setInt8:function(t,n){rt.call(this,t,n<<24>>24)},setUint8:function(t,n){rt.call(this,t,n<<24>>24)}},!0)}else S=function(t){var n=$(this,t);this._b=d.call(Array(n),0),this[C]=n},_=function(t,n,r){s(this,_,b),s(t,S,b);var e=t[C],i=l(n);if(i<0||i>e)throw E("Wrong offset!");if(r=void 0===r?e-i:h(r),i+r>e)throw E(x);this[R]=t,this[D]=i,this[C]=r},i&&(J(S,k,"_l"),J(_,I,"_b"),J(_,k,"_l"),J(_,L,"_o")),f(_[m],{getInt8:function(t){return Y(this,1,t)[0]<<24>>24},getUint8:function(t){return Y(this,1,t)[0]},getInt16:function(t){var n=Y(this,2,t,arguments[1]);return(n[1]<<8|n[0])<<16>>16},getUint16:function(t){var n=Y(this,2,t,arguments[1]);return n[1]<<8|n[0]},getInt32:function(t){return G(Y(this,4,t,arguments[1]))},getUint32:function(t){return G(Y(this,4,t,arguments[1]))>>>0},getFloat32:function(t){return W(Y(this,4,t,arguments[1]),23,4)},getFloat64:function(t){return W(Y(this,8,t,arguments[1]),52,8)},setInt8:function(t,n){H(this,1,t,B,n)},setUint8:function(t,n){H(this,1,t,B,n)},setInt16:function(t,n){H(this,2,t,V,n,arguments[2])},setUint16:function(t,n){H(this,2,t,V,n,arguments[2])},setInt32:function(t,n){H(this,4,t,z,n,arguments[2])},setUint32:function(t,n){H(this,4,t,z,n,arguments[2])},setFloat32:function(t,n){H(this,4,t,K,n,arguments[2])},setFloat64:function(t,n){H(this,8,t,q,n,arguments[2])}});y(S,g),y(_,b),c(_[m],u.VIEW,!0),n[g]=S,n[b]=_},function(t,n,r){var e=r(3),i=r(52),o=r(69),u=r(182),c=r(11).f;t.exports=function(t){var n=i.Symbol||(i.Symbol=o?{}:e.Symbol||{});"_"==t.charAt(0)||t in n||c(n,t,{value:u.f(t)})}},function(t,n,r){var e=r(114),i=r(7)("iterator"),o=r(80);t.exports=r(52).getIteratorMethod=function(t){if(void 0!=t)return t[i]||t["@@iterator"]||o[e(t)]}},function(t,n,r){"use strict";var e=r(78),i=r(170),o=r(80),u=r(30);t.exports=r(140)(Array,"Array",function(t,n){this._t=u(t),this._i=0,this._k=n},function(){var t=this._t,n=this._k,r=this._i++;return!t||r>=t.length?(this._t=void 0,i(1)):"keys"==n?i(0,r):"values"==n?i(0,t[r]):i(0,[r,t[r]])},"values"),o.Arguments=o.Array,e("keys"),e("values"),e("entries")},function(t,n){function r(t,n){t.classList?t.classList.add(n):t.className+=" "+n}t.exports=r},function(t,n){function r(t,n){if(t.classList)t.classList.remove(n);else{var r=new RegExp("(^|\\b)"+n.split(" ").join("|")+"(\\b|$)","gi");t.className=t.className.replace(r," ")}}t.exports=r},function(t,n){function r(){throw new Error("setTimeout has not been defined")}function e(){throw new Error("clearTimeout has not been defined")}function i(t){if(s===setTimeout)return setTimeout(t,0);if((s===r||!s)&&setTimeout)return s=setTimeout,setTimeout(t,0);try{return s(t,0)}catch(n){try{return s.call(null,t,0)}catch(n){return s.call(this,t,0)}}}function o(t){if(l===clearTimeout)return clearTimeout(t);if((l===e||!l)&&clearTimeout)return l=clearTimeout,clearTimeout(t);try{return l(t)}catch(n){try{return l.call(null,t)}catch(n){return l.call(this,t)}}}function u(){d&&v&&(d=!1,v.length?p=v.concat(p):y=-1,p.length&&c())}function c(){if(!d){var t=i(u);d=!0;for(var n=p.length;n;){for(v=p,p=[];++y<n;)v&&v[y].run();y=-1,n=p.length}v=null,d=!1,o(t)}}function f(t,n){this.fun=t,this.array=n}function a(){}var s,l,h=t.exports={};!function(){try{s="function"==typeof setTimeout?setTimeout:r}catch(t){s=r}try{l="function"==typeof clearTimeout?clearTimeout:e}catch(t){l=e}}();var v,p=[],d=!1,y=-1;h.nextTick=function(t){var n=new Array(arguments.length-1);if(arguments.length>1)for(var r=1;r<arguments.length;r++)n[r-1]=arguments[r];p.push(new f(t,n)),1!==p.length||d||i(c)},f.prototype.run=function(){this.fun.apply(null,this.array)},h.title="browser",h.browser=!0,h.env={},h.argv=[],h.version="",h.versions={},h.on=a,h.addListener=a,h.once=a,h.off=a,h.removeListener=a,h.removeAllListeners=a,h.emit=a,h.prependListener=a,h.prependOnceListener=a,h.listeners=function(t){return[]},h.binding=function(t){throw new Error("process.binding is not supported")},h.cwd=function(){return"/"},h.chdir=function(t){throw new Error("process.chdir is not supported")},h.umask=function(){return 0}},function(t,n,r){var e=r(45);t.exports=function(t,n){if("number"!=typeof t&&"Number"!=e(t))throw TypeError(n);return+t}},function(t,n,r){"use strict";var e=r(17),i=r(75),o=r(16);t.exports=[].copyWithin||function(t,n){var r=e(this),u=o(r.length),c=i(t,u),f=i(n,u),a=arguments.length>2?arguments[2]:void 0,s=Math.min((void 0===a?u:i(a,u))-f,u-c),l=1;for(f<c&&c<f+s&&(l=-1,f+=s-1,c+=s-1);s-- >0;)f in r?r[c]=r[f]:delete r[c],c+=l,f+=l;return r}},function(t,n,r){var e=r(79);t.exports=function(t,n){var r=[];return e(t,!1,r.push,r,n),r}},function(t,n,r){var e=r(26),i=r(17),o=r(115),u=r(16);t.exports=function(t,n,r,c,f){e(n);var a=i(t),s=o(a),l=u(a.length),h=f?l-1:0,v=f?-1:1;if(r<2)for(;;){if(h in s){c=s[h],h+=v;break}if(h+=v,f?h<0:l<=h)throw TypeError("Reduce of empty array with no initial value")}for(;f?h>=0:l>h;h+=v)h in s&&(c=n(c,s[h],h,a));return c}},function(t,n,r){"use strict";var e=r(26),i=r(6),o=r(121),u=[].slice,c={},f=function(t,n,r){if(!(n in c)){for(var e=[],i=0;i<n;i++)e[i]="a["+i+"]";c[n]=Function("F,a","return new F("+e.join(",")+")")}return c[n](t,r)};t.exports=Function.bind||function(t){var n=e(this),r=u.call(arguments,1),c=function(){var e=r.concat(u.call(arguments));return this instanceof c?f(n,e.length,e):o(n,e,t)};return i(n.prototype)&&(c.prototype=n.prototype),c}},function(t,n,r){"use strict";var e=r(11).f,i=r(70),o=r(73),u=r(53),c=r(68),f=r(46),a=r(79),s=r(140),l=r(170),h=r(74),v=r(10),p=r(65).fastKey,d=v?"_s":"size",y=function(t,n){var r,e=p(n);if("F"!==e)return t._i[e];for(r=t._f;r;r=r.n)if(r.k==n)return r};t.exports={getConstructor:function(t,n,r,s){var l=t(function(t,e){c(t,l,n,"_i"),t._i=i(null),t._f=void 0,t._l=void 0,t[d]=0,void 0!=e&&a(e,r,t[s],t)});return o(l.prototype,{clear:function(){for(var t=this,n=t._i,r=t._f;r;r=r.n)r.r=!0,r.p&&(r.p=r.p.n=void 0),delete n[r.i];t._f=t._l=void 0,t[d]=0},delete:function(t){var n=this,r=y(n,t);if(r){var e=r.n,i=r.p;delete n._i[r.i],r.r=!0,i&&(i.n=e),e&&(e.p=i),n._f==r&&(n._f=e),n._l==r&&(n._l=i),n[d]--}return!!r},forEach:function(t){c(this,l,"forEach");for(var n,r=u(t,arguments.length>1?arguments[1]:void 0,3);n=n?n.n:this._f;)for(r(n.v,n.k,this);n&&n.r;)n=n.p},has:function(t){return!!y(this,t)}}),v&&e(l.prototype,"size",{get:function(){return f(this[d])}}),l},def:function(t,n,r){var e,i,o=y(t,n);return o?o.v=r:(t._l=o={i:i=p(n,!0),k:n,v:r,p:e=t._l,n:void 0,r:!1},t._f||(t._f=o),e&&(e.n=o),t[d]++,"F"!==i&&(t._i[i]=o)),t},getEntry:y,setStrong:function(t,n,r){s(t,n,function(t,n){this._t=t,this._k=n,this._l=void 0},function(){for(var t=this,n=t._k,r=t._l;r&&r.r;)r=r.p;return t._t&&(t._l=r=r?r.n:t._t._f)?"keys"==n?l(0,r.k):"values"==n?l(0,r.v):l(0,[r.k,r.v]):(t._t=void 0,l(1))},r?"entries":"values",!r,!0),h(n)}}},function(t,n,r){var e=r(114),i=r(161);t.exports=function(t){return function(){if(e(this)!=t)throw TypeError(t+"#toJSON isn't generic");return i(this)}}},function(t,n,r){"use strict";var e=r(73),i=r(65).getWeak,o=r(2),u=r(6),c=r(68),f=r(79),a=r(48),s=r(24),l=a(5),h=a(6),v=0,p=function(t){return t._l||(t._l=new d)},d=function(){this.a=[]},y=function(t,n){return l(t.a,function(t){return t[0]===n})};d.prototype={get:function(t){var n=y(this,t);if(n)return n[1]},has:function(t){return!!y(this,t)},set:function(t,n){var r=y(this,t);r?r[1]=n:this.a.push([t,n])},delete:function(t){var n=h(this.a,function(n){return n[0]===t});return~n&&this.a.splice(n,1),!!~n}},t.exports={getConstructor:function(t,n,r,o){var a=t(function(t,e){c(t,a,n,"_i"),t._i=v++,t._l=void 0,void 0!=e&&f(e,r,t[o],t)});return e(a.prototype,{delete:function(t){if(!u(t))return!1;var n=i(t);return!0===n?p(this).delete(t):n&&s(n,this._i)&&delete n[this._i]},has:function(t){if(!u(t))return!1;var n=i(t);return!0===n?p(this).has(t):n&&s(n,this._i)}}),a},def:function(t,n,r){var e=i(o(n),!0);return!0===e?p(t).set(n,r):e[t._i]=r,t},ufstore:p}},function(t,n,r){t.exports=!r(10)&&!r(4)(function(){return 7!=Object.defineProperty(r(132)("div"),"a",{get:function(){return 7}}).a})},function(t,n,r){var e=r(6),i=Math.floor;t.exports=function(t){return!e(t)&&isFinite(t)&&i(t)===t}},function(t,n,r){var e=r(2);t.exports=function(t,n,r,i){try{return i?n(e(r)[0],r[1]):n(r)}catch(n){var o=t.return;throw void 0!==o&&e(o.call(t)),n}}},function(t,n){t.exports=function(t,n){return{value:n,done:!!t}}},function(t,n){t.exports=Math.log1p||function(t){return(t=+t)>-1e-8&&t<1e-8?t-t*t/2:Math.log(1+t)}},function(t,n,r){"use strict";var e=r(72),i=r(125),o=r(116),u=r(17),c=r(115),f=Object.assign;t.exports=!f||r(4)(function(){var t={},n={},r=Symbol(),e="abcdefghijklmnopqrst";return t[r]=7,e.split("").forEach(function(t){n[t]=t}),7!=f({},t)[r]||Object.keys(f({},n)).join("")!=e})?function(t,n){for(var r=u(t),f=arguments.length,a=1,s=i.f,l=o.f;f>a;)for(var h,v=c(arguments[a++]),p=s?e(v).concat(s(v)):e(v),d=p.length,y=0;d>y;)l.call(v,h=p[y++])&&(r[h]=v[h]);return r}:f},function(t,n,r){var e=r(11),i=r(2),o=r(72);t.exports=r(10)?Object.defineProperties:function(t,n){i(t);for(var r,u=o(n),c=u.length,f=0;c>f;)e.f(t,r=u[f++],n[r]);return t}},function(t,n,r){var e=r(30),i=r(71).f,o={}.toString,u="object"==typeof window&&window&&Object.getOwnPropertyNames?Object.getOwnPropertyNames(window):[],c=function(t){try{return i(t)}catch(t){return u.slice()}};t.exports.f=function(t){return u&&"[object Window]"==o.call(t)?c(t):i(e(t))}},function(t,n,r){var e=r(24),i=r(30),o=r(117)(!1),u=r(145)("IE_PROTO");t.exports=function(t,n){var r,c=i(t),f=0,a=[];for(r in c)r!=u&&e(c,r)&&a.push(r);for(;n.length>f;)e(c,r=n[f++])&&(~o(a,r)||a.push(r));return a}},function(t,n,r){var e=r(72),i=r(30),o=r(116).f;t.exports=function(t){return function(n){for(var r,u=i(n),c=e(u),f=c.length,a=0,s=[];f>a;)o.call(u,r=c[a++])&&s.push(t?[r,u[r]]:u[r]);return s}}},function(t,n,r){var e=r(71),i=r(125),o=r(2),u=r(3).Reflect;t.exports=u&&u.ownKeys||function(t){var n=e.f(o(t)),r=i.f;return r?n.concat(r(t)):n}},function(t,n,r){var e=r(3).parseFloat,i=r(82).trim;t.exports=1/e(r(150)+"-0")!=-1/0?function(t){var n=i(String(t),3),r=e(n);return 0===r&&"-"==n.charAt(0)?-0:r}:e},function(t,n,r){var e=r(3).parseInt,i=r(82).trim,o=r(150),u=/^[\-+]?0[xX]/;t.exports=8!==e(o+"08")||22!==e(o+"0x16")?function(t,n){var r=i(String(t),3);return e(r,n>>>0||(u.test(r)?16:10))}:e},function(t,n){t.exports=Object.is||function(t,n){return t===n?0!==t||1/t==1/n:t!=t&&n!=n}},function(t,n,r){var e=r(16),i=r(149),o=r(46);t.exports=function(t,n,r,u){var c=String(o(t)),f=c.length,a=void 0===r?" ":String(r),s=e(n);if(s<=f||""==a)return c;var l=s-f,h=i.call(a,Math.ceil(l/a.length));return h.length>l&&(h=h.slice(0,l)),u?h+c:c+h}},function(t,n,r){n.f=r(7)},function(t,n,r){"use strict";var e=r(164);t.exports=r(118)("Map",function(t){return function(){return t(this,arguments.length>0?arguments[0]:void 0)}},{get:function(t){var n=e.getEntry(this,t);return n&&n.v},set:function(t,n){return e.def(this,0===t?0:t,n)}},e,!0)},function(t,n,r){r(10)&&"g"!=/./g.flags&&r(11).f(RegExp.prototype,"flags",{configurable:!0,get:r(120)})},function(t,n,r){"use strict";var e=r(164);t.exports=r(118)("Set",function(t){return function(){return t(this,arguments.length>0?arguments[0]:void 0)}},{add:function(t){return e.def(this,t=0===t?0:t,t)}},e)},function(t,n,r){"use strict";var e,i=r(48)(0),o=r(28),u=r(65),c=r(172),f=r(166),a=r(6),s=u.getWeak,l=Object.isExtensible,h=f.ufstore,v={},p=function(t){return function(){return t(this,arguments.length>0?arguments[0]:void 0)}},d={get:function(t){if(a(t)){var n=s(t);return!0===n?h(this).get(t):n?n[this._i]:void 0}},set:function(t,n){return f.def(this,t,n)}},y=t.exports=r(118)("WeakMap",p,d,f,!0,!0);7!=(new y).set((Object.freeze||Object)(v),7).get(v)&&(e=f.getConstructor(p),c(e.prototype,d),u.NEED=!0,i(["delete","has","get","set"],function(t){var n=y.prototype,r=n[t];o(n,t,function(n,i){if(a(n)&&!l(n)){this._f||(this._f=new e);var o=this._f[t](n,i);return"set"==t?this:o}return r.call(this,n,i)})}))},,,,function(t,n){"use strict";function r(){var t=document.querySelector("#page-nav");if(t&&!document.querySelector("#page-nav .extend.prev")&&(t.innerHTML='<a class="extend prev disabled" rel="prev">&laquo; Prev</a>'+t.innerHTML),t&&!document.querySelector("#page-nav .extend.next")&&(t.innerHTML=t.innerHTML+'<a class="extend next disabled" rel="next">Next &raquo;</a>'),yiliaConfig&&yiliaConfig.open_in_new){document.querySelectorAll(".article-entry a:not(.article-more-a)").forEach(function(t){var n=t.getAttribute("target");n&&""!==n||t.setAttribute("target","_blank")})}if(yiliaConfig&&yiliaConfig.toc_hide_index){document.querySelectorAll(".toc-number").forEach(function(t){t.style.display="none"})}var n=document.querySelector("#js-aboutme");n&&0!==n.length&&(n.innerHTML=n.innerText)}t.exports={init:r}},function(t,n,r){"use strict";function e(t){return t&&t.__esModule?t:{default:t}}function i(t,n){var r=/\/|index.html/g;return t.replace(r,"")===n.replace(r,"")}function o(){for(var t=document.querySelectorAll(".js-header-menu li a"),n=window.location.pathname,r=0,e=t.length;r<e;r++){var o=t[r];i(n,o.getAttribute("href"))&&(0,h.default)(o,"active")}}function u(t){for(var n=t.offsetLeft,r=t.offsetParent;null!==r;)n+=r.offsetLeft,r=r.offsetParent;return n}function c(t){for(var n=t.offsetTop,r=t.offsetParent;null!==r;)n+=r.offsetTop,r=r.offsetParent;return n}function f(t,n,r,e,i){var o=u(t),f=c(t)-n;if(f-r<=i){var a=t.$newDom;a||(a=t.cloneNode(!0),(0,d.default)(t,a),t.$newDom=a,a.style.position="fixed",a.style.top=(r||f)+"px",a.style.left=o+"px",a.style.zIndex=e||2,a.style.width="100%",a.style.color="#fff"),a.style.visibility="visible",t.style.visibility="hidden"}else{t.style.visibility="visible";var s=t.$newDom;s&&(s.style.visibility="hidden")}}function a(){var t=document.querySelector(".js-overlay"),n=document.querySelector(".js-header-menu");f(t,document.body.scrollTop,-63,2,0),f(n,document.body.scrollTop,1,3,0)}function s(){document.querySelector("#container").addEventListener("scroll",function(t){a()}),window.addEventListener("scroll",function(t){a()}),a()}var l=r(156),h=e(l),v=r(157),p=(e(v),r(382)),d=e(p),y=r(128),g=e(y),b=r(190),m=e(b),x=r(129);(function(){g.default.versions.mobile&&window.screen.width<800&&(o(),s())})(),(0,x.addLoadEvent)(function(){m.default.init()}),t.exports={}},,,,function(t,n,r){(function(t){"use strict";function n(t,n,r){t[n]||Object[e](t,n,{writable:!0,configurable:!0,value:r})}if(r(381),r(391),r(198),t._babelPolyfill)throw new Error("only one instance of babel-polyfill is allowed");t._babelPolyfill=!0;var e="defineProperty";n(String.prototype,"padLeft","".padStart),n(String.prototype,"padRight","".padEnd),"pop,reverse,shift,keys,values,entries,indexOf,every,some,forEach,map,filter,find,findIndex,includes,join,slice,concat,push,splice,unshift,sort,lastIndexOf,reduce,reduceRight,copyWithin,fill".split(",").forEach(function(t){[][t]&&n(Array,t,Function.call.bind([][t]))})}).call(n,function(){return this}())},,,function(t,n,r){r(210),t.exports=r(52).RegExp.escape},,,,function(t,n,r){var e=r(6),i=r(138),o=r(7)("species");t.exports=function(t){var n;return i(t)&&(n=t.constructor,"function"!=typeof n||n!==Array&&!i(n.prototype)||(n=void 0),e(n)&&null===(n=n[o])&&(n=void 0)),void 0===n?Array:n}},function(t,n,r){var e=r(202);t.exports=function(t,n){return new(e(t))(n)}},function(t,n,r){"use strict";var e=r(2),i=r(50),o="number";t.exports=function(t){if("string"!==t&&t!==o&&"default"!==t)throw TypeError("Incorrect hint");return i(e(this),t!=o)}},function(t,n,r){var e=r(72),i=r(125),o=r(116);t.exports=function(t){var n=e(t),r=i.f;if(r)for(var u,c=r(t),f=o.f,a=0;c.length>a;)f.call(t,u=c[a++])&&n.push(u);return n}},function(t,n,r){var e=r(72),i=r(30);t.exports=function(t,n){for(var r,o=i(t),u=e(o),c=u.length,f=0;c>f;)if(o[r=u[f++]]===n)return r}},function(t,n,r){"use strict";var e=r(208),i=r(121),o=r(26);t.exports=function(){for(var t=o(this),n=arguments.length,r=Array(n),u=0,c=e._,f=!1;n>u;)(r[u]=arguments[u++])===c&&(f=!0);return function(){var e,o=this,u=arguments.length,a=0,s=0;if(!f&&!u)return i(t,r,o);if(e=r.slice(),f)for(;n>a;a++)e[a]===c&&(e[a]=arguments[s++]);for(;u>s;)e.push(arguments[s++]);return i(t,e,o)}}},function(t,n,r){t.exports=r(3)},function(t,n){t.exports=function(t,n){var r=n===Object(n)?function(t){return n[t]}:n;return function(n){return String(n).replace(t,r)}}},function(t,n,r){var e=r(1),i=r(209)(/[\\^$*+?.()|[\]{}]/g,"\\$&");e(e.S,"RegExp",{escape:function(t){return i(t)}})},function(t,n,r){var e=r(1);e(e.P,"Array",{copyWithin:r(160)}),r(78)("copyWithin")},function(t,n,r){"use strict";var e=r(1),i=r(48)(4);e(e.P+e.F*!r(47)([].every,!0),"Array",{every:function(t){return i(this,t,arguments[1])}})},function(t,n,r){var e=r(1);e(e.P,"Array",{fill:r(130)}),r(78)("fill")},function(t,n,r){"use strict";var e=r(1),i=r(48)(2);e(e.P+e.F*!r(47)([].filter,!0),"Array",{filter:function(t){return i(this,t,arguments[1])}})},function(t,n,r){"use strict";var e=r(1),i=r(48)(6),o="findIndex",u=!0;o in[]&&Array(1)[o](function(){u=!1}),e(e.P+e.F*u,"Array",{findIndex:function(t){return i(this,t,arguments.length>1?arguments[1]:void 0)}}),r(78)(o)},function(t,n,r){"use strict";var e=r(1),i=r(48)(5),o="find",u=!0;o in[]&&Array(1)[o](function(){u=!1}),e(e.P+e.F*u,"Array",{find:function(t){return i(this,t,arguments.length>1?arguments[1]:void 0)}}),r(78)(o)},function(t,n,r){"use strict";var e=r(1),i=r(48)(0),o=r(47)([].forEach,!0);e(e.P+e.F*!o,"Array",{forEach:function(t){return i(this,t,arguments[1])}})},function(t,n,r){"use strict";var e=r(53),i=r(1),o=r(17),u=r(169),c=r(137),f=r(16),a=r(131),s=r(154);i(i.S+i.F*!r(123)(function(t){Array.from(t)}),"Array",{from:function(t){var n,r,i,l,h=o(t),v="function"==typeof this?this:Array,p=arguments.length,d=p>1?arguments[1]:void 0,y=void 0!==d,g=0,b=s(h);if(y&&(d=e(d,p>2?arguments[2]:void 0,2)),void 0==b||v==Array&&c(b))for(n=f(h.length),r=new v(n);n>g;g++)a(r,g,y?d(h[g],g):h[g]);else for(l=b.call(h),r=new v;!(i=l.next()).done;g++)a(r,g,y?u(l,d,[i.value,g],!0):i.value);return r.length=g,r}})},function(t,n,r){"use strict";var e=r(1),i=r(117)(!1),o=[].indexOf,u=!!o&&1/[1].indexOf(1,-0)<0;e(e.P+e.F*(u||!r(47)(o)),"Array",{indexOf:function(t){return u?o.apply(this,arguments)||0:i(this,t,arguments[1])}})},function(t,n,r){var e=r(1);e(e.S,"Array",{isArray:r(138)})},function(t,n,r){"use strict";var e=r(1),i=r(30),o=[].join;e(e.P+e.F*(r(115)!=Object||!r(47)(o)),"Array",{join:function(t){return o.call(i(this),void 0===t?",":t)}})},function(t,n,r){"use strict";var e=r(1),i=r(30),o=r(67),u=r(16),c=[].lastIndexOf,f=!!c&&1/[1].lastIndexOf(1,-0)<0;e(e.P+e.F*(f||!r(47)(c)),"Array",{lastIndexOf:function(t){if(f)return c.apply(this,arguments)||0;var n=i(this),r=u(n.length),e=r-1;for(arguments.length>1&&(e=Math.min(e,o(arguments[1]))),e<0&&(e=r+e);e>=0;e--)if(e in n&&n[e]===t)return e||0;return-1}})},function(t,n,r){"use strict";var e=r(1),i=r(48)(1);e(e.P+e.F*!r(47)([].map,!0),"Array",{map:function(t){return i(this,t,arguments[1])}})},function(t,n,r){"use strict";var e=r(1),i=r(131);e(e.S+e.F*r(4)(function(){function t(){}return!(Array.of.call(t)instanceof t)}),"Array",{of:function(){for(var t=0,n=arguments.length,r=new("function"==typeof this?this:Array)(n);n>t;)i(r,t,arguments[t++]);return r.length=n,r}})},function(t,n,r){"use strict";var e=r(1),i=r(162);e(e.P+e.F*!r(47)([].reduceRight,!0),"Array",{reduceRight:function(t){return i(this,t,arguments.length,arguments[1],!0)}})},function(t,n,r){"use strict";var e=r(1),i=r(162);e(e.P+e.F*!r(47)([].reduce,!0),"Array",{reduce:function(t){return i(this,t,arguments.length,arguments[1],!1)}})},function(t,n,r){"use strict";var e=r(1),i=r(135),o=r(45),u=r(75),c=r(16),f=[].slice;e(e.P+e.F*r(4)(function(){i&&f.call(i)}),"Array",{slice:function(t,n){var r=c(this.length),e=o(this);if(n=void 0===n?r:n,"Array"==e)return f.call(this,t,n);for(var i=u(t,r),a=u(n,r),s=c(a-i),l=Array(s),h=0;h<s;h++)l[h]="String"==e?this.charAt(i+h):this[i+h];return l}})},function(t,n,r){"use strict";var e=r(1),i=r(48)(3);e(e.P+e.F*!r(47)([].some,!0),"Array",{some:function(t){return i(this,t,arguments[1])}})},function(t,n,r){"use strict";var e=r(1),i=r(26),o=r(17),u=r(4),c=[].sort,f=[1,2,3];e(e.P+e.F*(u(function(){f.sort(void 0)})||!u(function(){f.sort(null)})||!r(47)(c)),"Array",{sort:function(t){return void 0===t?c.call(o(this)):c.call(o(this),i(t))}})},function(t,n,r){r(74)("Array")},function(t,n,r){var e=r(1);e(e.S,"Date",{now:function(){return(new Date).getTime()}})},function(t,n,r){"use strict";var e=r(1),i=r(4),o=Date.prototype.getTime,u=function(t){return t>9?t:"0"+t};e(e.P+e.F*(i(function(){return"0385-07-25T07:06:39.999Z"!=new Date(-5e13-1).toISOString()})||!i(function(){new Date(NaN).toISOString()})),"Date",{toISOString:function(){
if(!isFinite(o.call(this)))throw RangeError("Invalid time value");var t=this,n=t.getUTCFullYear(),r=t.getUTCMilliseconds(),e=n<0?"-":n>9999?"+":"";return e+("00000"+Math.abs(n)).slice(e?-6:-4)+"-"+u(t.getUTCMonth()+1)+"-"+u(t.getUTCDate())+"T"+u(t.getUTCHours())+":"+u(t.getUTCMinutes())+":"+u(t.getUTCSeconds())+"."+(r>99?r:"0"+u(r))+"Z"}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(50);e(e.P+e.F*r(4)(function(){return null!==new Date(NaN).toJSON()||1!==Date.prototype.toJSON.call({toISOString:function(){return 1}})}),"Date",{toJSON:function(t){var n=i(this),r=o(n);return"number"!=typeof r||isFinite(r)?n.toISOString():null}})},function(t,n,r){var e=r(7)("toPrimitive"),i=Date.prototype;e in i||r(27)(i,e,r(204))},function(t,n,r){var e=Date.prototype,i="Invalid Date",o="toString",u=e[o],c=e.getTime;new Date(NaN)+""!=i&&r(28)(e,o,function(){var t=c.call(this);return t===t?u.call(this):i})},function(t,n,r){var e=r(1);e(e.P,"Function",{bind:r(163)})},function(t,n,r){"use strict";var e=r(6),i=r(32),o=r(7)("hasInstance"),u=Function.prototype;o in u||r(11).f(u,o,{value:function(t){if("function"!=typeof this||!e(t))return!1;if(!e(this.prototype))return t instanceof this;for(;t=i(t);)if(this.prototype===t)return!0;return!1}})},function(t,n,r){var e=r(11).f,i=r(66),o=r(24),u=Function.prototype,c="name",f=Object.isExtensible||function(){return!0};c in u||r(10)&&e(u,c,{configurable:!0,get:function(){try{var t=this,n=(""+t).match(/^\s*function ([^ (]*)/)[1];return o(t,c)||!f(t)||e(t,c,i(5,n)),n}catch(t){return""}}})},function(t,n,r){var e=r(1),i=r(171),o=Math.sqrt,u=Math.acosh;e(e.S+e.F*!(u&&710==Math.floor(u(Number.MAX_VALUE))&&u(1/0)==1/0),"Math",{acosh:function(t){return(t=+t)<1?NaN:t>94906265.62425156?Math.log(t)+Math.LN2:i(t-1+o(t-1)*o(t+1))}})},function(t,n,r){function e(t){return isFinite(t=+t)&&0!=t?t<0?-e(-t):Math.log(t+Math.sqrt(t*t+1)):t}var i=r(1),o=Math.asinh;i(i.S+i.F*!(o&&1/o(0)>0),"Math",{asinh:e})},function(t,n,r){var e=r(1),i=Math.atanh;e(e.S+e.F*!(i&&1/i(-0)<0),"Math",{atanh:function(t){return 0==(t=+t)?t:Math.log((1+t)/(1-t))/2}})},function(t,n,r){var e=r(1),i=r(142);e(e.S,"Math",{cbrt:function(t){return i(t=+t)*Math.pow(Math.abs(t),1/3)}})},function(t,n,r){var e=r(1);e(e.S,"Math",{clz32:function(t){return(t>>>=0)?31-Math.floor(Math.log(t+.5)*Math.LOG2E):32}})},function(t,n,r){var e=r(1),i=Math.exp;e(e.S,"Math",{cosh:function(t){return(i(t=+t)+i(-t))/2}})},function(t,n,r){var e=r(1),i=r(141);e(e.S+e.F*(i!=Math.expm1),"Math",{expm1:i})},function(t,n,r){var e=r(1),i=r(142),o=Math.pow,u=o(2,-52),c=o(2,-23),f=o(2,127)*(2-c),a=o(2,-126),s=function(t){return t+1/u-1/u};e(e.S,"Math",{fround:function(t){var n,r,e=Math.abs(t),o=i(t);return e<a?o*s(e/a/c)*a*c:(n=(1+c/u)*e,r=n-(n-e),r>f||r!=r?o*(1/0):o*r)}})},function(t,n,r){var e=r(1),i=Math.abs;e(e.S,"Math",{hypot:function(t,n){for(var r,e,o=0,u=0,c=arguments.length,f=0;u<c;)r=i(arguments[u++]),f<r?(e=f/r,o=o*e*e+1,f=r):r>0?(e=r/f,o+=e*e):o+=r;return f===1/0?1/0:f*Math.sqrt(o)}})},function(t,n,r){var e=r(1),i=Math.imul;e(e.S+e.F*r(4)(function(){return-5!=i(4294967295,5)||2!=i.length}),"Math",{imul:function(t,n){var r=65535,e=+t,i=+n,o=r&e,u=r&i;return 0|o*u+((r&e>>>16)*u+o*(r&i>>>16)<<16>>>0)}})},function(t,n,r){var e=r(1);e(e.S,"Math",{log10:function(t){return Math.log(t)/Math.LN10}})},function(t,n,r){var e=r(1);e(e.S,"Math",{log1p:r(171)})},function(t,n,r){var e=r(1);e(e.S,"Math",{log2:function(t){return Math.log(t)/Math.LN2}})},function(t,n,r){var e=r(1);e(e.S,"Math",{sign:r(142)})},function(t,n,r){var e=r(1),i=r(141),o=Math.exp;e(e.S+e.F*r(4)(function(){return-2e-17!=!Math.sinh(-2e-17)}),"Math",{sinh:function(t){return Math.abs(t=+t)<1?(i(t)-i(-t))/2:(o(t-1)-o(-t-1))*(Math.E/2)}})},function(t,n,r){var e=r(1),i=r(141),o=Math.exp;e(e.S,"Math",{tanh:function(t){var n=i(t=+t),r=i(-t);return n==1/0?1:r==1/0?-1:(n-r)/(o(t)+o(-t))}})},function(t,n,r){var e=r(1);e(e.S,"Math",{trunc:function(t){return(t>0?Math.floor:Math.ceil)(t)}})},function(t,n,r){"use strict";var e=r(3),i=r(24),o=r(45),u=r(136),c=r(50),f=r(4),a=r(71).f,s=r(31).f,l=r(11).f,h=r(82).trim,v="Number",p=e[v],d=p,y=p.prototype,g=o(r(70)(y))==v,b="trim"in String.prototype,m=function(t){var n=c(t,!1);if("string"==typeof n&&n.length>2){n=b?n.trim():h(n,3);var r,e,i,o=n.charCodeAt(0);if(43===o||45===o){if(88===(r=n.charCodeAt(2))||120===r)return NaN}else if(48===o){switch(n.charCodeAt(1)){case 66:case 98:e=2,i=49;break;case 79:case 111:e=8,i=55;break;default:return+n}for(var u,f=n.slice(2),a=0,s=f.length;a<s;a++)if((u=f.charCodeAt(a))<48||u>i)return NaN;return parseInt(f,e)}}return+n};if(!p(" 0o1")||!p("0b1")||p("+0x1")){p=function(t){var n=arguments.length<1?0:t,r=this;return r instanceof p&&(g?f(function(){y.valueOf.call(r)}):o(r)!=v)?u(new d(m(n)),r,p):m(n)};for(var x,w=r(10)?a(d):"MAX_VALUE,MIN_VALUE,NaN,NEGATIVE_INFINITY,POSITIVE_INFINITY,EPSILON,isFinite,isInteger,isNaN,isSafeInteger,MAX_SAFE_INTEGER,MIN_SAFE_INTEGER,parseFloat,parseInt,isInteger".split(","),S=0;w.length>S;S++)i(d,x=w[S])&&!i(p,x)&&l(p,x,s(d,x));p.prototype=y,y.constructor=p,r(28)(e,v,p)}},function(t,n,r){var e=r(1);e(e.S,"Number",{EPSILON:Math.pow(2,-52)})},function(t,n,r){var e=r(1),i=r(3).isFinite;e(e.S,"Number",{isFinite:function(t){return"number"==typeof t&&i(t)}})},function(t,n,r){var e=r(1);e(e.S,"Number",{isInteger:r(168)})},function(t,n,r){var e=r(1);e(e.S,"Number",{isNaN:function(t){return t!=t}})},function(t,n,r){var e=r(1),i=r(168),o=Math.abs;e(e.S,"Number",{isSafeInteger:function(t){return i(t)&&o(t)<=9007199254740991}})},function(t,n,r){var e=r(1);e(e.S,"Number",{MAX_SAFE_INTEGER:9007199254740991})},function(t,n,r){var e=r(1);e(e.S,"Number",{MIN_SAFE_INTEGER:-9007199254740991})},function(t,n,r){var e=r(1),i=r(178);e(e.S+e.F*(Number.parseFloat!=i),"Number",{parseFloat:i})},function(t,n,r){var e=r(1),i=r(179);e(e.S+e.F*(Number.parseInt!=i),"Number",{parseInt:i})},function(t,n,r){"use strict";var e=r(1),i=r(67),o=r(159),u=r(149),c=1..toFixed,f=Math.floor,a=[0,0,0,0,0,0],s="Number.toFixed: incorrect invocation!",l="0",h=function(t,n){for(var r=-1,e=n;++r<6;)e+=t*a[r],a[r]=e%1e7,e=f(e/1e7)},v=function(t){for(var n=6,r=0;--n>=0;)r+=a[n],a[n]=f(r/t),r=r%t*1e7},p=function(){for(var t=6,n="";--t>=0;)if(""!==n||0===t||0!==a[t]){var r=String(a[t]);n=""===n?r:n+u.call(l,7-r.length)+r}return n},d=function(t,n,r){return 0===n?r:n%2==1?d(t,n-1,r*t):d(t*t,n/2,r)},y=function(t){for(var n=0,r=t;r>=4096;)n+=12,r/=4096;for(;r>=2;)n+=1,r/=2;return n};e(e.P+e.F*(!!c&&("0.000"!==8e-5.toFixed(3)||"1"!==.9.toFixed(0)||"1.25"!==1.255.toFixed(2)||"1000000000000000128"!==(0xde0b6b3a7640080).toFixed(0))||!r(4)(function(){c.call({})})),"Number",{toFixed:function(t){var n,r,e,c,f=o(this,s),a=i(t),g="",b=l;if(a<0||a>20)throw RangeError(s);if(f!=f)return"NaN";if(f<=-1e21||f>=1e21)return String(f);if(f<0&&(g="-",f=-f),f>1e-21)if(n=y(f*d(2,69,1))-69,r=n<0?f*d(2,-n,1):f/d(2,n,1),r*=4503599627370496,(n=52-n)>0){for(h(0,r),e=a;e>=7;)h(1e7,0),e-=7;for(h(d(10,e,1),0),e=n-1;e>=23;)v(1<<23),e-=23;v(1<<e),h(1,1),v(2),b=p()}else h(0,r),h(1<<-n,0),b=p()+u.call(l,a);return a>0?(c=b.length,b=g+(c<=a?"0."+u.call(l,a-c)+b:b.slice(0,c-a)+"."+b.slice(c-a))):b=g+b,b}})},function(t,n,r){"use strict";var e=r(1),i=r(4),o=r(159),u=1..toPrecision;e(e.P+e.F*(i(function(){return"1"!==u.call(1,void 0)})||!i(function(){u.call({})})),"Number",{toPrecision:function(t){var n=o(this,"Number#toPrecision: incorrect invocation!");return void 0===t?u.call(n):u.call(n,t)}})},function(t,n,r){var e=r(1);e(e.S+e.F,"Object",{assign:r(172)})},function(t,n,r){var e=r(1);e(e.S,"Object",{create:r(70)})},function(t,n,r){var e=r(1);e(e.S+e.F*!r(10),"Object",{defineProperties:r(173)})},function(t,n,r){var e=r(1);e(e.S+e.F*!r(10),"Object",{defineProperty:r(11).f})},function(t,n,r){var e=r(6),i=r(65).onFreeze;r(49)("freeze",function(t){return function(n){return t&&e(n)?t(i(n)):n}})},function(t,n,r){var e=r(30),i=r(31).f;r(49)("getOwnPropertyDescriptor",function(){return function(t,n){return i(e(t),n)}})},function(t,n,r){r(49)("getOwnPropertyNames",function(){return r(174).f})},function(t,n,r){var e=r(17),i=r(32);r(49)("getPrototypeOf",function(){return function(t){return i(e(t))}})},function(t,n,r){var e=r(6);r(49)("isExtensible",function(t){return function(n){return!!e(n)&&(!t||t(n))}})},function(t,n,r){var e=r(6);r(49)("isFrozen",function(t){return function(n){return!e(n)||!!t&&t(n)}})},function(t,n,r){var e=r(6);r(49)("isSealed",function(t){return function(n){return!e(n)||!!t&&t(n)}})},function(t,n,r){var e=r(1);e(e.S,"Object",{is:r(180)})},function(t,n,r){var e=r(17),i=r(72);r(49)("keys",function(){return function(t){return i(e(t))}})},function(t,n,r){var e=r(6),i=r(65).onFreeze;r(49)("preventExtensions",function(t){return function(n){return t&&e(n)?t(i(n)):n}})},function(t,n,r){var e=r(6),i=r(65).onFreeze;r(49)("seal",function(t){return function(n){return t&&e(n)?t(i(n)):n}})},function(t,n,r){var e=r(1);e(e.S,"Object",{setPrototypeOf:r(144).set})},function(t,n,r){"use strict";var e=r(114),i={};i[r(7)("toStringTag")]="z",i+""!="[object z]"&&r(28)(Object.prototype,"toString",function(){return"[object "+e(this)+"]"},!0)},function(t,n,r){var e=r(1),i=r(178);e(e.G+e.F*(parseFloat!=i),{parseFloat:i})},function(t,n,r){var e=r(1),i=r(179);e(e.G+e.F*(parseInt!=i),{parseInt:i})},function(t,n,r){"use strict";var e,i,o,u=r(69),c=r(3),f=r(53),a=r(114),s=r(1),l=r(6),h=r(26),v=r(68),p=r(79),d=r(146),y=r(151).set,g=r(143)(),b="Promise",m=c.TypeError,x=c.process,w=c[b],x=c.process,S="process"==a(x),_=function(){},O=!!function(){try{var t=w.resolve(1),n=(t.constructor={})[r(7)("species")]=function(t){t(_,_)};return(S||"function"==typeof PromiseRejectionEvent)&&t.then(_)instanceof n}catch(t){}}(),E=function(t,n){return t===n||t===w&&n===o},P=function(t){var n;return!(!l(t)||"function"!=typeof(n=t.then))&&n},j=function(t){return E(w,t)?new F(t):new i(t)},F=i=function(t){var n,r;this.promise=new t(function(t,e){if(void 0!==n||void 0!==r)throw m("Bad Promise constructor");n=t,r=e}),this.resolve=h(n),this.reject=h(r)},M=function(t){try{t()}catch(t){return{error:t}}},A=function(t,n){if(!t._n){t._n=!0;var r=t._c;g(function(){for(var e=t._v,i=1==t._s,o=0;r.length>o;)!function(n){var r,o,u=i?n.ok:n.fail,c=n.resolve,f=n.reject,a=n.domain;try{u?(i||(2==t._h&&I(t),t._h=1),!0===u?r=e:(a&&a.enter(),r=u(e),a&&a.exit()),r===n.promise?f(m("Promise-chain cycle")):(o=P(r))?o.call(r,c,f):c(r)):f(e)}catch(t){f(t)}}(r[o++]);t._c=[],t._n=!1,n&&!t._h&&N(t)})}},N=function(t){y.call(c,function(){var n,r,e,i=t._v;if(T(t)&&(n=M(function(){S?x.emit("unhandledRejection",i,t):(r=c.onunhandledrejection)?r({promise:t,reason:i}):(e=c.console)&&e.error&&e.error("Unhandled promise rejection",i)}),t._h=S||T(t)?2:1),t._a=void 0,n)throw n.error})},T=function(t){if(1==t._h)return!1;for(var n,r=t._a||t._c,e=0;r.length>e;)if(n=r[e++],n.fail||!T(n.promise))return!1;return!0},I=function(t){y.call(c,function(){var n;S?x.emit("rejectionHandled",t):(n=c.onrejectionhandled)&&n({promise:t,reason:t._v})})},k=function(t){var n=this;n._d||(n._d=!0,n=n._w||n,n._v=t,n._s=2,n._a||(n._a=n._c.slice()),A(n,!0))},L=function(t){var n,r=this;if(!r._d){r._d=!0,r=r._w||r;try{if(r===t)throw m("Promise can't be resolved itself");(n=P(t))?g(function(){var e={_w:r,_d:!1};try{n.call(t,f(L,e,1),f(k,e,1))}catch(t){k.call(e,t)}}):(r._v=t,r._s=1,A(r,!1))}catch(t){k.call({_w:r,_d:!1},t)}}};O||(w=function(t){v(this,w,b,"_h"),h(t),e.call(this);try{t(f(L,this,1),f(k,this,1))}catch(t){k.call(this,t)}},e=function(t){this._c=[],this._a=void 0,this._s=0,this._d=!1,this._v=void 0,this._h=0,this._n=!1},e.prototype=r(73)(w.prototype,{then:function(t,n){var r=j(d(this,w));return r.ok="function"!=typeof t||t,r.fail="function"==typeof n&&n,r.domain=S?x.domain:void 0,this._c.push(r),this._a&&this._a.push(r),this._s&&A(this,!1),r.promise},catch:function(t){return this.then(void 0,t)}}),F=function(){var t=new e;this.promise=t,this.resolve=f(L,t,1),this.reject=f(k,t,1)}),s(s.G+s.W+s.F*!O,{Promise:w}),r(81)(w,b),r(74)(b),o=r(52)[b],s(s.S+s.F*!O,b,{reject:function(t){var n=j(this);return(0,n.reject)(t),n.promise}}),s(s.S+s.F*(u||!O),b,{resolve:function(t){if(t instanceof w&&E(t.constructor,this))return t;var n=j(this);return(0,n.resolve)(t),n.promise}}),s(s.S+s.F*!(O&&r(123)(function(t){w.all(t).catch(_)})),b,{all:function(t){var n=this,r=j(n),e=r.resolve,i=r.reject,o=M(function(){var r=[],o=0,u=1;p(t,!1,function(t){var c=o++,f=!1;r.push(void 0),u++,n.resolve(t).then(function(t){f||(f=!0,r[c]=t,--u||e(r))},i)}),--u||e(r)});return o&&i(o.error),r.promise},race:function(t){var n=this,r=j(n),e=r.reject,i=M(function(){p(t,!1,function(t){n.resolve(t).then(r.resolve,e)})});return i&&e(i.error),r.promise}})},function(t,n,r){var e=r(1),i=r(26),o=r(2),u=(r(3).Reflect||{}).apply,c=Function.apply;e(e.S+e.F*!r(4)(function(){u(function(){})}),"Reflect",{apply:function(t,n,r){var e=i(t),f=o(r);return u?u(e,n,f):c.call(e,n,f)}})},function(t,n,r){var e=r(1),i=r(70),o=r(26),u=r(2),c=r(6),f=r(4),a=r(163),s=(r(3).Reflect||{}).construct,l=f(function(){function t(){}return!(s(function(){},[],t)instanceof t)}),h=!f(function(){s(function(){})});e(e.S+e.F*(l||h),"Reflect",{construct:function(t,n){o(t),u(n);var r=arguments.length<3?t:o(arguments[2]);if(h&&!l)return s(t,n,r);if(t==r){switch(n.length){case 0:return new t;case 1:return new t(n[0]);case 2:return new t(n[0],n[1]);case 3:return new t(n[0],n[1],n[2]);case 4:return new t(n[0],n[1],n[2],n[3])}var e=[null];return e.push.apply(e,n),new(a.apply(t,e))}var f=r.prototype,v=i(c(f)?f:Object.prototype),p=Function.apply.call(t,v,n);return c(p)?p:v}})},function(t,n,r){var e=r(11),i=r(1),o=r(2),u=r(50);i(i.S+i.F*r(4)(function(){Reflect.defineProperty(e.f({},1,{value:1}),1,{value:2})}),"Reflect",{defineProperty:function(t,n,r){o(t),n=u(n,!0),o(r);try{return e.f(t,n,r),!0}catch(t){return!1}}})},function(t,n,r){var e=r(1),i=r(31).f,o=r(2);e(e.S,"Reflect",{deleteProperty:function(t,n){var r=i(o(t),n);return!(r&&!r.configurable)&&delete t[n]}})},function(t,n,r){"use strict";var e=r(1),i=r(2),o=function(t){this._t=i(t),this._i=0;var n,r=this._k=[];for(n in t)r.push(n)};r(139)(o,"Object",function(){var t,n=this,r=n._k;do{if(n._i>=r.length)return{value:void 0,done:!0}}while(!((t=r[n._i++])in n._t));return{value:t,done:!1}}),e(e.S,"Reflect",{enumerate:function(t){return new o(t)}})},function(t,n,r){var e=r(31),i=r(1),o=r(2);i(i.S,"Reflect",{getOwnPropertyDescriptor:function(t,n){return e.f(o(t),n)}})},function(t,n,r){var e=r(1),i=r(32),o=r(2);e(e.S,"Reflect",{getPrototypeOf:function(t){return i(o(t))}})},function(t,n,r){function e(t,n){var r,c,s=arguments.length<3?t:arguments[2];return a(t)===s?t[n]:(r=i.f(t,n))?u(r,"value")?r.value:void 0!==r.get?r.get.call(s):void 0:f(c=o(t))?e(c,n,s):void 0}var i=r(31),o=r(32),u=r(24),c=r(1),f=r(6),a=r(2);c(c.S,"Reflect",{get:e})},function(t,n,r){var e=r(1);e(e.S,"Reflect",{has:function(t,n){return n in t}})},function(t,n,r){var e=r(1),i=r(2),o=Object.isExtensible;e(e.S,"Reflect",{isExtensible:function(t){return i(t),!o||o(t)}})},function(t,n,r){var e=r(1);e(e.S,"Reflect",{ownKeys:r(177)})},function(t,n,r){var e=r(1),i=r(2),o=Object.preventExtensions;e(e.S,"Reflect",{preventExtensions:function(t){i(t);try{return o&&o(t),!0}catch(t){return!1}}})},function(t,n,r){var e=r(1),i=r(144);i&&e(e.S,"Reflect",{setPrototypeOf:function(t,n){i.check(t,n);try{return i.set(t,n),!0}catch(t){return!1}}})},function(t,n,r){function e(t,n,r){var f,h,v=arguments.length<4?t:arguments[3],p=o.f(s(t),n);if(!p){if(l(h=u(t)))return e(h,n,r,v);p=a(0)}return c(p,"value")?!(!1===p.writable||!l(v)||(f=o.f(v,n)||a(0),f.value=r,i.f(v,n,f),0)):void 0!==p.set&&(p.set.call(v,r),!0)}var i=r(11),o=r(31),u=r(32),c=r(24),f=r(1),a=r(66),s=r(2),l=r(6);f(f.S,"Reflect",{set:e})},function(t,n,r){var e=r(3),i=r(136),o=r(11).f,u=r(71).f,c=r(122),f=r(120),a=e.RegExp,s=a,l=a.prototype,h=/a/g,v=/a/g,p=new a(h)!==h;if(r(10)&&(!p||r(4)(function(){return v[r(7)("match")]=!1,a(h)!=h||a(v)==v||"/a/i"!=a(h,"i")}))){a=function(t,n){var r=this instanceof a,e=c(t),o=void 0===n;return!r&&e&&t.constructor===a&&o?t:i(p?new s(e&&!o?t.source:t,n):s((e=t instanceof a)?t.source:t,e&&o?f.call(t):n),r?this:l,a)};for(var d=u(s),y=0;d.length>y;)!function(t){t in a||o(a,t,{configurable:!0,get:function(){return s[t]},set:function(n){s[t]=n}})}(d[y++]);l.constructor=a,a.prototype=l,r(28)(e,"RegExp",a)}r(74)("RegExp")},function(t,n,r){r(119)("match",1,function(t,n,r){return[function(r){"use strict";var e=t(this),i=void 0==r?void 0:r[n];return void 0!==i?i.call(r,e):new RegExp(r)[n](String(e))},r]})},function(t,n,r){r(119)("replace",2,function(t,n,r){return[function(e,i){"use strict";var o=t(this),u=void 0==e?void 0:e[n];return void 0!==u?u.call(e,o,i):r.call(String(o),e,i)},r]})},function(t,n,r){r(119)("search",1,function(t,n,r){return[function(r){"use strict";var e=t(this),i=void 0==r?void 0:r[n];return void 0!==i?i.call(r,e):new RegExp(r)[n](String(e))},r]})},function(t,n,r){r(119)("split",2,function(t,n,e){"use strict";var i=r(122),o=e,u=[].push,c="split",f="length",a="lastIndex";if("c"=="abbc"[c](/(b)*/)[1]||4!="test"[c](/(?:)/,-1)[f]||2!="ab"[c](/(?:ab)*/)[f]||4!="."[c](/(.?)(.?)/)[f]||"."[c](/()()/)[f]>1||""[c](/.?/)[f]){var s=void 0===/()??/.exec("")[1];e=function(t,n){var r=String(this);if(void 0===t&&0===n)return[];if(!i(t))return o.call(r,t,n);var e,c,l,h,v,p=[],d=(t.ignoreCase?"i":"")+(t.multiline?"m":"")+(t.unicode?"u":"")+(t.sticky?"y":""),y=0,g=void 0===n?4294967295:n>>>0,b=new RegExp(t.source,d+"g");for(s||(e=new RegExp("^"+b.source+"$(?!\\s)",d));(c=b.exec(r))&&!((l=c.index+c[0][f])>y&&(p.push(r.slice(y,c.index)),!s&&c[f]>1&&c[0].replace(e,function(){for(v=1;v<arguments[f]-2;v++)void 0===arguments[v]&&(c[v]=void 0)}),c[f]>1&&c.index<r[f]&&u.apply(p,c.slice(1)),h=c[0][f],y=l,p[f]>=g));)b[a]===c.index&&b[a]++;return y===r[f]?!h&&b.test("")||p.push(""):p.push(r.slice(y)),p[f]>g?p.slice(0,g):p}}else"0"[c](void 0,0)[f]&&(e=function(t,n){return void 0===t&&0===n?[]:o.call(this,t,n)});return[function(r,i){var o=t(this),u=void 0==r?void 0:r[n];return void 0!==u?u.call(r,o,i):e.call(String(o),r,i)},e]})},function(t,n,r){"use strict";r(184);var e=r(2),i=r(120),o=r(10),u="toString",c=/./[u],f=function(t){r(28)(RegExp.prototype,u,t,!0)};r(4)(function(){return"/a/b"!=c.call({source:"a",flags:"b"})})?f(function(){var t=e(this);return"/".concat(t.source,"/","flags"in t?t.flags:!o&&t instanceof RegExp?i.call(t):void 0)}):c.name!=u&&f(function(){return c.call(this)})},function(t,n,r){"use strict";r(29)("anchor",function(t){return function(n){return t(this,"a","name",n)}})},function(t,n,r){"use strict";r(29)("big",function(t){return function(){return t(this,"big","","")}})},function(t,n,r){"use strict";r(29)("blink",function(t){return function(){return t(this,"blink","","")}})},function(t,n,r){"use strict";r(29)("bold",function(t){return function(){return t(this,"b","","")}})},function(t,n,r){"use strict";var e=r(1),i=r(147)(!1);e(e.P,"String",{codePointAt:function(t){return i(this,t)}})},function(t,n,r){"use strict";var e=r(1),i=r(16),o=r(148),u="endsWith",c=""[u];e(e.P+e.F*r(134)(u),"String",{endsWith:function(t){var n=o(this,t,u),r=arguments.length>1?arguments[1]:void 0,e=i(n.length),f=void 0===r?e:Math.min(i(r),e),a=String(t);return c?c.call(n,a,f):n.slice(f-a.length,f)===a}})},function(t,n,r){"use strict";r(29)("fixed",function(t){return function(){return t(this,"tt","","")}})},function(t,n,r){"use strict";r(29)("fontcolor",function(t){return function(n){return t(this,"font","color",n)}})},function(t,n,r){"use strict";r(29)("fontsize",function(t){return function(n){return t(this,"font","size",n)}})},function(t,n,r){var e=r(1),i=r(75),o=String.fromCharCode,u=String.fromCodePoint;e(e.S+e.F*(!!u&&1!=u.length),"String",{fromCodePoint:function(t){for(var n,r=[],e=arguments.length,u=0;e>u;){if(n=+arguments[u++],i(n,1114111)!==n)throw RangeError(n+" is not a valid code point");r.push(n<65536?o(n):o(55296+((n-=65536)>>10),n%1024+56320))}return r.join("")}})},function(t,n,r){"use strict";var e=r(1),i=r(148),o="includes";e(e.P+e.F*r(134)(o),"String",{includes:function(t){return!!~i(this,t,o).indexOf(t,arguments.length>1?arguments[1]:void 0)}})},function(t,n,r){"use strict";r(29)("italics",function(t){return function(){return t(this,"i","","")}})},function(t,n,r){"use strict";var e=r(147)(!0);r(140)(String,"String",function(t){this._t=String(t),this._i=0},function(){var t,n=this._t,r=this._i;return r>=n.length?{value:void 0,done:!0}:(t=e(n,r),this._i+=t.length,{value:t,done:!1})})},function(t,n,r){"use strict";r(29)("link",function(t){return function(n){return t(this,"a","href",n)}})},function(t,n,r){var e=r(1),i=r(30),o=r(16);e(e.S,"String",{raw:function(t){for(var n=i(t.raw),r=o(n.length),e=arguments.length,u=[],c=0;r>c;)u.push(String(n[c++])),c<e&&u.push(String(arguments[c]));return u.join("")}})},function(t,n,r){var e=r(1);e(e.P,"String",{repeat:r(149)})},function(t,n,r){"use strict";r(29)("small",function(t){return function(){return t(this,"small","","")}})},function(t,n,r){"use strict";var e=r(1),i=r(16),o=r(148),u="startsWith",c=""[u];e(e.P+e.F*r(134)(u),"String",{startsWith:function(t){var n=o(this,t,u),r=i(Math.min(arguments.length>1?arguments[1]:void 0,n.length)),e=String(t);return c?c.call(n,e,r):n.slice(r,r+e.length)===e}})},function(t,n,r){"use strict";r(29)("strike",function(t){return function(){return t(this,"strike","","")}})},function(t,n,r){"use strict";r(29)("sub",function(t){return function(){return t(this,"sub","","")}})},function(t,n,r){"use strict";r(29)("sup",function(t){return function(){return t(this,"sup","","")}})},function(t,n,r){"use strict";r(82)("trim",function(t){return function(){return t(this,3)}})},function(t,n,r){"use strict";var e=r(3),i=r(24),o=r(10),u=r(1),c=r(28),f=r(65).KEY,a=r(4),s=r(126),l=r(81),h=r(76),v=r(7),p=r(182),d=r(153),y=r(206),g=r(205),b=r(138),m=r(2),x=r(30),w=r(50),S=r(66),_=r(70),O=r(174),E=r(31),P=r(11),j=r(72),F=E.f,M=P.f,A=O.f,N=e.Symbol,T=e.JSON,I=T&&T.stringify,k="prototype",L=v("_hidden"),R=v("toPrimitive"),C={}.propertyIsEnumerable,D=s("symbol-registry"),U=s("symbols"),W=s("op-symbols"),G=Object[k],B="function"==typeof N,V=e.QObject,z=!V||!V[k]||!V[k].findChild,q=o&&a(function(){return 7!=_(M({},"a",{get:function(){return M(this,"a",{value:7}).a}})).a})?function(t,n,r){var e=F(G,n);e&&delete G[n],M(t,n,r),e&&t!==G&&M(G,n,e)}:M,K=function(t){var n=U[t]=_(N[k]);return n._k=t,n},J=B&&"symbol"==typeof N.iterator?function(t){return"symbol"==typeof t}:function(t){return t instanceof N},Y=function(t,n,r){return t===G&&Y(W,n,r),m(t),n=w(n,!0),m(r),i(U,n)?(r.enumerable?(i(t,L)&&t[L][n]&&(t[L][n]=!1),r=_(r,{enumerable:S(0,!1)})):(i(t,L)||M(t,L,S(1,{})),t[L][n]=!0),q(t,n,r)):M(t,n,r)},H=function(t,n){m(t);for(var r,e=g(n=x(n)),i=0,o=e.length;o>i;)Y(t,r=e[i++],n[r]);return t},$=function(t,n){return void 0===n?_(t):H(_(t),n)},X=function(t){var n=C.call(this,t=w(t,!0));return!(this===G&&i(U,t)&&!i(W,t))&&(!(n||!i(this,t)||!i(U,t)||i(this,L)&&this[L][t])||n)},Q=function(t,n){if(t=x(t),n=w(n,!0),t!==G||!i(U,n)||i(W,n)){var r=F(t,n);return!r||!i(U,n)||i(t,L)&&t[L][n]||(r.enumerable=!0),r}},Z=function(t){for(var n,r=A(x(t)),e=[],o=0;r.length>o;)i(U,n=r[o++])||n==L||n==f||e.push(n);return e},tt=function(t){for(var n,r=t===G,e=A(r?W:x(t)),o=[],u=0;e.length>u;)!i(U,n=e[u++])||r&&!i(G,n)||o.push(U[n]);return o};B||(N=function(){if(this instanceof N)throw TypeError("Symbol is not a constructor!");var t=h(arguments.length>0?arguments[0]:void 0),n=function(r){this===G&&n.call(W,r),i(this,L)&&i(this[L],t)&&(this[L][t]=!1),q(this,t,S(1,r))};return o&&z&&q(G,t,{configurable:!0,set:n}),K(t)},c(N[k],"toString",function(){return this._k}),E.f=Q,P.f=Y,r(71).f=O.f=Z,r(116).f=X,r(125).f=tt,o&&!r(69)&&c(G,"propertyIsEnumerable",X,!0),p.f=function(t){return K(v(t))}),u(u.G+u.W+u.F*!B,{Symbol:N});for(var nt="hasInstance,isConcatSpreadable,iterator,match,replace,search,species,split,toPrimitive,toStringTag,unscopables".split(","),rt=0;nt.length>rt;)v(nt[rt++]);for(var nt=j(v.store),rt=0;nt.length>rt;)d(nt[rt++]);u(u.S+u.F*!B,"Symbol",{for:function(t){return i(D,t+="")?D[t]:D[t]=N(t)},keyFor:function(t){if(J(t))return y(D,t);throw TypeError(t+" is not a symbol!")},useSetter:function(){z=!0},useSimple:function(){z=!1}}),u(u.S+u.F*!B,"Object",{create:$,defineProperty:Y,defineProperties:H,getOwnPropertyDescriptor:Q,getOwnPropertyNames:Z,getOwnPropertySymbols:tt}),T&&u(u.S+u.F*(!B||a(function(){var t=N();return"[null]"!=I([t])||"{}"!=I({a:t})||"{}"!=I(Object(t))})),"JSON",{stringify:function(t){if(void 0!==t&&!J(t)){for(var n,r,e=[t],i=1;arguments.length>i;)e.push(arguments[i++]);return n=e[1],"function"==typeof n&&(r=n),!r&&b(n)||(n=function(t,n){if(r&&(n=r.call(this,t,n)),!J(n))return n}),e[1]=n,I.apply(T,e)}}}),N[k][R]||r(27)(N[k],R,N[k].valueOf),l(N,"Symbol"),l(Math,"Math",!0),l(e.JSON,"JSON",!0)},function(t,n,r){"use strict";var e=r(1),i=r(127),o=r(152),u=r(2),c=r(75),f=r(16),a=r(6),s=r(3).ArrayBuffer,l=r(146),h=o.ArrayBuffer,v=o.DataView,p=i.ABV&&s.isView,d=h.prototype.slice,y=i.VIEW,g="ArrayBuffer";e(e.G+e.W+e.F*(s!==h),{ArrayBuffer:h}),e(e.S+e.F*!i.CONSTR,g,{isView:function(t){return p&&p(t)||a(t)&&y in t}}),e(e.P+e.U+e.F*r(4)(function(){return!new h(2).slice(1,void 0).byteLength}),g,{slice:function(t,n){if(void 0!==d&&void 0===n)return d.call(u(this),t);for(var r=u(this).byteLength,e=c(t,r),i=c(void 0===n?r:n,r),o=new(l(this,h))(f(i-e)),a=new v(this),s=new v(o),p=0;e<i;)s.setUint8(p++,a.getUint8(e++));return o}}),r(74)(g)},function(t,n,r){var e=r(1);e(e.G+e.W+e.F*!r(127).ABV,{DataView:r(152).DataView})},function(t,n,r){r(55)("Float32",4,function(t){return function(n,r,e){return t(this,n,r,e)}})},function(t,n,r){r(55)("Float64",8,function(t){return function(n,r,e){return t(this,n,r,e)}})},function(t,n,r){r(55)("Int16",2,function(t){return function(n,r,e){return t(this,n,r,e)}})},function(t,n,r){r(55)("Int32",4,function(t){return function(n,r,e){return t(this,n,r,e)}})},function(t,n,r){r(55)("Int8",1,function(t){return function(n,r,e){return t(this,n,r,e)}})},function(t,n,r){r(55)("Uint16",2,function(t){return function(n,r,e){return t(this,n,r,e)}})},function(t,n,r){r(55)("Uint32",4,function(t){return function(n,r,e){return t(this,n,r,e)}})},function(t,n,r){r(55)("Uint8",1,function(t){return function(n,r,e){return t(this,n,r,e)}})},function(t,n,r){r(55)("Uint8",1,function(t){return function(n,r,e){return t(this,n,r,e)}},!0)},function(t,n,r){"use strict";var e=r(166);r(118)("WeakSet",function(t){return function(){return t(this,arguments.length>0?arguments[0]:void 0)}},{add:function(t){return e.def(this,t,!0)}},e,!1,!0)},function(t,n,r){"use strict";var e=r(1),i=r(117)(!0);e(e.P,"Array",{includes:function(t){return i(this,t,arguments.length>1?arguments[1]:void 0)}}),r(78)("includes")},function(t,n,r){var e=r(1),i=r(143)(),o=r(3).process,u="process"==r(45)(o);e(e.G,{asap:function(t){var n=u&&o.domain;i(n?n.bind(t):t)}})},function(t,n,r){var e=r(1),i=r(45);e(e.S,"Error",{isError:function(t){return"Error"===i(t)}})},function(t,n,r){var e=r(1);e(e.P+e.R,"Map",{toJSON:r(165)("Map")})},function(t,n,r){var e=r(1);e(e.S,"Math",{iaddh:function(t,n,r,e){var i=t>>>0,o=n>>>0,u=r>>>0;return o+(e>>>0)+((i&u|(i|u)&~(i+u>>>0))>>>31)|0}})},function(t,n,r){var e=r(1);e(e.S,"Math",{imulh:function(t,n){var r=65535,e=+t,i=+n,o=e&r,u=i&r,c=e>>16,f=i>>16,a=(c*u>>>0)+(o*u>>>16);return c*f+(a>>16)+((o*f>>>0)+(a&r)>>16)}})},function(t,n,r){var e=r(1);e(e.S,"Math",{isubh:function(t,n,r,e){var i=t>>>0,o=n>>>0,u=r>>>0;return o-(e>>>0)-((~i&u|~(i^u)&i-u>>>0)>>>31)|0}})},function(t,n,r){var e=r(1);e(e.S,"Math",{umulh:function(t,n){var r=65535,e=+t,i=+n,o=e&r,u=i&r,c=e>>>16,f=i>>>16,a=(c*u>>>0)+(o*u>>>16);return c*f+(a>>>16)+((o*f>>>0)+(a&r)>>>16)}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(26),u=r(11);r(10)&&e(e.P+r(124),"Object",{__defineGetter__:function(t,n){u.f(i(this),t,{get:o(n),enumerable:!0,configurable:!0})}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(26),u=r(11);r(10)&&e(e.P+r(124),"Object",{__defineSetter__:function(t,n){u.f(i(this),t,{set:o(n),enumerable:!0,configurable:!0})}})},function(t,n,r){var e=r(1),i=r(176)(!0);e(e.S,"Object",{entries:function(t){return i(t)}})},function(t,n,r){var e=r(1),i=r(177),o=r(30),u=r(31),c=r(131);e(e.S,"Object",{getOwnPropertyDescriptors:function(t){for(var n,r=o(t),e=u.f,f=i(r),a={},s=0;f.length>s;)c(a,n=f[s++],e(r,n));return a}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(50),u=r(32),c=r(31).f;r(10)&&e(e.P+r(124),"Object",{__lookupGetter__:function(t){var n,r=i(this),e=o(t,!0);do{if(n=c(r,e))return n.get}while(r=u(r))}})},function(t,n,r){"use strict";var e=r(1),i=r(17),o=r(50),u=r(32),c=r(31).f;r(10)&&e(e.P+r(124),"Object",{__lookupSetter__:function(t){var n,r=i(this),e=o(t,!0);do{if(n=c(r,e))return n.set}while(r=u(r))}})},function(t,n,r){var e=r(1),i=r(176)(!1);e(e.S,"Object",{values:function(t){return i(t)}})},function(t,n,r){"use strict";var e=r(1),i=r(3),o=r(52),u=r(143)(),c=r(7)("observable"),f=r(26),a=r(2),s=r(68),l=r(73),h=r(27),v=r(79),p=v.RETURN,d=function(t){return null==t?void 0:f(t)},y=function(t){var n=t._c;n&&(t._c=void 0,n())},g=function(t){return void 0===t._o},b=function(t){g(t)||(t._o=void 0,y(t))},m=function(t,n){a(t),this._c=void 0,this._o=t,t=new x(this);try{var r=n(t),e=r;null!=r&&("function"==typeof r.unsubscribe?r=function(){e.unsubscribe()}:f(r),this._c=r)}catch(n){return void t.error(n)}g(this)&&y(this)};m.prototype=l({},{unsubscribe:function(){b(this)}});var x=function(t){this._s=t};x.prototype=l({},{next:function(t){var n=this._s;if(!g(n)){var r=n._o;try{var e=d(r.next);if(e)return e.call(r,t)}catch(t){try{b(n)}finally{throw t}}}},error:function(t){var n=this._s;if(g(n))throw t;var r=n._o;n._o=void 0;try{var e=d(r.error);if(!e)throw t;t=e.call(r,t)}catch(t){try{y(n)}finally{throw t}}return y(n),t},complete:function(t){var n=this._s;if(!g(n)){var r=n._o;n._o=void 0;try{var e=d(r.complete);t=e?e.call(r,t):void 0}catch(t){try{y(n)}finally{throw t}}return y(n),t}}});var w=function(t){s(this,w,"Observable","_f")._f=f(t)};l(w.prototype,{subscribe:function(t){return new m(t,this._f)},forEach:function(t){var n=this;return new(o.Promise||i.Promise)(function(r,e){f(t);var i=n.subscribe({next:function(n){try{return t(n)}catch(t){e(t),i.unsubscribe()}},error:e,complete:r})})}}),l(w,{from:function(t){var n="function"==typeof this?this:w,r=d(a(t)[c]);if(r){var e=a(r.call(t));return e.constructor===n?e:new n(function(t){return e.subscribe(t)})}return new n(function(n){var r=!1;return u(function(){if(!r){try{if(v(t,!1,function(t){if(n.next(t),r)return p})===p)return}catch(t){if(r)throw t;return void n.error(t)}n.complete()}}),function(){r=!0}})},of:function(){for(var t=0,n=arguments.length,r=Array(n);t<n;)r[t]=arguments[t++];return new("function"==typeof this?this:w)(function(t){var n=!1;return u(function(){if(!n){for(var e=0;e<r.length;++e)if(t.next(r[e]),n)return;t.complete()}}),function(){n=!0}})}}),h(w.prototype,c,function(){return this}),e(e.G,{Observable:w}),r(74)("Observable")},function(t,n,r){var e=r(54),i=r(2),o=e.key,u=e.set;e.exp({defineMetadata:function(t,n,r,e){u(t,n,i(r),o(e))}})},function(t,n,r){var e=r(54),i=r(2),o=e.key,u=e.map,c=e.store;e.exp({deleteMetadata:function(t,n){var r=arguments.length<3?void 0:o(arguments[2]),e=u(i(n),r,!1);if(void 0===e||!e.delete(t))return!1;if(e.size)return!0;var f=c.get(n);return f.delete(r),!!f.size||c.delete(n)}})},function(t,n,r){var e=r(185),i=r(161),o=r(54),u=r(2),c=r(32),f=o.keys,a=o.key,s=function(t,n){var r=f(t,n),o=c(t);if(null===o)return r;var u=s(o,n);return u.length?r.length?i(new e(r.concat(u))):u:r};o.exp({getMetadataKeys:function(t){return s(u(t),arguments.length<2?void 0:a(arguments[1]))}})},function(t,n,r){var e=r(54),i=r(2),o=r(32),u=e.has,c=e.get,f=e.key,a=function(t,n,r){if(u(t,n,r))return c(t,n,r);var e=o(n);return null!==e?a(t,e,r):void 0};e.exp({getMetadata:function(t,n){return a(t,i(n),arguments.length<3?void 0:f(arguments[2]))}})},function(t,n,r){var e=r(54),i=r(2),o=e.keys,u=e.key;e.exp({getOwnMetadataKeys:function(t){
return o(i(t),arguments.length<2?void 0:u(arguments[1]))}})},function(t,n,r){var e=r(54),i=r(2),o=e.get,u=e.key;e.exp({getOwnMetadata:function(t,n){return o(t,i(n),arguments.length<3?void 0:u(arguments[2]))}})},function(t,n,r){var e=r(54),i=r(2),o=r(32),u=e.has,c=e.key,f=function(t,n,r){if(u(t,n,r))return!0;var e=o(n);return null!==e&&f(t,e,r)};e.exp({hasMetadata:function(t,n){return f(t,i(n),arguments.length<3?void 0:c(arguments[2]))}})},function(t,n,r){var e=r(54),i=r(2),o=e.has,u=e.key;e.exp({hasOwnMetadata:function(t,n){return o(t,i(n),arguments.length<3?void 0:u(arguments[2]))}})},function(t,n,r){var e=r(54),i=r(2),o=r(26),u=e.key,c=e.set;e.exp({metadata:function(t,n){return function(r,e){c(t,n,(void 0!==e?i:o)(r),u(e))}}})},function(t,n,r){var e=r(1);e(e.P+e.R,"Set",{toJSON:r(165)("Set")})},function(t,n,r){"use strict";var e=r(1),i=r(147)(!0);e(e.P,"String",{at:function(t){return i(this,t)}})},function(t,n,r){"use strict";var e=r(1),i=r(46),o=r(16),u=r(122),c=r(120),f=RegExp.prototype,a=function(t,n){this._r=t,this._s=n};r(139)(a,"RegExp String",function(){var t=this._r.exec(this._s);return{value:t,done:null===t}}),e(e.P,"String",{matchAll:function(t){if(i(this),!u(t))throw TypeError(t+" is not a regexp!");var n=String(this),r="flags"in f?String(t.flags):c.call(t),e=new RegExp(t.source,~r.indexOf("g")?r:"g"+r);return e.lastIndex=o(t.lastIndex),new a(e,n)}})},function(t,n,r){"use strict";var e=r(1),i=r(181);e(e.P,"String",{padEnd:function(t){return i(this,t,arguments.length>1?arguments[1]:void 0,!1)}})},function(t,n,r){"use strict";var e=r(1),i=r(181);e(e.P,"String",{padStart:function(t){return i(this,t,arguments.length>1?arguments[1]:void 0,!0)}})},function(t,n,r){"use strict";r(82)("trimLeft",function(t){return function(){return t(this,1)}},"trimStart")},function(t,n,r){"use strict";r(82)("trimRight",function(t){return function(){return t(this,2)}},"trimEnd")},function(t,n,r){r(153)("asyncIterator")},function(t,n,r){r(153)("observable")},function(t,n,r){var e=r(1);e(e.S,"System",{global:r(3)})},function(t,n,r){for(var e=r(155),i=r(28),o=r(3),u=r(27),c=r(80),f=r(7),a=f("iterator"),s=f("toStringTag"),l=c.Array,h=["NodeList","DOMTokenList","MediaList","StyleSheetList","CSSRuleList"],v=0;v<5;v++){var p,d=h[v],y=o[d],g=y&&y.prototype;if(g){g[a]||u(g,a,l),g[s]||u(g,s,d),c[d]=l;for(p in e)g[p]||i(g,p,e[p],!0)}}},function(t,n,r){var e=r(1),i=r(151);e(e.G+e.B,{setImmediate:i.set,clearImmediate:i.clear})},function(t,n,r){var e=r(3),i=r(1),o=r(121),u=r(207),c=e.navigator,f=!!c&&/MSIE .\./.test(c.userAgent),a=function(t){return f?function(n,r){return t(o(u,[].slice.call(arguments,2),"function"==typeof n?n:Function(n)),r)}:t};i(i.G+i.B+i.F*f,{setTimeout:a(e.setTimeout),setInterval:a(e.setInterval)})},function(t,n,r){r(330),r(269),r(271),r(270),r(273),r(275),r(280),r(274),r(272),r(282),r(281),r(277),r(278),r(276),r(268),r(279),r(283),r(284),r(236),r(238),r(237),r(286),r(285),r(256),r(266),r(267),r(257),r(258),r(259),r(260),r(261),r(262),r(263),r(264),r(265),r(239),r(240),r(241),r(242),r(243),r(244),r(245),r(246),r(247),r(248),r(249),r(250),r(251),r(252),r(253),r(254),r(255),r(317),r(322),r(329),r(320),r(312),r(313),r(318),r(323),r(325),r(308),r(309),r(310),r(311),r(314),r(315),r(316),r(319),r(321),r(324),r(326),r(327),r(328),r(231),r(233),r(232),r(235),r(234),r(220),r(218),r(224),r(221),r(227),r(229),r(217),r(223),r(214),r(228),r(212),r(226),r(225),r(219),r(222),r(211),r(213),r(216),r(215),r(230),r(155),r(302),r(307),r(184),r(303),r(304),r(305),r(306),r(287),r(183),r(185),r(186),r(342),r(331),r(332),r(337),r(340),r(341),r(335),r(338),r(336),r(339),r(333),r(334),r(288),r(289),r(290),r(291),r(292),r(295),r(293),r(294),r(296),r(297),r(298),r(299),r(301),r(300),r(343),r(369),r(372),r(371),r(373),r(374),r(370),r(375),r(376),r(354),r(357),r(353),r(351),r(352),r(355),r(356),r(346),r(368),r(377),r(345),r(347),r(349),r(348),r(350),r(359),r(360),r(362),r(361),r(364),r(363),r(365),r(366),r(367),r(344),r(358),r(380),r(379),r(378),t.exports=r(52)},function(t,n){function r(t,n){if("string"==typeof n)return t.insertAdjacentHTML("afterend",n);var r=t.nextSibling;return r?t.parentNode.insertBefore(n,r):t.parentNode.appendChild(n)}t.exports=r},,,,,,,,,function(t,n,r){(function(n,r){!function(n){"use strict";function e(t,n,r,e){var i=n&&n.prototype instanceof o?n:o,u=Object.create(i.prototype),c=new p(e||[]);return u._invoke=s(t,r,c),u}function i(t,n,r){try{return{type:"normal",arg:t.call(n,r)}}catch(t){return{type:"throw",arg:t}}}function o(){}function u(){}function c(){}function f(t){["next","throw","return"].forEach(function(n){t[n]=function(t){return this._invoke(n,t)}})}function a(t){function n(r,e,o,u){var c=i(t[r],t,e);if("throw"!==c.type){var f=c.arg,a=f.value;return a&&"object"==typeof a&&m.call(a,"__await")?Promise.resolve(a.__await).then(function(t){n("next",t,o,u)},function(t){n("throw",t,o,u)}):Promise.resolve(a).then(function(t){f.value=t,o(f)},u)}u(c.arg)}function e(t,r){function e(){return new Promise(function(e,i){n(t,r,e,i)})}return o=o?o.then(e,e):e()}"object"==typeof r&&r.domain&&(n=r.domain.bind(n));var o;this._invoke=e}function s(t,n,r){var e=P;return function(o,u){if(e===F)throw new Error("Generator is already running");if(e===M){if("throw"===o)throw u;return y()}for(r.method=o,r.arg=u;;){var c=r.delegate;if(c){var f=l(c,r);if(f){if(f===A)continue;return f}}if("next"===r.method)r.sent=r._sent=r.arg;else if("throw"===r.method){if(e===P)throw e=M,r.arg;r.dispatchException(r.arg)}else"return"===r.method&&r.abrupt("return",r.arg);e=F;var a=i(t,n,r);if("normal"===a.type){if(e=r.done?M:j,a.arg===A)continue;return{value:a.arg,done:r.done}}"throw"===a.type&&(e=M,r.method="throw",r.arg=a.arg)}}}function l(t,n){var r=t.iterator[n.method];if(r===g){if(n.delegate=null,"throw"===n.method){if(t.iterator.return&&(n.method="return",n.arg=g,l(t,n),"throw"===n.method))return A;n.method="throw",n.arg=new TypeError("The iterator does not provide a 'throw' method")}return A}var e=i(r,t.iterator,n.arg);if("throw"===e.type)return n.method="throw",n.arg=e.arg,n.delegate=null,A;var o=e.arg;return o?o.done?(n[t.resultName]=o.value,n.next=t.nextLoc,"return"!==n.method&&(n.method="next",n.arg=g),n.delegate=null,A):o:(n.method="throw",n.arg=new TypeError("iterator result is not an object"),n.delegate=null,A)}function h(t){var n={tryLoc:t[0]};1 in t&&(n.catchLoc=t[1]),2 in t&&(n.finallyLoc=t[2],n.afterLoc=t[3]),this.tryEntries.push(n)}function v(t){var n=t.completion||{};n.type="normal",delete n.arg,t.completion=n}function p(t){this.tryEntries=[{tryLoc:"root"}],t.forEach(h,this),this.reset(!0)}function d(t){if(t){var n=t[w];if(n)return n.call(t);if("function"==typeof t.next)return t;if(!isNaN(t.length)){var r=-1,e=function n(){for(;++r<t.length;)if(m.call(t,r))return n.value=t[r],n.done=!1,n;return n.value=g,n.done=!0,n};return e.next=e}}return{next:y}}function y(){return{value:g,done:!0}}var g,b=Object.prototype,m=b.hasOwnProperty,x="function"==typeof Symbol?Symbol:{},w=x.iterator||"@@iterator",S=x.asyncIterator||"@@asyncIterator",_=x.toStringTag||"@@toStringTag",O="object"==typeof t,E=n.regeneratorRuntime;if(E)return void(O&&(t.exports=E));E=n.regeneratorRuntime=O?t.exports:{},E.wrap=e;var P="suspendedStart",j="suspendedYield",F="executing",M="completed",A={},N={};N[w]=function(){return this};var T=Object.getPrototypeOf,I=T&&T(T(d([])));I&&I!==b&&m.call(I,w)&&(N=I);var k=c.prototype=o.prototype=Object.create(N);u.prototype=k.constructor=c,c.constructor=u,c[_]=u.displayName="GeneratorFunction",E.isGeneratorFunction=function(t){var n="function"==typeof t&&t.constructor;return!!n&&(n===u||"GeneratorFunction"===(n.displayName||n.name))},E.mark=function(t){return Object.setPrototypeOf?Object.setPrototypeOf(t,c):(t.__proto__=c,_ in t||(t[_]="GeneratorFunction")),t.prototype=Object.create(k),t},E.awrap=function(t){return{__await:t}},f(a.prototype),a.prototype[S]=function(){return this},E.AsyncIterator=a,E.async=function(t,n,r,i){var o=new a(e(t,n,r,i));return E.isGeneratorFunction(n)?o:o.next().then(function(t){return t.done?t.value:o.next()})},f(k),k[_]="Generator",k.toString=function(){return"[object Generator]"},E.keys=function(t){var n=[];for(var r in t)n.push(r);return n.reverse(),function r(){for(;n.length;){var e=n.pop();if(e in t)return r.value=e,r.done=!1,r}return r.done=!0,r}},E.values=d,p.prototype={constructor:p,reset:function(t){if(this.prev=0,this.next=0,this.sent=this._sent=g,this.done=!1,this.delegate=null,this.method="next",this.arg=g,this.tryEntries.forEach(v),!t)for(var n in this)"t"===n.charAt(0)&&m.call(this,n)&&!isNaN(+n.slice(1))&&(this[n]=g)},stop:function(){this.done=!0;var t=this.tryEntries[0],n=t.completion;if("throw"===n.type)throw n.arg;return this.rval},dispatchException:function(t){function n(n,e){return o.type="throw",o.arg=t,r.next=n,e&&(r.method="next",r.arg=g),!!e}if(this.done)throw t;for(var r=this,e=this.tryEntries.length-1;e>=0;--e){var i=this.tryEntries[e],o=i.completion;if("root"===i.tryLoc)return n("end");if(i.tryLoc<=this.prev){var u=m.call(i,"catchLoc"),c=m.call(i,"finallyLoc");if(u&&c){if(this.prev<i.catchLoc)return n(i.catchLoc,!0);if(this.prev<i.finallyLoc)return n(i.finallyLoc)}else if(u){if(this.prev<i.catchLoc)return n(i.catchLoc,!0)}else{if(!c)throw new Error("try statement without catch or finally");if(this.prev<i.finallyLoc)return n(i.finallyLoc)}}}},abrupt:function(t,n){for(var r=this.tryEntries.length-1;r>=0;--r){var e=this.tryEntries[r];if(e.tryLoc<=this.prev&&m.call(e,"finallyLoc")&&this.prev<e.finallyLoc){var i=e;break}}i&&("break"===t||"continue"===t)&&i.tryLoc<=n&&n<=i.finallyLoc&&(i=null);var o=i?i.completion:{};return o.type=t,o.arg=n,i?(this.method="next",this.next=i.finallyLoc,A):this.complete(o)},complete:function(t,n){if("throw"===t.type)throw t.arg;return"break"===t.type||"continue"===t.type?this.next=t.arg:"return"===t.type?(this.rval=this.arg=t.arg,this.method="return",this.next="end"):"normal"===t.type&&n&&(this.next=n),A},finish:function(t){for(var n=this.tryEntries.length-1;n>=0;--n){var r=this.tryEntries[n];if(r.finallyLoc===t)return this.complete(r.completion,r.afterLoc),v(r),A}},catch:function(t){for(var n=this.tryEntries.length-1;n>=0;--n){var r=this.tryEntries[n];if(r.tryLoc===t){var e=r.completion;if("throw"===e.type){var i=e.arg;v(r)}return i}}throw new Error("illegal catch attempt")},delegateYield:function(t,n,r){return this.delegate={iterator:d(t),resultName:n,nextLoc:r},"next"===this.method&&(this.arg=g),A}}}("object"==typeof n?n:"object"==typeof window?window:"object"==typeof self?self:this)}).call(n,function(){return this}(),r(158))}])</script><script src="/./main.0cf68a.js"></script><script>!function(){!function(e){var t=document.createElement("script");document.getElementsByTagName("body")[0].appendChild(t),t.setAttribute("src",e)}("/slider.e37972.js")}()</script>


    
<div class="tools-col" q-class="show:isShow,hide:isShow|isFalse" q-on="click:stop(e)">
  <div class="tools-nav header-menu">
    
    
      
      
      
    
      
      
      
    
      
      
      
    
    

    <ul style="width: 70%">
    
    
      
      <li style="width: 33.333333333333336%" q-on="click: openSlider(e, 'innerArchive')"><a href="javascript:void(0)" q-class="active:innerArchive">所有文章</a></li>
      
        
      
      <li style="width: 33.333333333333336%" q-on="click: openSlider(e, 'friends')"><a href="javascript:void(0)" q-class="active:friends">友链</a></li>
      
        
      
      <li style="width: 33.333333333333336%" q-on="click: openSlider(e, 'aboutme')"><a href="javascript:void(0)" q-class="active:aboutme">关于我</a></li>
      
        
    </ul>
  </div>
  <div class="tools-wrap">
    
    	<section class="tools-section tools-section-all" q-show="innerArchive">
        <div class="search-wrap">
          <input class="search-ipt" q-model="search" type="text" placeholder="find something…">
          <i class="icon-search icon" q-show="search|isEmptyStr"></i>
          <i class="icon-close icon" q-show="search|isNotEmptyStr" q-on="click:clearChose(e)"></i>
        </div>
        <div class="widget tagcloud search-tag">
          <p class="search-tag-wording">tag:</p>
          <label class="search-switch">
            <input type="checkbox" q-on="click:toggleTag(e)" q-attr="checked:showTags">
          </label>
          <ul class="article-tag-list" q-show="showTags">
            
            <div class="clearfix"></div>
          </ul>
        </div>
        <ul class="search-ul">
          <p q-show="jsonFail" style="padding: 20px; font-size: 12px;">
            缺失模块。<br/>1、请确保node版本大于6.2<br/>2、在博客根目录（注意不是yilia根目录）执行以下命令：<br/> npm i hexo-generator-json-content --save<br/><br/>
            3、在根目录_config.yml里添加配置：
<pre style="font-size: 12px;" q-show="jsonFail">
  jsonContent:
    meta: false
    pages: false
    posts:
      title: true
      date: true
      path: true
      text: false
      raw: false
      content: false
      slug: false
      updated: false
      comments: false
      link: false
      permalink: false
      excerpt: false
      categories: false
      tags: true
</pre>
          </p>
          <li class="search-li" q-repeat="items" q-show="isShow">
            <a q-attr="href:path|urlformat" class="search-title"><i class="icon-quo-left icon"></i><span q-text="title"></span></a>
            <p class="search-time">
              <i class="icon-calendar icon"></i>
              <span q-text="date|dateformat"></span>
            </p>
            <p class="search-tag">
              <i class="icon-price-tags icon"></i>
              <span q-repeat="tags" q-on="click:choseTag(e, name)" q-text="name|tagformat"></span>
            </p>
          </li>
        </ul>
    	</section>
    

    
    	<section class="tools-section tools-section-friends" q-show="friends">
  		
        <ul class="search-ul">
          
            <li class="search-li">
              <a href="http://localhost:4000/" target="_blank" class="search-title"><i class="icon-quo-left icon"></i>友情链接1</a>
            </li>
          
            <li class="search-li">
              <a href="http://localhost:4000/" target="_blank" class="search-title"><i class="icon-quo-left icon"></i>友情链接2</a>
            </li>
          
            <li class="search-li">
              <a href="http://localhost:4000/" target="_blank" class="search-title"><i class="icon-quo-left icon"></i>友情链接3</a>
            </li>
          
            <li class="search-li">
              <a href="http://localhost:4000/" target="_blank" class="search-title"><i class="icon-quo-left icon"></i>友情链接4</a>
            </li>
          
            <li class="search-li">
              <a href="http://localhost:4000/" target="_blank" class="search-title"><i class="icon-quo-left icon"></i>友情链接5</a>
            </li>
          
            <li class="search-li">
              <a href="http://localhost:4000/" target="_blank" class="search-title"><i class="icon-quo-left icon"></i>友情链接6</a>
            </li>
          
        </ul>
  		
    	</section>
    

    
    	<section class="tools-section tools-section-me" q-show="aboutme">
  	  	
  	  		<div class="aboutme-wrap" id="js-aboutme">很惭愧&lt;br&gt;&lt;br&gt;只做了一点微小的工作&lt;br&gt;谢谢大家</div>
  	  	
    	</section>
    
  </div>
  
</div>
    <!-- Root element of PhotoSwipe. Must have class pswp. -->
<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

    <!-- Background of PhotoSwipe. 
         It's a separate element as animating opacity is faster than rgba(). -->
    <div class="pswp__bg"></div>

    <!-- Slides wrapper with overflow:hidden. -->
    <div class="pswp__scroll-wrap">

        <!-- Container that holds slides. 
            PhotoSwipe keeps only 3 of them in the DOM to save memory.
            Don't modify these 3 pswp__item elements, data is added later on. -->
        <div class="pswp__container">
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
        </div>

        <!-- Default (PhotoSwipeUI_Default) interface on top of sliding area. Can be changed. -->
        <div class="pswp__ui pswp__ui--hidden">

            <div class="pswp__top-bar">

                <!--  Controls are self-explanatory. Order can be changed. -->

                <div class="pswp__counter"></div>

                <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>

                <button class="pswp__button pswp__button--share" style="display:none" title="Share"></button>

                <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>

                <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>

                <!-- Preloader demo http://codepen.io/dimsemenov/pen/yyBWoR -->
                <!-- element will get class pswp__preloader--active when preloader is running -->
                <div class="pswp__preloader">
                    <div class="pswp__preloader__icn">
                      <div class="pswp__preloader__cut">
                        <div class="pswp__preloader__donut"></div>
                      </div>
                    </div>
                </div>
            </div>

            <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
                <div class="pswp__share-tooltip"></div> 
            </div>

            <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
            </button>

            <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
            </button>

            <div class="pswp__caption">
                <div class="pswp__caption__center"></div>
            </div>

        </div>

    </div>

</div>
  </div>
</body>
</html>